<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="quarto-1.7.29" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />


<title>4  Plots of Multivariate Data – Visualizing Multivariate Data and Models in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<!-- htmldependencies:E3FAD763 -->
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn"
      data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <h1 class="quarto-secondary-nav-title no-breadcrumbs"></h1>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" role="link"
        aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
        onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="/">
      Visualizing Multivariate Data and Models in R
      </a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/friendly/vis-MLM-book"  title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/00-Author.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Author</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1"  role="navigation" aria-expanded="true">
 <span class="menu-text">Orienting Ideas</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/01-Prelude.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>1</span>  <span class='chapter-title'>Warm-up Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/02-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>2</span>  <span class='chapter-title'>Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/03-getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>3</span>  <span class='chapter-title'>Getting Started</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2"  role="navigation" aria-expanded="true">
 <span class="menu-text">Exploratory Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/04-multivariate_plots.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class='chapter-number'>4</span>  <span class='chapter-title'>Plots of Multivariate Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/05-pca-biplot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>5</span>  <span class='chapter-title'>Dimension Reduction</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3"  role="navigation" aria-expanded="true">
 <span class="menu-text">Univariate Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/06-linear_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>6</span>  <span class='chapter-title'>Overview of Linear models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/07-linear_models-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>7</span>  <span class='chapter-title'>Plots for Univariate Response Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/08-lin-mod-topics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>8</span>  <span class='chapter-title'>Topics in Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/09-collinearity-ridge.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>9</span>  <span class='chapter-title'>Collinearity & Ridge Regression</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4"  role="navigation" aria-expanded="true">
 <span class="menu-text">Multivariate Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/10-hotelling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>10</span>  <span class='chapter-title'>Hotelling's $T^2$</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/11-mlm-review.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>11</span>  <span class='chapter-title'>Multivariate Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/12-mlm-viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>12</span>  <span class='chapter-title'>Visualizing Multivariate Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/13-eqcov.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>13</span>  <span class='chapter-title'>Visualizing Equality of Covariance Matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/14-infl-robust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>14</span>  <span class='chapter-title'>Multiviate Influence and Robust Estimation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/15-case-studies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>15</span>  <span class='chapter-title'>Case Studies</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5"  role="navigation" aria-expanded="true">
 <span class="menu-text">End Matter</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/91-colophon.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Colophon</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/95-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6"  role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/21-discrim.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>A</span>  <span class='chapter-title'>Discriminant analysis</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" ></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <div id="quarto-toc-target"></div>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">


<header id="title-block-header" class="quarto-title-block default">
  <div class="quarto-title-banner">
    <div class="quarto-title column-body">
      <h1 class="title"><span id="sec-multivariate_plots" class="quarto-section-identifier"><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span></h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header>

<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-bivariate_summaries" id="toc-sec-bivariate_summaries"><span class="header-section-number">4.1</span> Bivariate summaries</a>
  <ul>
  <li><a href="#smoothers" id="toc-smoothers"><span class="header-section-number">4.1.1</span> Smoothers</a></li>
  <li><a href="#stratifiers" id="toc-stratifiers"><span class="header-section-number">4.1.2</span> Stratifiers</a></li>
  <li><a href="#conditioning" id="toc-conditioning"><span class="header-section-number">4.1.3</span> Conditioning</a></li>
  </ul></li>
  <li><a href="#sec-data-ellipse" id="toc-sec-data-ellipse"><span class="header-section-number">4.2</span> Data Ellipses</a>
  <ul>
  <li><a href="#drawing-data-ellipses" id="toc-drawing-data-ellipses"><span class="header-section-number">4.2.1</span> Drawing data ellipses</a></li>
  <li><a href="#ellipse-properties" id="toc-ellipse-properties"><span class="header-section-number">4.2.2</span> Ellipse properties</a></li>
  <li><a href="#r-functions-for-data-ellipses" id="toc-r-functions-for-data-ellipses"><span class="header-section-number">4.2.3</span> R functions for data ellipses</a></li>
  <li><a href="#sec-log-scale" id="toc-sec-log-scale"><span class="header-section-number">4.2.4</span> Handling nonlinearity: Plotting on a log scale</a></li>
  <li><a href="#sec-stratifying" id="toc-sec-stratifying"><span class="header-section-number">4.2.5</span> Stratifying</a></li>
  <li><a href="#sec-penguins" id="toc-sec-penguins"><span class="header-section-number">4.2.6</span> Meet the Penguins</a></li>
  <li><a href="#sec-visual-thinning" id="toc-sec-visual-thinning"><span class="header-section-number">4.2.7</span> Visual thinning</a></li>
  </ul></li>
  <li><a href="#sec-bagplots" id="toc-sec-bagplots"><span class="header-section-number">4.3</span> Bagplots</a></li>
  <li><a href="#sec-bivar-density" id="toc-sec-bivar-density"><span class="header-section-number">4.4</span> Non-parametric bivariate density plots</a></li>
  <li><a href="#sec-simpsons" id="toc-sec-simpsons"><span class="header-section-number">4.5</span> Simpson’s paradox: marginal and conditional relationships</a></li>
  <li><a href="#sec-multivar-normality" id="toc-sec-multivar-normality"><span class="header-section-number">4.6</span> Multivariate normality and outliers</a>
  <ul>
  <li><a href="#galton-data" id="toc-galton-data"><span class="header-section-number">4.6.1</span> Galton data</a></li>
  <li><a href="#sec-multnorm-penguin" id="toc-sec-multnorm-penguin"><span class="header-section-number">4.6.2</span> Penguin data</a></li>
  </ul></li>
  <li><a href="#sec-scatmat" id="toc-sec-scatmat"><span class="header-section-number">4.7</span> Scatterplot matrices</a>
  <ul>
  <li><a href="#visual-thinning" id="toc-visual-thinning"><span class="header-section-number">4.7.1</span> Visual thinning</a></li>
  </ul></li>
  <li><a href="#sec-corrgram" id="toc-sec-corrgram"><span class="header-section-number">4.8</span> Corrgrams</a></li>
  <li><a href="#sec-ggpairs" id="toc-sec-ggpairs"><span class="header-section-number">4.9</span> Generalized pairs plots</a></li>
  <li><a href="#sec-parcoord" id="toc-sec-parcoord"><span class="header-section-number">4.10</span> Parallel coordinate plots</a></li>
  <li><a href="#sec-animated-tours" id="toc-sec-animated-tours"><span class="header-section-number">4.11</span> Animated tours</a>
  <ul>
  <li><a href="#sec-projections" id="toc-sec-projections"><span class="header-section-number">4.11.1</span> Projections</a></li>
  <li><a href="#touring-methods" id="toc-touring-methods"><span class="header-section-number">4.11.2</span> Touring methods</a></li>
  </ul></li>
  <li><a href="#sec-network" id="toc-sec-network"><span class="header-section-number">4.12</span> Network diagrams</a>
  <ul>
  <li><a href="#crime-data" id="toc-crime-data"><span class="header-section-number">4.12.1</span> Crime data</a></li>
  <li><a href="#sec-partial-cor" id="toc-sec-partial-cor"><span class="header-section-number">4.12.2</span> Partial correlations</a></li>
  <li><a href="#sec-pvPlot" id="toc-sec-pvPlot"><span class="header-section-number">4.12.3</span> Visualizing partial correlations</a></li>
  </ul></li>
  <li><a href="#what-have-we-learned" id="toc-what-have-we-learned"><span class="header-section-number">4.13</span> What have we learned?</a></li>
  <li><a href="#exercises" id="toc-exercises"><span class="header-section-number">4.14</span> Exercises</a></li>
  </ul>
</nav>
<!--- For HTML Only --->
<!-- \require{newcommand} -->
<!-- %\renewcommand*{\det}[1]{\mathrm{det} (#1)} -->
<!-- %\renewcommand*{\det}[1]{|#1|} -->
<!-- \providecommand{\sizedmat}[2]{\mathord{\mathop{\mat{#1}}\limits_{(#2)}}} -->
<!-- \newcommand*{\E}{\mathcal{E}} -->
<!-- Index generation -->
<!-- % R packages:  indexed under both package name and packages! -->
<!-- % data sets:  -->
<!-- % R stuff -->
<blockquote>
<p><em>There is no excuse for failing to plot and look.</em> <em>The greatest value of a picture is when it forces us to notice what we never expected to see.</em> — John W. Tukey <span class="citation" data-cites="Tukey:77">(<a href="#ref-Tukey:77" role="doc-biblioref">1977</a>)</span>, <em>Exploratory Data Analysis</em> </p>
</blockquote>
<p>These quotes from John Tukey remind us that data analysis should nearly always start with graphs to help us understand the main features of our data. It is important to understand the general <em>patterns</em> and <em>trends</em>: Are relationships increasing or decreasing? Are they approximately linear or non-linear? But it is also important to spot <em>anomalies</em>: “unusual” observations, groups of points that seem to differ from the rest, and so forth. As we saw with Anscombe’s quartet (<a href="#sec-anscombe" class="quarto-xref"><span class="quarto-unresolved-ref">sec-anscombe</span></a>) and the Davis weight data (<a href="#sec-davis" class="quarto-xref"><span class="quarto-unresolved-ref">sec-davis</span></a>) numerical summaries hide features that are immediately apparent in a plot. </p>
<p>This chapter introduces a toolbox of basic graphical methods for visualizing multivariate datasets. It starts with some simple techniques to enhance the basic scatterplot with graphical <em>annotations</em> such as fitted lines, curves (<a href="#sec-bivariate_summaries" class="quarto-xref"><span class="quarto-unresolved-ref">sec-bivariate_summaries</span></a>) and data ellipses (<a href="#sec-data-ellipse" class="quarto-xref"><span class="quarto-unresolved-ref">sec-data-ellipse</span></a>) which serve to add <em>visual summaries</em> of the relation between two variables.</p>
<p>To visualize more than two variables, we can view all pairs of variables in a scatterplot matrix (<a href="#sec-scatmat" class="quarto-xref"><span class="quarto-unresolved-ref">sec-scatmat</span></a>) or shift gears entirely to show multiple variables along a set of parallel axes (<a href="#sec-parcoord" class="quarto-xref"><span class="quarto-unresolved-ref">sec-parcoord</span></a>).</p>
<p>As the number of variables increases, we may need to suppress details with stronger summaries (<a href="#sec-visual-thinning" class="quarto-xref"><span class="quarto-unresolved-ref">sec-visual-thinning</span></a>) for a high-level reconnaissance of our data terrain, as we do by zooming out on a map. For example, we can simply remove the data points or make them nearly transparent to focus on the visual summaries provided by fitted lines or other graphical summaries. </p>
<p>Another approach to visualizing high-D data uses <em>animated tours</em> (<a href="#sec-animated-tours" class="quarto-xref"><span class="quarto-unresolved-ref">sec-animated-tours</span></a>) to show the data in a sequence of low-D projections, typically 2D, designed to reveal interesting features that might not otherwise be visible. At a higher level of abstraction, <em>network diagrams</em> (<a href="#sec-network" class="quarto-xref"><span class="quarto-unresolved-ref">sec-network</span></a>) representing the structure of correlations among a possibly large set of variables offer another useful technique.</p>
<p><strong>Packages</strong></p>
<p>In this chapter I use the following packages. Load them now:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrgram)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdensity)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpcp)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tourr)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(heplots)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gggda)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># set basic ggplot theme</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span></code></pre></div>
</div>
<!--

-->
<section id="sec-bivariate_summaries" class="level2" data-number="4.1">
<h2 data-number="4.1"><span class="header-section-number">4.1</span> Bivariate summaries</h2>
<p>The basic scatterplot is the workhorse of multivariate data visualization, showing how one variable, <span class="math inline">\(y\)</span>, often an outcome to be explained by or varies with another, <span class="math inline">\(x\)</span>. It is a building block for many useful techniques, so it is helpful to understand how it can be used as a tool for thinking in a wider, multivariate context.</p>
<p>The essential idea is that we can start with a simple version of the scatterplot and add annotations to show interesting features more clearly. I consider the following here:</p>
<ul>
<li><strong>Smoothers</strong>: Showing overall trends, perhaps in several forms, as visual summaries such as fitted regression lines or curves and nonparametric smoothers.</li>
<li><strong>Stratifiers</strong>: Using color, shape or other features to identify subgroups; more generally, <em>conditioning</em> on other variables in multi-panel displays;</li>
<li><strong>Data ellipses</strong>: A compact 2D visual summary of bivariate linear relations and uncertainty assuming normality; more generally, contour plots of bivariate density. </li>
</ul>
<div id="exm-salaries1" class="theorem example">
<p><span class="theorem-title"><strong>Example 4.1</strong></span> <strong>Academic salaries</strong></p>
<p>Let’s start with data on the academic salaries of faculty members collected at a U.S. college for the purpose of assessing salary differences between male and female faculty members, and perhaps address anomalies in compensation. The dataset <code>carData::Salaries</code> gives data on nine-month salaries and other variables for 397 faculty members in the 2008-2009 academic year.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Salaries, <span class="at">package =</span> <span class="st">&quot;carData&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(Salaries)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;data.frame&#39;: 397 obs. of  6 variables:</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ rank         : Factor w/ 3 levels &quot;AsstProf&quot;,&quot;AssocProf&quot;,..: 3 3 1 3 3 2 3 3 3 3 ...</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ discipline   : Factor w/ 2 levels &quot;A&quot;,&quot;B&quot;: 2 2 2 2 2 2 2 2 2 2 ...</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ yrs.since.phd: int  19 20 4 45 40 6 30 45 21 18 ...</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ yrs.service  : int  18 16 3 39 41 6 23 45 20 18 ...</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ sex          : Factor w/ 2 levels &quot;Female&quot;,&quot;Male&quot;: 2 2 2 2 2 2 2 2 2 1 ...</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ salary       : int  139750 173200 79750 115000 141500 97000 175000 147765 119250 129000 ...</span></span></code></pre></div>
</div>
<p>The most obvious, but perhaps naive, predictor of <code>salary</code> is <code>years.since.phd</code>. For simplicity, I’ll refer to this as years of “experience.” Before looking at differences between males and females, we would want consider faculty <code>rank</code> (related also to <code>yrs.service</code>) and <code>discipline</code>, recorded here as <code>"A"</code> (“theoretical” departments) or <code>"B"</code> (“applied” departments). But, for a basic plot, we will ignore these for now to focus on what can be learned from plot annotations.</p>
<!-- fig.code: `R/Salaries-scatterplots.R`  -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Salaries, </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> yrs.since.phd, <span class="at">y =</span> salary)) <span class="sc">+</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">dollar_format</span>(</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">prefix=</span><span class="st">&quot;$&quot;</span>, <span class="at">scale =</span> <span class="fl">0.001</span>, <span class="at">suffix =</span> <span class="st">&quot;K&quot;</span>)) <span class="sc">+</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Years since PhD&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Salary&quot;</span>) </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> <span class="fu">geom_rug</span>(<span class="at">position =</span> <span class="st">&quot;jitter&quot;</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Salaries-scat" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-scat-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-scat-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-scat-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.1: Naive scatterplot of Salary vs. years since PhD, ignoring other variables, and without graphical annotations.
</figcaption>
</figure>
</div>
</div>
</div>
<p>There is quite a lot we can see “just by looking” at this simple plot, but the main things are:</p>
<ul>
<li>Salary increases generally from 0 - 40 years since the PhD, but then maybe begins to drop off (partial retirement?);</li>
<li>Variability in salary increases among those with the same experience, a “fan-shaped” pattern that signals a violation of homogeneity of variance in simple regression;</li>
<li>Data beyond 50 years is thin, but there are some quite low salaries there. Adding rug plots to the X and Y axes is a simple but effective way to show the marginal distributions of the observations. Jitter and transparency helps to avoid overplotting due to discrete values.</li>
</ul>
</div>
<section id="smoothers" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1"><span class="header-section-number">4.1.1</span> Smoothers</h3>
<p>Smoothers are among the most useful graphical annotations you can add to such plots, giving a visual summary of how <span class="math inline">\(y\)</span> changes with <span class="math inline">\(x\)</span>. The most common smoother is a line showing the linear regression for <span class="math inline">\(y\)</span> given <span class="math inline">\(x\)</span>, expressed in math notation as <span class="math inline">\(\mathbb{E} (y | x) = b_0 + b_1 x\)</span>. If there is doubt that a linear relation is an adequate summary, you can try a quadratic or other polynomial smoothers.</p>
<p>In <span style="color: brown;"><strong>ggplot2</strong></span>, these are easily added to a plot using <code>geom_smooth()</code> with <code>method = "lm"</code>, and a model <code>formula</code>, which (by default) is <code>y ~ x</code> for a linear relation or <code>y ~ poly(x, k)</code> for a polynomial of degree <span class="math inline">\(k\)</span>.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode" id="cb4" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill=</span> <span class="st">&quot;pink&quot;</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ poly(x,2)&quot;</span>, </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&quot;darkgreen&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;lightgreen&quot;</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>) </span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-lm" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-lm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-lm-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-lm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.2: Scatterplot of Salary vs. years since PhD, showing <span style="color: red;">linear</span> and <span style="color: darkgreen;">quadratic</span> smooths with 95% confidence bands.
</figcaption>
</figure>
</div>
</div>
</div>
<!--# UA: This is a fantastic graph. My only (very minor) suggestion is to replace one of the colours because the most common type of colour vision deficiency makes it hard to tell the difference between red and green. So, maybe green and purple (purple would match the inline code highlight colour)-->
<p>This serves to highlight some of our impressions from the basic scatterplot shown in <a href="#fig-Salaries-scat" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-scat</span></a>, making them more apparent. And that’s precisely the point: the regression smoother draws attention to a possible pattern that we can consider as a visual summary of the data. You can think of this as showing what a linear (or quadratic) regression “sees” in the data. Statistical tests <!--# (secref?) --> can help you decide if there is more evidence for a quadratic fit compared to the simpler linear relation. <!--# UA: Great paragraph!--></p>
<p>It is useful to also show some indication of <em>uncertainty</em> (or inversely, <em>precision</em>) associated with the predicted values. Both the linear and quadratic trends are shown in <a href="#fig-Salaries-lm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-lm</span></a> with 95% pointwise confidence bands.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> These are necessarily narrower in the center of the range of <span class="math inline">\(x\)</span> where there is typically more data; they get wider toward the highest values of experience where the data are thinner. </p>
<section id="non-parametric-smoothers" class="level4 unnumbered">
<h4 class="unnumbered">Non-parametric smoothers</h4>
<p>The most generally useful idea is a smoother that tracks an average value, <span class="math inline">\(\mathbb{E} (y | x)\)</span>, of <span class="math inline">\(y\)</span> as <span class="math inline">\(x\)</span> varies across its’ range <em>without</em> assuming any particular functional form, and so avoiding the necessity to choose among <code>y ~ poly(x, 1)</code>, or <code>y ~ poly(x, 2)</code>, or <code>y ~ poly(x, 3)</code>, etc.</p>
<p>Non-parametric smoothers attempt to estimate <span class="math inline">\(\mathbb{E} (y | x) = f(x)\)</span> where <span class="math inline">\(f(x)\)</span> is some smooth function. These typically use a collection of weighted <em>local regressions</em> for each <span class="math inline">\(x_i\)</span> within a window centered at that value. In the method called <em>lowess</em> or <em>loess</em> <span class="citation" data-cites="Cleveland:79 ClevelandDevlin:88">(<a href="#ref-Cleveland:79" role="doc-biblioref">Cleveland, 1979</a>; <a href="#ref-ClevelandDevlin:88" role="doc-biblioref">Cleveland &amp; Devlin, 1988</a>)</span>, a weight function is applied, giving greatest weight to <span class="math inline">\(x_i\)</span> and a weight of 0 outside a window containing a certain fraction, <span class="math inline">\(s\)</span>, called <em>span</em>, of the nearest neighbors of <span class="math inline">\(x_i\)</span>. The fraction, <span class="math inline">\(s\)</span>, is usually within the range <span class="math inline">\(1/3 \le s \le 2/3\)</span>, and it determines the smoothness of the resulting curve; smaller values produce a wigglier curve and larger values giving a smoother fit (an optimal span can be determined by <span class="math inline">\(k\)</span>-fold cross-validation to minimize a measure of overall error of approximation).</p>
<p>Non-parametric regression is a broad topic; see <span class="citation" data-cites="Fox:2016:ARA">Fox (<a href="#ref-Fox:2016:ARA" role="doc-biblioref">2016</a>)</span>, Ch. 18 for a more general treatment including smoothing splines, and <span class="citation" data-cites="Wood:2006">Wood (<a href="#ref-Wood:2006" role="doc-biblioref">2006</a>)</span> for generalized additive models, fit using <code>method = "gam"</code> in <strong>ggplot2</strong>, which is the default when the largest group has more than 1,000 observations.</p>
<p><a href="#fig-Salaries-loess" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-loess</span></a> shows the addition of a loess smooth to the plot in <a href="#fig-Salaries-lm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-lm</span></a>, suppressing the confidence band for the linear regression. The loess fit is nearly coincident with the quadratic fit, but has a slightly wider confidence band.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb5" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>gg1 <span class="sc">+</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;loess&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">fill =</span> scales<span class="sc">::</span><span class="fu">muted</span>(<span class="st">&quot;blue&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, <span class="at">se =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&quot;red&quot;</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ poly(x,2)&quot;</span>, </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&quot;darkgreen&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;lightgreen&quot;</span>,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>) </span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-loess" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-loess-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-loess-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-loess-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.3: Scatterplot of Salary vs. years since PhD, adding the loess smooth. The loess smooth curve and confidence band in <span style="color: green;">green</span> is nearly indistinguishable from a quadratic fit in <span style="color: blue;">blue</span>.
</figcaption>
</figure>
</div>
</div>
</div>
<p>But now comes an important question: is it reasonable that academic salary should increase up to about 40 years since the PhD degree and then decline? The predicted salary for someone still working 50 years after earning their degree is about the same as a person at 15 years. What else is going on here?</p>
</section>
</section>
<section id="stratifiers" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2"><span class="header-section-number">4.1.2</span> Stratifiers</h3>
<p>Very often, we have a main relationship of interest, but various groups in the data are identified by discrete factors (like faculty <code>rank</code> and <code>sex</code>, their type of <code>discipline</code> here), or there are quantitative predictors for which the main relation might vary. In the language of statistical models such effects are <em>interaction</em> terms, as in <code>y ~ group + x + group:x</code>, where the term <code>group:x</code> fits a different slope for each group and the grouping variable is often called a <em>moderator</em> variable. Common moderator variables are ethnicity, health status, social class and level of education. Moderators can also be continuous variables as in <code>y ~ x1 + x2 + x1:x2</code>.</p>
<p>I call these <em>stratifiers</em>, recognizing that we should consider breaking down the overall relation to see whether and how it changes over such “other” variables. Such variables are most often factors, but we can cut a continuous variable into ranges (<em>shingles</em>) and do the same graphically. There are two general graphical techniques for stratifying:</p>
<ul>
<li><p><strong>Grouping</strong>: Identify subgroups in the data by assigning different visual attributes, such as color, shape, line style, etc. within a single plot, as shown in <a href="#fig-Salaries-rank" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-rank</span></a> below. This is quite natural for factors; quantitative predictors can be accommodated by cutting their range into ordered intervals. Grouping has the advantage that the levels of a grouping variable can be shown within the same plot, facilitating direct comparison.</p></li>
<li><p><strong>Conditioning</strong>: Showing subgroups in different plot panels, as in <a href="#fig-Salaries-faceted" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-faceted</span></a>. This has the advantages that relations for the individual groups more easily discerned and one can easily stratify by two (or more) other variables jointly. But visual comparison is more difficult because the eye must scan from one panel to another.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="History Corner: Coplots and faceting">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
History Corner: Coplots and faceting
</div>
</div>
<div class="callout-body-container callout-body">
<p>The syntax you use in specifying a plot plays an important role in visual thinking. Ideally, you want the shortest path between having a graphic idea in your head and specifying that in code to see the result rendered on your screen. </p>
<p>Recognition of the roles of visual grouping by factors within a panel and conditioning in multi-panel displays was an important advance in the development of modern statistical graphics software. It began at A.T.&amp; T. Bell Labs in Murray Hill, NJ in conjunction with the <strong>S</strong> language, the mother of R.</p>
<p>Conditioning displays (originally called <em>coplots</em> <span class="citation" data-cites="ChambersHastie1991">(<a href="#ref-ChambersHastie1991" role="doc-biblioref">Chambers &amp; Hastie, 1991</a>)</span>) are simply a collection of 1D, 2D or 3D plots separate panels for subsets of the data broken down by one or more factors, or, for quantitative variables, subdivided into a factor with several overlapping intervals (<em>shingles</em>). The first implementation was in <em>Trellis</em> plots <span class="citation" data-cites="Becker:1996:VDC Cleveland:85">(<a href="#ref-Becker:1996:VDC" role="doc-biblioref">Becker et al., 1996</a>; <a href="#ref-Cleveland:85" role="doc-biblioref">Cleveland, 1985</a>)</span>. </p>
<p>Trellis displays were extended in the <span style="color: brown;"><strong>lattice</strong></span> package <span class="citation" data-cites="R-lattice">(<a href="#ref-R-lattice" role="doc-biblioref">Sarkar, 2008</a>)</span>, which offered:</p>
<ul>
<li>A <strong>graphing syntax</strong> similar to that used in statistical model formulas: <code>y ~ x | g</code> conditions the data by the levels of <code>g</code>, with <code>|</code> read as “given”; two or more conditioning are specified as <code>y ~ x | g1 + g2 + ...</code>, with <code>+</code> read as “and”.</li>
<li><strong>Panel functions</strong> define what is plotted in a given panel. <code>panel.xyplot()</code> is the default for scatterplots, plotting points, but you can add <code>panel.lmline()</code> for regression lines, <code>latticeExtra::panel.smoother()</code> for loess smooths and a wide variety of others.</li>
</ul>
<p>The <span style="color: brown;"><strong>car</strong></span> package <span class="citation" data-cites="R-car">(<a href="#ref-R-car" role="doc-biblioref">Fox &amp; Weisberg, 2019</a>)</span> supports this graphing syntax in many of its functions. <span style="color: brown;"><strong>ggplot2</strong></span> does not; it uses aesthetics (<code>aes()</code>), which map variables in the data to visual characteristics in displays.</p>
</div>
</div>
<p><strong>Using grouping</strong> For the <code>Salaries</code> data, the most obvious variable that affects academic salary is <code>rank</code>, because faculty typically get an increase in salary with a promotion that carries through in their future salary. What can we see if we group by <code>rank</code> and fit a separate smoothed curve for each?</p>
<p>In <code>ggplot2</code> thinking, grouping is accomplished simply by adding an aesthetic, such as <code>color = rank</code>. What happens then is that points, lines, smooths and other <code>geom_*()</code> inherit the feature that they are differentiated by color. In the case of <code>geom_smooth()</code>, we get a separate fit for each subset of the data, according to <code>rank</code>.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb6" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make some re-useable pieces to avoid repetitions</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>scale_salary <span class="ot">&lt;-</span>   <span class="fu">scale_y_continuous</span>(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">dollar_format</span>(<span class="at">prefix=</span><span class="st">&quot;$&quot;</span>, </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">scale =</span> <span class="fl">0.001</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">suffix =</span> <span class="st">&quot;K&quot;</span>)) </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># position the legend inside the plot</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>legend_pos <span class="ot">&lt;-</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;inside&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">legend.position.inside =</span> <span class="fu">c</span>(.<span class="dv">1</span>, <span class="fl">0.95</span>), </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">legend.justification =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Salaries, </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> yrs.since.phd, <span class="at">y =</span> salary, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> rank, <span class="at">shape =</span> rank)) <span class="sc">+</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  scale_salary <span class="sc">+</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Years since PhD&quot;</span>,</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Salary&quot;</span>) <span class="sc">+</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">fill =</span> rank),</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>                  <span class="at">method =</span> <span class="st">&quot;loess&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, </span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>                  <span class="at">linewidth =</span> <span class="dv">2</span>)  <span class="sc">+</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  legend_pos</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-rank" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-rank-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-rank-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-rank-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.4: Scatterplot of Salary vs. years since PhD, grouped by rank.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Well, there is a different story here. Salaries generally occupy separate vertical levels, increasing with academic rank. The horizontal extents of the smoothed curves show their ranges. Within each rank there is some initial increase after promotion, and then some tendency to decline with increasing years. But, by and large, years since the PhD doesn’t make as much difference once we’ve taken academic rank into account.</p>
<p>What about the <code>discipline</code> which is classified, perhaps peculiarly, as “theoretical” vs. “applied”? The values are just <code>"A"</code> and <code>"B"</code>, so I map these to more meaningful labels before making the plot.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb7" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>Salaries <span class="ot">&lt;-</span> Salaries <span class="sc">|&gt;</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">discipline =</span> </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>           <span class="fu">factor</span>(discipline, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;A: Theoretical&quot;</span>, <span class="st">&quot;B: Applied&quot;</span>)))</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>Salaries <span class="sc">|&gt;</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> yrs.since.phd, <span class="at">y =</span> salary, <span class="at">color =</span> discipline)) <span class="sc">+</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  scale_salary <span class="sc">+</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">fill =</span> discipline ),</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">method =</span> <span class="st">&quot;loess&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Years since PhD&quot;</span>,</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Salary&quot;</span>) <span class="sc">+</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  legend_pos </span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-discipline" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-discipline-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-discipline-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-discipline-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.5: Scatterplot of Salary vs. years since PhD, grouped by discipline.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The story in <a href="#fig-Salaries-discipline" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-discipline</span></a> is again different. Faculty in applied disciplines on average earn about 10,000$ more per year on average than their theoretical colleagues.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>Salaries <span class="sc">|&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(discipline) <span class="sc">|&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean =</span> <span class="fu">mean</span>(salary)) </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # A tibble: 2 × 2</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   discipline        mean</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   &lt;fct&gt;            &lt;dbl&gt;</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 A: Theoretical 108548.</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 B: Applied     118029.</span></span></code></pre></div>
</div>
<p>For both groups, there is an approximately linear relation up to about 30–40 years, but the smoothed curves then diverge into the region where the data is thinner.</p>
<p>This result is more surprising than differences among faculty ranks. The effect of annotation with smoothed curves as visual summaries is apparent, and provides a stimulus to think about <em>why</em> these differences (if they are real) exist between theoretical and applied professors, and maybe <em>should</em> theoreticians be paid more!</p>
</section>
<section id="conditioning" class="level3" data-number="4.1.3">
<h3 data-number="4.1.3"><span class="header-section-number">4.1.3</span> Conditioning</h3>
<p>The previous plots use grouping by color to plot the data for different subsets inside the same plot window, making comparison among groups easier, because they can be directly compared along a common vertical scale <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. This gets messy, however, when there are more than just a few levels, or worse—when there are two (or more) variables for which we want to show separate effects. In such cases, we can plot separate panels using the <code>ggplot2</code> concept of <em>faceting</em>. There are two options: <code>facet_wrap()</code> takes one or more conditioning variables and produces a ribbon of plots for each combination of levels; <code>facet_grid(row ~ col)</code> takes two or more conditioning variables and arranges the plots in a 2D array identified by the <code>row</code> and <code>col</code> variables.</p>
<p>Let’s look at salary broken down by the combinations of discipline and rank. Here, I chose to stratify using color by rank within each of panels faceting by discipline. Because there is more going on in this plot, a linear smooth is used to represent the trend.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb9" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>Salaries <span class="sc">|&gt;</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> yrs.since.phd, <span class="at">y =</span> salary, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> rank, <span class="at">shape =</span> rank)) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  scale_salary <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Years since PhD&quot;</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Salary&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">fill =</span> rank),</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>, </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> discipline) <span class="sc">+</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  legend_pos</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-faceted" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-faceted-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-faceted-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-faceted-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.6: Scatterplot of Salary vs. years since PhD, grouped by rank, with separate panels for discipline.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Once both of these factors are taken into account, there does not seem to be much impact of years of service. Salaries in theoretical disciplines are noticeably greater than those in applied disciplines at all ranks, and there are even greater differences among ranks.</p>
<p>Finally, to shed light on the question that motivated this example— are there anomalous differences in salary for men and women— we can look at differences in salary according to sex, when discipline and rank are taken into account. To do this graphically, condition by both variables, but use <code>facet_grid(discipline ~ rank)</code> to arrange their combinations in a grid whose rows are the levels of <code>discipline</code> and columns are those of <code>rank</code>. I want to make the comparison of males and females most direct, so I use <code>color = sex</code> to stratify the panels. The smoothed regression lines and error bands are calculated separately for each combination of discipline, rank and sex.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb10" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Salaries <span class="sc">|&gt;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> yrs.since.phd, <span class="at">y =</span> salary, <span class="at">color =</span> sex)) <span class="sc">+</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  scale_salary <span class="sc">+</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Years since PhD&quot;</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Salary&quot;</span>) <span class="sc">+</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="fu">aes</span>(<span class="at">fill =</span> sex),</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;y ~ x&quot;</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(discipline <span class="sc">~</span> rank) <span class="sc">+</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  legend_pos</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-Salaries-facet-sex" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Salaries-facet-sex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Salaries-facet-sex-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Salaries-facet-sex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.7: Scatterplot of Salary vs. years since PhD, grouped by sex, faceted by discipline and rank.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="sec-data-ellipse" class="level2" data-number="4.2">
<h2 data-number="4.2"><span class="header-section-number">4.2</span> Data Ellipses</h2>
<p>The <em>data ellipse</em> <span class="citation" data-cites="Monette:90">(<a href="#ref-Monette:90" role="doc-biblioref">Monette, 1990</a>)</span>, or <em>concentration ellipse</em> <span class="citation" data-cites="Dempster:69">(<a href="#ref-Dempster:69" role="doc-biblioref">Dempster, 1969</a>)</span> is a remarkably simple and effective display for viewing and understanding bivariate relationships in multivariate data. The data ellipse is typically used to add a visual summary to a scatterplot, that shows all together the means, standard deviations, correlation, and slope of the regression line for two variables, perhaps stratified by another variable.</p>
<p>Under the classical assumption that the data are bivariate normally distributed, the data ellipse is also a <strong>sufficient</strong> visual summary, in the sense that it captures <strong>all</strong> relevant features of the data. See <span class="citation" data-cites="Friendly-etal:ellipses:2013">Friendly et al. (<a href="#ref-Friendly-etal:ellipses:2013" role="doc-biblioref">2013</a>)</span> for a complete discussion of the role of ellipsoids in statistical data visualization.</p>
<p>The data ellipse is based on the idea that in a bivariate normal distribution, the contours of equal probability form a series of concentric ellipses. If the variables were uncorrelated and had the same variances, these would be circles, and Euclidean distance would measure the distance of each observation from the mean. When the variables are correlated, a different measure, <em>Mahalanobis distance</em> is the proper measure of how far a point is from the mean, taking the correlation into account. </p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-mahalanobis" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-mahalanobis-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/mahalanobis.png" class="img-fluid quarto-figure quarto-figure-center" style="width:50.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mahalanobis-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.8: 2D data with curves of constant distance from the centroid. The blue solid ellipse shows a contour of constant Mahalanobis distance, taking the correlation into account; the dashed red circle is a contour of equal Euclidean distance. Given the data points, which of the points <strong>A</strong> and <strong>B</strong> is further from the mean (X)? <em>Source</em>: Re-drawn from <a href="https://ouzhang.rbind.io/2020/11/16/outliers-part4/">Ou Zhang</a>
</figcaption>
</figure>
</div>
</div>
</div>
<!--
This doesn't work

#| fig-cap: !expr glue::glue("2D data with curves of constant distance from the centroid. The {blue} solid ellipse shows a contour of constant Mahalanobis distance, taking the correlation into account; the dashed {red} circle is a contour of equal Euclidean distance. Given the data points,  which of the points **A** and **B** is further from the mean (X)? _Source_: Re-drawn from [Ou Zhang](https://ouzhang.rbind.io/2020/11/16/outliers-part4/)")

#| fig-cap: !expr paste("2D data with curves of constant distance from the centroid. The", colorize('blue'), "solid ellipse shows a contour of constant Mahalanobis distance, taking the correlation into account; the dashed", colorize('blue'), "circle is a contour of equal Euclidean distance. Given the data points,  Which of the points **A** and **B** is further from the mean (X)? _Source_: Re-drawn from [Ou Zhang](https://ouzhang.rbind.io/2020/11/16/outliers-part4/)")
-->
<p>To illustrate, <a href="#fig-mahalanobis" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-mahalanobis</span></a> shows a scatterplot with labels for two points, “A” and “B”. Which is further from the mean, “X”? A contour of constant Euclidean distance, shown by the <span style="color: red;">red</span> dashed circle, ignores the apparent negative correlation, so point “A” is further. The <span style="color: blue;">blue</span> ellipse for Mahalanobis distance takes the correlation into account, so point “B” has a greater distance from the mean.</p>
<p>Mathematically, Euclidean (squared) distance (<span class="math inline">\(D_E^2(y)\)</span>) for <span class="math inline">\(p\)</span> variables, <span class="math inline">\(j = 1, 2, \dots , p\)</span>, is just a generalization of the square of a univariate standardized (<span class="math inline">\(z\)</span>) score, <span class="math inline">\(z^2 = [(y - \bar{y}) / s]^2\)</span>,</p>
<p><span class="math display">\[
D_E^2 (\mathbf{y}) = \sum_j^p z_j^2 = \mathbf{z}^\textsf{T}  \mathbf{z} = (\mathbf{y} - \bar{\mathbf{y}})^\textsf{T} \operatorname{diag}(\mathbf{S})^{-1} (\mathbf{y} - \bar{\mathbf{y}}) \; ,
\]</span> where <span class="math inline">\(\mathbf{S}\)</span> is the sample variance-covariance matrix, <span class="math inline">\(\mathbf{S} = ({n-1})^{-1} \sum_{i=1}^n (\mathbf{y}_i - \bar{\mathbf{y}})^\textsf{T} (\mathbf{y}_i - \bar{\mathbf{y}})\)</span>.</p>
<p>Mahalanobis distance takes the correlations into account simply by using the covariances as well as the variances,</p>
<p><span id="eq-Dsq"><span class="math display">\[
D_M^2 (\mathbf{y}) = (\mathbf{y} - \bar{\mathbf{y}})^\mathsf{T} S^{-1} (\mathbf{y} - \bar{\mathbf{y}}) \; .
\tag{4.1}\]</span></span></p>
<p>In <a href="#eq-Dsq" class="quarto-xref">Equation <span class="quarto-unresolved-ref">eq-Dsq</span></a>, the inverse <span class="math inline">\(S^{-1}\)</span> serves to “divide” the matrix <span class="math inline">\((\mathbf{y} - \bar{\mathbf{y}})^\mathsf{T} (\mathbf{y} - \bar{\mathbf{y}})\)</span> of squared distances by the variances (and covariances) of the variables, as in the univariate case.</p>
<p>For <span class="math inline">\(p\)</span> variables, the data <em>ellipsoid</em> <span class="math inline">\(\mathcal{E}_c\)</span> of size <span class="math inline">\(c\)</span> is a <span class="math inline">\(p\)</span>-dimensional ellipse, defined as the set of points <span class="math inline">\(\mathbf{y} = (y_1, y_2, \dots y_p)\)</span> whose squared Mahalanobis distance, <span class="math inline">\(D_M^2 ( \mathbf{y} )\)</span> is less than or equal to <span class="math inline">\(c^2\)</span>,</p>
<p><span class="math display">\[
\mathcal{E}_c (\bar{\mathbf{y}}, \mathbf{S}) := \{ D_M^2 (\mathbf{y}) \le c^2 \} \; .
\]</span></p>
<section id="drawing-data-ellipses" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1"><span class="header-section-number">4.2.1</span> Drawing data ellipses</h3>
<p>A computational definition of the data ellipsoid recognizes that the boundary of an ellipsoid can be found by (a) starting with a unit a unit sphere <span class="math inline">\(\mathcal{P}\)</span> centered at the origin, (b) transforming that by a “square root” of the covariance matrix, denoted <span class="math inline">\(\mathbf{S}^{1/2}\)</span>, and then (c) shifting that to centroid of the data.</p>
<p>The unit sphere is defined as the contour of Euclidean distance 1 from the origin, <span class="math inline">\(\mathcal{P} := \{ \mathbf{x}^\textsf{T} \mathbf{x}= 1\}\)</span>. Using the notation <span class="math inline">\(\oplus\)</span> to represent translation to a new centroid at the variable means, <span class="math inline">\(\bar{\mathbf{y}}\)</span>, the data ellipsoid becomes,</p>
<p><span class="math display">\[
\mathcal{E}_c (\bar{\mathbf{y}}, \mathbf{S}) = \bar{\mathbf{y}} \; \oplus \; \mathbf{S}^{1/2} \, \mathcal{P} \:\: .
\]</span> The matrix <span class="math inline">\(\mathbf{S}^{1/2}\)</span> represents a rotation and scaling of the sphere and is commonly computed as the Cholesky factor of <span class="math inline">\(\mathbf{S}\)</span> given in R by <code>chol(S)</code>. You can imagine this in 2D by thinking of how the dashed <span style="color: red;">red</span> circle in <a href="#fig-mahalanobis" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-mahalanobis</span></a> could be transformed into the <span style="color: blue;">blue</span> ellipse.</p>
<p>Slightly abusing notation and taking the unit sphere <span class="math inline">\(\mathcal{P}\)</span> like an identity matrix <span class="math inline">\(\mathbf{I}\)</span> that vanishes in multiplication, we can write the data ellipsoid as simply:</p>
<p><span id="eq-ellE"><span class="math display">\[
\mathcal{E}_c (\bar{\mathbf{y}}, \mathbf{S}) = \bar{\mathbf{y}} \; \oplus \; c\, \sqrt{\mathbf{S}} \:\: .
\tag{4.2}\]</span></span></p>
<p><strong>Contour levels (<span class="math inline">\(c\)</span>)</strong></p>
<p>When <span class="math inline">\(\mathbf{y}\)</span> is (at least approximately) bivariate normal, <span class="math inline">\(D_M^2(\mathbf{y})\)</span> has a large-sample <span class="math inline">\(\chi^2_2\)</span> distribution (<span class="math inline">\(\chi^2\)</span> with 2 df), so ellipses of various conventional sizes can be calculated using <code>qchisq()</code>:</p>
<ul>
<li><span class="math inline">\(c^2 = \chi^2_2 (0.5) = 1.39\)</span> gives a data ellipse covering 50% of the data points, a bivariate analog of the central box of a boxplot.</li>
<li><span class="math inline">\(c^2 = \chi^2_2 (0.68) = 2.28\)</span> gives a “1 standard deviation bivariate ellipse,” an analog of the standard interval <span class="math inline">\(\bar{y} \pm 1 s\)</span>,</li>
<li><span class="math inline">\(c^2 = \chi^2_2 (0.95) = 5.99 \approx 6\)</span> gives a data ellipse of 95% coverage.</li>
</ul>
<p>In not-large samples, the radius <span class="math inline">\(c\)</span> of the ellipsoid is better approximated by a multiple of a <span class="math inline">\(F_{p, n-p}\)</span> distribution, with <span class="math inline">\(p\)</span> variables and <span class="math inline">\(n-p\)</span> degrees of freedom for <span class="math inline">\(\mathbf{S}\)</span>. This gives a radius <span class="math inline">\(c =\sqrt{ 2 F_{2, n-2}^{1-\alpha} }\)</span> in the bivariate case (<span class="math inline">\(p=2\)</span>) for coverage <span class="math inline">\(1-\alpha\)</span>.</p>
<p>These three conventional cases are illustrated in <a href="#fig-ellipses-coverage" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-ellipses-coverage</span></a> with ellipses of 50%, 68% and 95% coverage for a the matrix <span class="math inline">\(\mathbf{S}\)</span> defined below and <span class="math inline">\(\bar{\mathbf{y}} = \mathbf{0}\)</span>. Here, the variance of <span class="math inline">\(\mathbf{y}_1\)</span> is twice that of <span class="math inline">\(\mathbf{y}_2\)</span> and the correlation works out to <span class="math inline">\(r(y_1, y_2) = 0.35\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ybar <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>S <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, .<span class="dv">5</span>, .<span class="dv">5</span>, <span class="dv">2</span>), <span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(S) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(S) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;y1&quot;</span>, <span class="st">&quot;y2&quot;</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>S</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#     y1  y2</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># y1 1.0 0.5</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># y2 0.5 2.0</span></span></code></pre></div>
</div>
<!-- fig.code: "R/ellipses-coverage.R -->
<p>Statistical ellipses are conveniently drawn using <code>car::ellipse()</code>. <code>heplots::ellipse.label()</code> provides flexible ways to add labels to ellipses at various locations around the ellipse shown in <a href="#fig-ellipses-coverage" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-ellipses-coverage</span></a>. These are called repeatedly to overlay the three ellipses.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode" id="cb12" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">0.50</span>, <span class="fl">0.68</span>, <span class="fl">0.95</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> <span class="fu">qchisq</span>(levels, <span class="at">df =</span> <span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">print</span>()</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 1.39 2.28 5.99</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># labels for ellipses, using plotmath</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>lab1 <span class="ot">&lt;-</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;c =&quot;</span>, chi[<span class="dv">2</span>]<span class="sc">^</span><span class="dv">2</span>, <span class="st">&quot;(&quot;</span>, .(levels[<span class="dv">1</span>]), <span class="st">&quot;) =&quot;</span>, .(c[<span class="dv">1</span>])))</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>lab2 <span class="ot">&lt;-</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;c =&quot;</span>, chi[<span class="dv">2</span>]<span class="sc">^</span><span class="dv">2</span>, <span class="st">&quot;(&quot;</span>, .(levels[<span class="dv">2</span>]), <span class="st">&quot;) =&quot;</span>, .(c[<span class="dv">2</span>])))</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>lab3 <span class="ot">&lt;-</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;c =&quot;</span>, chi[<span class="dv">2</span>]<span class="sc">^</span><span class="dv">2</span>, <span class="st">&quot;(&quot;</span>, .(levels[<span class="dv">3</span>]), <span class="st">&quot;) =&quot;</span>, .(c[<span class="dv">3</span>])))</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>e1 <span class="ot">&lt;-</span> <span class="fu">ellipse</span>(ybar, S, <span class="at">radius=</span><span class="fu">qchisq</span>(levels[<span class="dv">1</span>], <span class="dv">2</span>), </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">fill.alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">add=</span><span class="cn">FALSE</span>, </span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">8</span>, <span class="dv">8</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">9.5</span>, <span class="fl">9.5</span>), </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">asp=</span><span class="dv">1</span>, <span class="at">grid =</span> <span class="cn">FALSE</span>,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="fu">expression</span>(y[<span class="dv">1</span>]), </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="fu">expression</span>(y[<span class="dv">2</span>]),</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex.lab =</span> <span class="fl">1.5</span>)</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="fu">label.ellipse</span>(e1, <span class="at">label =</span> lab1, <span class="at">label.pos =</span> <span class="st">&quot;S&quot;</span>,</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">cex =</span> <span class="fl">1.2</span>)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>e2 <span class="ot">&lt;-</span> <span class="fu">ellipse</span>(ybar, S, <span class="at">radius=</span><span class="fu">qchisq</span>(levels[<span class="dv">2</span>], <span class="dv">2</span>), </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">col=</span><span class="st">&quot;blue&quot;</span>, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">fill.alpha=</span><span class="fl">0.3</span>)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="fu">label.ellipse</span>(e2, <span class="at">label =</span> lab2, <span class="at">label.pos =</span> <span class="st">&quot;N&quot;</span>,</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">cex =</span> <span class="fl">1.2</span>)</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>e3 <span class="ot">&lt;-</span> <span class="fu">ellipse</span>(ybar, S, <span class="at">radius=</span><span class="fu">qchisq</span>(levels[<span class="dv">3</span>], <span class="dv">2</span>), </span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">col=</span><span class="st">&quot;blue&quot;</span>, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">fill.alpha=</span><span class="fl">0.1</span>)</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="fu">label.ellipse</span>(e3, <span class="at">label =</span> lab3, <span class="at">label.pos =</span> <span class="st">&quot;N&quot;</span>,</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>              <span class="at">cex =</span> <span class="fl">1.2</span>)</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-ellipses-coverage" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-ellipses-coverage-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-ellipses-coverage-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ellipses-coverage-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.9: Data ellipses of 50%, 68% and 95% coverage when the means are <span class="math inline">\(\bar{\mathbf{y}} = \mathbf{0}\)</span> and the variance-covariance matrix is <span class="math inline">\(\mathbf{S}\)</span>.
</figcaption>
</figure>
</div>
</div>
</div>
<p>As always, graphic details matter. <a href="#fig-ellipses-coverage" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-ellipses-coverage</span></a> uses <code>asp = 1</code> so that units in the plot are the same for <span class="math inline">\(\mathbf{y}_1\)</span> and <span class="math inline">\(\mathbf{y}_2\)</span>, and we can see the greater variance for <span class="math inline">\(\mathbf{y}_2\)</span> as well as the correlation. Facilities of <code>grDevices::plotmath()</code> are used to provide mathematical annotations in the plot.</p>
</section>
<section id="ellipse-properties" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2"><span class="header-section-number">4.2.2</span> Ellipse properties</h3>
<p>The essential ideas of correlation and regression and their relation to ellipses go back to <span class="citation" data-cites="Galton:1886">Galton (<a href="#ref-Galton:1886" role="doc-biblioref">1886</a>)</span>. Galton’s goal was to predict (or explain) how a heritable trait, <span class="math inline">\(Y\)</span>, (e.g., height) of children was related to that of their parents, <span class="math inline">\(X\)</span>. He made a semi-graphic table of the frequencies of 928 observations of the average height of father and mother versus the height of their child, shown in <a href="#fig-galton-corr" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-corr</span></a>. (Today, we would put child height on the <span class="math inline">\(y\)</span> axis, but Galton was working from the table, so he organized it with parent height as the rows.) He then drew smoothed contour lines of equal frequencies and had the wonderful visual insight that these formed concentric shapes that were tolerably close to ellipses.</p>
<p>He then calculated summaries, <span class="math inline">\(\text{Ave}(Y | X)\)</span>, and, for symmetry, <span class="math inline">\(\text{Ave}(X | Y)\)</span>, and plotted these as lines of means on his diagram. Lo and behold, he had a second visual insight: the lines of means of (<span class="math inline">\(Y | X\)</span>) and (<span class="math inline">\(X | Y\)</span>) corresponded approximately to the loci of horizontal and vertical tangents to the concentric ellipses. To complete the picture, he added lines showing the geometric major and minor axes of the family of ellipses (which turned out to be the principal components) with the result shown in <a href="#fig-galton-corr" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-corr</span></a>. </p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-galton-corr" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-galton-corr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/galton-corr.jpg" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-galton-corr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.10: Galton’s 1886 diagram, showing the relationship of height of children to the average of their parents’ height. The diagram is essentially an overlay of a geometrical interpretation on a bivariate grouped frequency distribution, shown as numbers.
</figcaption>
</figure>
</div>
</div>
</div>
<p>For two variables, <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, the remarkable properties of the data ellipse are illustrated in <a href="#fig-galton-ellipse-r" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-ellipse-r</span></a>, a modern reconstruction of Galton’s diagram.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Partial code for this figure</summary>
<div class="sourceCode" id="cb13" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Galton, <span class="at">package =</span> <span class="st">&quot;HistData&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sunflowerplot</span>(parent <span class="sc">~</span> child, <span class="at">data=</span>Galton, </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">61</span>,<span class="dv">75</span>), </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">61</span>,<span class="dv">75</span>), </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">seg.col=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">&quot;Child height&quot;</span>, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylab=</span><span class="st">&quot;Mid Parent height&quot;</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>y.x <span class="ot">&lt;-</span> <span class="fu">lm</span>(parent <span class="sc">~</span> child, <span class="at">data=</span>Galton)     <span class="co"># regression of y on x</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(y.x, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>x.y <span class="ot">&lt;-</span> <span class="fu">lm</span>(child <span class="sc">~</span> parent, <span class="at">data=</span>Galton)     <span class="co"># regression of x on y</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>cc <span class="ot">&lt;-</span> <span class="fu">coef</span>(x.y)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="sc">-</span>cc[<span class="dv">1</span>]<span class="sc">/</span>cc[<span class="dv">2</span>], <span class="dv">1</span><span class="sc">/</span>cc[<span class="dv">2</span>], <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">&quot;gray&quot;</span>)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(Galton, </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>     car<span class="sc">::</span><span class="fu">dataEllipse</span>(child, parent, </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.points=</span><span class="cn">FALSE</span>, </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">levels=</span><span class="fu">c</span>(<span class="fl">0.40</span>, <span class="fl">0.68</span>, <span class="fl">0.95</span>), </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">lty=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-galton-ellipse-r" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-galton-ellipse-r-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/galton-ellipse-r.jpg" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-galton-ellipse-r-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.11: Sunflower plot of Galton’s data on heights of parents and their children (in.), with 40%, 68% and 95% data ellipses and the regression lines of <span class="math inline">\(y\)</span> on <span class="math inline">\(x\)</span> (black) and <span class="math inline">\(x\)</span> on <span class="math inline">\(y\)</span> (grey).
</figcaption>
</figure>
</div>
</div>
</div>
<ul>
<li><p>The ellipses have the mean vector <span class="math inline">\((\bar{x}, \bar{y})\)</span> as their center.</p></li>
<li><p>The lengths of arms of the <span style="color: blue;">blue</span> dashed central cross show the standard deviations of the variables, which correspond to the shadows of the ellipse covering 40% of the data. These are the bivariate analogs of the standard intervals <span class="math inline">\(\bar{x} \pm 1 s_x\)</span> and <span class="math inline">\(\bar{y} \pm 1 s_y\)</span>.</p></li>
<li><p>More generally, shadows (projections) on the coordinate axes, or any linear combination of them, give any standard interval, <span class="math inline">\(\bar{x} \pm k s_x\)</span> and <span class="math inline">\(\bar{y} \pm k s_y\)</span>. Those with <span class="math inline">\(k=1, 1.5, 2.45\)</span>, have bivariate coverage 40%, 68% and 95% respectively, corresponding to these quantiles of the <span class="math inline">\(\chi^2\)</span> distribution with 2 degrees of freedom, i.e., <span class="math inline">\(\chi^2_2 (.40) \approx 1^2\)</span>, <span class="math inline">\(\chi^2_2 (.68) \approx 1.5^2\)</span>, and <span class="math inline">\(\chi^2_2 (.95) \approx 2.45\)</span>. The shadows of the 68% ellipse are the bivariate analog of a univariate <span class="math inline">\(\bar{x} \pm 1 s_x\)</span> interval.</p>
<!--# and univariate coverage 68\%, 87\% and 98.6\% respectively. --></li>
<li><p>The regression line predicting <span class="math inline">\(y\)</span> from <span class="math inline">\(x\)</span> goes through the points where the ellipses have vertical tangents. The <em>other</em> regression line, predicting <span class="math inline">\(x\)</span> from <span class="math inline">\(y\)</span> goes through the points of horizontal tangency.</p></li>
<li><p>The correlation <span class="math inline">\(r(x, y)\)</span> is the ratio of the vertical segment from the mean of <span class="math inline">\(y\)</span> to the regression line to the vertical segment going to the top of the ellipse as shown at the right of the figure. It is <span class="math inline">\(r = 0.46\)</span> in this example.</p></li>
<li><p>The residual standard deviation, <span class="math inline">\(s_e = \sqrt{MSE} = \sqrt{\Sigma (y - \bar{y})^2 / n-2}\)</span>, is the half-length of the ellipse at the mean <span class="math inline">\(\bar{x}\)</span>.</p></li>
</ul>
<p>Because Galton’s values of <code>parent</code> and <code>child</code> height were recorded in class intervals of 1 in., they are shown as sunflower symbols in <a href="#fig-galton-ellipse-r" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-ellipse-r</span></a>, with multiple ‘petals’ reflecting the number of observations at each location. This plot (except for annotations) is constructed using <code>sunflowerplot()</code> and <code>car::dataEllipse()</code> for the ellipses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sunflowerplot</span>(parent <span class="sc">~</span> child, <span class="at">data=</span>Galton, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">61</span>,<span class="dv">75</span>), </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">61</span>,<span class="dv">75</span>), </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">seg.col=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">&quot;Child height&quot;</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylab=</span><span class="st">&quot;Mid Parent height&quot;</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>y.x <span class="ot">&lt;-</span> <span class="fu">lm</span>(parent <span class="sc">~</span> child, <span class="at">data=</span>Galton)    <span class="co"># regression of y on x</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(y.x, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>x.y <span class="ot">&lt;-</span> <span class="fu">lm</span>(child <span class="sc">~</span> parent, <span class="at">data=</span>Galton)    <span class="co"># regression of x on y</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>cc <span class="ot">&lt;-</span> <span class="fu">coef</span>(x.y)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="sc">-</span>cc[<span class="dv">1</span>]<span class="sc">/</span>cc[<span class="dv">2</span>], <span class="dv">1</span><span class="sc">/</span>cc[<span class="dv">2</span>], <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">&quot;gray&quot;</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(Galton, </span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>     car<span class="sc">::</span><span class="fu">dataEllipse</span>(child, parent, </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.points=</span><span class="cn">FALSE</span>, </span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">levels=</span><span class="fu">c</span>(<span class="fl">0.40</span>, <span class="fl">0.68</span>, <span class="fl">0.95</span>), </span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">lty=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
</div>
<p>Finally, as Galton noted in his diagram, the principal major and minor axes of the ellipse have important statistical properties. <span class="citation" data-cites="Pearson:1901">Pearson (<a href="#ref-Pearson:1901" role="doc-biblioref">1901</a>)</span> would later show that their directions are determined by the eigenvectors <span class="math inline">\(\mathbf{v}_1, \mathbf{v}_2, \dots\)</span> of the covariance matrix <span class="math inline">\(\mathbf{S}\)</span> and their radii by the square roots, <span class="math inline">\(\sqrt{\lambda_1}, \sqrt{\lambda_2}, \dots\)</span> of the corresponding eigenvalues.</p>
</section>
<section id="r-functions-for-data-ellipses" class="level3" data-number="4.2.3">
<h3 data-number="4.2.3"><span class="header-section-number">4.2.3</span> R functions for data ellipses</h3>
<p>A number of packages provide functions for drawing data ellipses in a scatterplot, with various features.</p>
<ul>
<li><code>car::scatterplot()</code>: uses base R graphics to draw 2D scatterplots, with a wide variety of plot enhancements including linear and non-parametric smoothers (loess, gam), a formula method, e.g., <code>y ~ x | group</code>, and marking points and lines using symbol shape, color, etc. Importantly, the <span style="color: brown;"><strong>car</strong></span> package generally allows automatic identification of “noteworthy” points by their labels in the plot using a variety of methods. For example, <code>method = "mahal"</code> labels cases with the most extreme Mahalanobis distances; <code>method = "r"</code> selects points according to their value of <code>abs(y)</code>, which is appropriate in residual plots.</li>
<li><code>car::dataEllipse()</code>: plots classical or robust data ellipses for one or more groups, with the same facilities for point identification. The robust version (<code>robust=TRUE</code>) uses the multivariate <span class="math inline">\(t\)</span> distribution (using <code>MASS::cov/trob()</code>) rather than the Gaussian,</li>
<li><code>heplots::covEllipses()</code>: draws classical or robust data ellipses for one or more groups in a one-way design and optionally for the pooled total sample, where the focus is on homogeneity of within-group covariance matrices.</li>
<li><code>ggplot2::stat_ellipse()</code>: uses the calculation methods of <code>car::dataEllipse()</code> to add unfilled (<code>geom = "path"</code>) or filled (<code>geom = polygon"</code>) data ellipses in a <code>ggplot</code> scatterplot, using inherited aesthetics.</li>
</ul>
<!-- ### Example: Canadian occupational prestige {#sec-prestige} -->
<div id="exm-prestige" class="theorem example">
<p><span class="theorem-title"><strong>Example 4.2</strong></span> <strong>Canadian occupational prestige</strong></p>
<p>These examples use the data on the prestige of 102 occupational categories and other measures from the 1971 Canadian Census, recorded in <code>Prestige</code>.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> Our interest is in understanding how <code>prestige</code> (the <span class="citation" data-cites="PineoPorter2008">Pineo &amp; Porter (<a href="#ref-PineoPorter2008" role="doc-biblioref">2008</a>)</span> prestige score for an occupational category, derived from a social survey) is related to census measures of the average education, income, percent women of incumbents in those occupations. Occupation <code>type</code> is a factor with levels <code>"bc"</code> (blue collar), <code>"wc"</code> (white collar) and <code>"prof"</code> (professional).</p>
<!-- fig.code: R/prestige.R -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Prestige, <span class="at">package=</span><span class="st">&quot;carData&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># `type` is really an ordered factor. Make it so.</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>Prestige<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">ordered</span>(Prestige<span class="sc">$</span>type,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">levels=</span><span class="fu">c</span>(<span class="st">&quot;bc&quot;</span>, <span class="st">&quot;wc&quot;</span>, <span class="st">&quot;prof&quot;</span>))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(Prestige)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;data.frame&#39;: 102 obs. of  6 variables:</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ education: num  13.1 12.3 12.8 11.4 14.6 ...</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ income   : int  12351 25879 9271 8865 8403 11030 8258 14163 11377 11023 ...</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ women    : num  11.16 4.02 15.7 9.11 11.68 ...</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ prestige : num  68.8 69.1 63.4 56.8 73.5 77.6 72.6 78.1 73.1 68.8 ...</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ census   : int  1113 1130 1171 1175 2111 2113 2133 2141 2143 2153 ...</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ type     : Ord.factor w/ 3 levels &quot;bc&quot;&lt;&quot;wc&quot;&lt;&quot;prof&quot;: 3 3 3 3 3 3 3 3 3 3 ...</span></span></code></pre></div>
</div>
</div>
<p>I first illustrate the relation between <code>income</code> and <code>prestige</code> in <a href="#fig-Prestige-scatterplot-income1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot-income1</span></a> using <code>car::scatterplot()</code> with many of its bells and whistles, including marginal boxplots for the variables, the linear regression line, loess smooth and the 68% data ellipse.</p>
<!-- fig.code: R/prestige.R -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot</span>(prestige <span class="sc">~</span> income, <span class="at">data=</span>Prestige,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex.lab =</span> <span class="fl">1.25</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>),</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth =</span> <span class="fu">list</span>(<span class="at">smoother=</span>loessLine, </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">lty.smooth =</span> <span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">col.smooth =</span> <span class="st">&quot;darkgreen&quot;</span>, </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">col.var =</span> <span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fl">0.68</span>),</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">4</span>, <span class="at">method =</span> <span class="st">&quot;mahal&quot;</span>, <span class="at">col=</span><span class="st">&quot;black&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>))</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co"># general.managers          lawyers        ministers       physicians </span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                2               17               20               24</span></span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Prestige-scatterplot-income1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Prestige-scatterplot-income1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Prestige-scatterplot-income1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Prestige-scatterplot-income1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.12: Scatterplot of prestige vs. income, showing the linear regression line (<span style="color: red;">red</span>), the loess smooth with a confidence envelope (<span style="color: darkgreen;">darkgreen</span>) and a 68% data ellipse. Points with the 4 largest Mahalanobis <span class="math inline">\(D^2\)</span> values are labeled.
</figcaption>
</figure>
</div>
</div>
</div>
<p>There is a lot that can be seen here:</p>
<ul>
<li><code>income</code> is positively skewed, as is often the case.</li>
<li>The loess smooth, on the scale of income, shows <code>prestige</code> increasing up to $15,000 (these are 1971 incomes), and then leveling off.</li>
<li>The bivariate 1 standard deviation data ellipse, centered at the means encloses approximately 68% of the data points. It adds visual information about the correlation and precision of the linear regression; but here, the non-linear trend for higher incomes strongly suggests a different approach.</li>
<li>The four points identified by their labels are those with the largest Mahalanobis distances. <code>scatterplot()</code> prints their labels to the console.</li>
</ul>
<p><a href="#fig-Prestige-scatterplot-educ1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot-educ1</span></a> shows a similar plot for education, which from the boxplot appears to be reasonably symmetric. The smoothed curve is quite close to the linear regression, according to which <code>prestige</code> increases on average <code>coef(lm(prestige ~ education, data=Prestige))["education"]</code> = 5.361 with each year of education.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot</span>(prestige <span class="sc">~</span> education, <span class="at">data=</span>Prestige,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex.lab =</span> <span class="fl">1.3</span>,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>),</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth =</span> <span class="fu">list</span>(<span class="at">smoother=</span>loessLine, </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">lty.smooth =</span> <span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">col.smooth =</span> <span class="st">&quot;darkgreen&quot;</span>, </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">col.var =</span> <span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fl">0.68</span>),</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">4</span>, <span class="at">method =</span> <span class="st">&quot;mahal&quot;</span>, <span class="at">col=</span><span class="st">&quot;black&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>))</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">#  physicians file.clerks    newsboys     farmers </span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co">#          24          41          53          67</span></span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Prestige-scatterplot-educ1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Prestige-scatterplot-educ1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Prestige-scatterplot-educ1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Prestige-scatterplot-educ1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.13: Scatterplot of prestige vs. education, showing the linear regression line (<span style="color: red;">red</span>), the loess smooth with a confidence envelope (<span style="color: darkgreen;">darkgreen</span>) and a 68% data ellipse.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In this plot, farmers, newsboys, file.clerks and physicians are identified as noteworthy, for being furthest from the mean by Mahalanobis distance. In relation to their typical level of education, these are mostly understandable, but it is nice that farmers are rated of higher prestige than their level of education would predict.</p>
<p>Note that the <code>method</code> argument for point identification can take a vector of case numbers indicating the points to be labeled. So, to label the observations with large absolute standardized residuals in the linear model <code>m</code>, you can use <code>method = which(abs(rstandard(m)) &gt; 2)</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">lm</span>(prestige <span class="sc">~</span> education, <span class="at">data=</span>Prestige)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot</span>(prestige <span class="sc">~</span> education, <span class="at">data=</span>Prestige,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex.lab =</span> <span class="fl">1.3</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">boxplots =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>),</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> <span class="fu">list</span>(<span class="at">smoother=</span>loessLine,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">lty.smooth =</span> <span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                  <span class="at">col.smooth =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">col.var =</span> <span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fl">0.68</span>),</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">4</span>, <span class="at">method =</span> <span class="fu">which</span>(<span class="fu">abs</span>(<span class="fu">rstandard</span>(m))<span class="sc">&gt;</span><span class="dv">2</span>), </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>              <span class="at">col=</span><span class="st">&quot;black&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">invisible</span>()</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Prestige-scatterplot-educ2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Prestige-scatterplot-educ2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Prestige-scatterplot-educ2-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Prestige-scatterplot-educ2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.14: Scatterplot of prestige vs. education, labeling points whose absolute standardized residual is &gt; 2.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-log-scale" class="level3" data-number="4.2.4">
<h3 data-number="4.2.4"><span class="header-section-number">4.2.4</span> Handling nonlinearity: Plotting on a log scale</h3>
<p>A typical remedy for the non-linear relationship of income to prestige is to plot income on a log scale. This usually makes sense, and expresses a belief that a <strong>multiple</strong> of or <strong>percentage increase</strong> in income has a constant impact on prestige, as opposed to the <strong>additive</strong> interpretation for income itself.</p>
<p>For example, the slope of the linear regression line in <a href="#fig-Prestige-scatterplot-income1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot-income1</span></a> is given by <code>coef(lm(prestige ~ income, data=Prestige))["income"]</code> = 0.003. Multiplying this by 1000 says that a $1000 increase in <code>income</code> is associated with with an average increase of <code>prestige</code> of 2.9.</p>
<p>In the plot below, <code>scatterplot(..., log = "x")</code> re-scales the x-axis to the <span class="math inline">\(\log_e()\)</span> scale. The slope, <code>coef(lm(prestige ~ log(income), data=Prestige))["log(income)"]</code> = 21.556 says that a 1% increase in salary is associated with an average change of 21.55 / 100 in prestige.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot</span>(prestige <span class="sc">~</span> income, <span class="at">data=</span>Prestige,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">log =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex.lab =</span> <span class="fl">1.3</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">col =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth =</span> <span class="fu">list</span>(<span class="at">smoother=</span>loessLine,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">lty.smooth =</span> <span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">col.smooth =</span> <span class="st">&quot;darkgreen&quot;</span>, <span class="at">col.var =</span> <span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fl">0.68</span>),</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">4</span>, <span class="at">method =</span> <span class="st">&quot;mahal&quot;</span>, <span class="at">col=</span><span class="st">&quot;black&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>))</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># general.managers        ministers         newsboys      babysitters </span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                2               20               53               63</span></span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Prestige-scatterplot2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Prestige-scatterplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Prestige-scatterplot2-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Prestige-scatterplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.15: Scatterplot of prestige vs. log(income).
</figcaption>
</figure>
</div>
</div>
</div>
<p>The smoothed curve in <a href="#fig-Prestige-scatterplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot2</span></a> exhibits a slight tendency to bend upwards, but a linear relation is a reasonable approximation.</p>
</section>
<section id="sec-stratifying" class="level3" data-number="4.2.5">
<h3 data-number="4.2.5"><span class="header-section-number">4.2.5</span> Stratifying</h3>
<p>Before going further, it is instructive to ask what we could see in the relationship between income and prestige if we stratified by type of occupation, fitting separate regressions and smooths for blue collar, white collar and professional incumbents in these occupations.</p>
<p>The formula <code>prestige ~ income | type</code> (read: income <em>given</em> type) is a natural way to specify grouping by <code>type</code>; separate linear regressions and smooths are calculated for each group, applying the color and point shapes specified by the <code>col</code> and <code>pch</code> arguments.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot</span>(prestige <span class="sc">~</span> income <span class="sc">|</span> type, <span class="at">data=</span>Prestige,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">15</span><span class="sc">:</span><span class="dv">17</span>, <span class="at">cex.lab =</span> <span class="fl">1.3</span>,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">grid =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend =</span> <span class="fu">list</span>(<span class="at">coords=</span><span class="st">&quot;bottomright&quot;</span>),</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty =</span> <span class="st">&quot;longdash&quot;</span>),</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth=</span><span class="fu">list</span>(<span class="at">smoother=</span>loessLine, </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">var=</span><span class="cn">FALSE</span>, <span class="at">lwd.smooth=</span><span class="dv">2</span>, <span class="at">lty.smooth=</span><span class="dv">1</span>))</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-Prestige-scatterplot3" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-Prestige-scatterplot3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-Prestige-scatterplot3-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Prestige-scatterplot3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.16: Scatterplot of prestige vs. income, stratified by occupational type. This implies a different interpretation, where occupation type is a moderator variable, so different regressions apply to each type. The linear regression line is dashed, while the loess smooth is solid.
</figcaption>
</figure>
</div>
</div>
</div>
<p>This visual analysis offers a different interpretation of the dependence of prestige on income, which appeared to be non-linear when occupation type was ignored. Instead, <a href="#fig-Prestige-scatterplot3" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot3</span></a> suggests an <em>interaction</em> of income by type. In a model formula this would be expressed as one of:</p>
<div class="sourceCode" id="cb21" data-code-line-numbers=""><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lm</span>(prestige <span class="sc">~</span> income <span class="sc">+</span> type <span class="sc">+</span> income<span class="sc">:</span>type, <span class="at">data =</span> Prestige)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lm</span>(prestige <span class="sc">~</span> income <span class="sc">*</span> type, <span class="at">data =</span> Prestige)</span></code></pre></div>
<p>These models signify that there are different slopes (and intercepts) for the three occupational types. In this interpretation, <code>type</code> is a <strong>moderator</strong> variable, with a different story. The slopes of the fitted lines suggest that among blue collar workers, prestige increases sharply with their income. For white collar and professional workers, there is still an increasing relation of prestige with income, but the effect of income (slope) diminishes with higher occupational category. A different fitted relationship entails a different story.</p>
</section>
<section id="sec-penguins" class="level3" data-number="4.2.6">
<h3 data-number="4.2.6"><span class="header-section-number">4.2.6</span> Meet the Penguins</h3>
<p>The <code>penguins</code> dataset from the <span style="color: brown;"><strong>palmerpenguins</strong></span> package <span class="citation" data-cites="R-palmerpenguins">(<a href="#ref-R-palmerpenguins" role="doc-biblioref">Horst et al., 2020</a>)</span> provides further instructive examples of plots and analyses of multivariate data. The data consists of measurements of body size (flipper length, body mass, bill length and depth) of 344 penguins collected at the <a href="https://pallter.marine.rutgers.edu/">Palmer Research Station</a> in Antarctica.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-penguin-species" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-penguin-species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/penguins-horst.png" class="img-fluid quarto-figure quarto-figure-center" style="width:50.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguin-species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.17: Penguin species observed in the Palmer Archipelago. This is a cartoon, but it illustrates some features of penguin body size measurements, and the colors typically used for species. Image credit: Allison Horst
</figcaption>
</figure>
</div>
</div>
</div>
<p>There were three different species of penguins (Adélie, Chinstrap &amp; Gentoo) collected from 3 islands in the Palmer Archipelago between 2007–2009 <span class="citation" data-cites="Gorman2014">(<a href="#ref-Gorman2014" role="doc-biblioref">Gorman et al., 2014</a>)</span>. The purpose was to examine differences in size or appearance of these species, particularly those between the sexes (sexual dimorphism) in relation to foraging and habitat.</p>
<p>Here, I use a slightly altered version of the dataset, <code>heplots::peng</code>, constructed by renaming variables to remove the units, making factors of character variables and deleting a few cases with missing data.<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(penguins, <span class="at">package =</span> <span class="st">&quot;palmerpenguins&quot;</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>peng <span class="ot">&lt;-</span> penguins <span class="sc">|&gt;</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">bill_length =</span> bill_length_mm, </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">bill_depth =</span> bill_depth_mm, </span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">flipper_length =</span> flipper_length_mm, </span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">body_mass =</span> body_mass_g</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">species =</span> <span class="fu">as.factor</span>(species),</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">island =</span> <span class="fu">as.factor</span>(island),</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">sex =</span> <span class="fu">as.factor</span>(<span class="fu">substr</span>(sex,<span class="dv">1</span>,<span class="dv">1</span>))) <span class="sc">|&gt;</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">drop_na</span>()</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(peng)</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Rows: 333</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Columns: 8</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="co"># $ species        &lt;fct&gt; Adelie, Adelie, Adelie, Adelie, Adelie, Ade…</span></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co"># $ island         &lt;fct&gt; Torgersen, Torgersen, Torgersen, Torgersen,…</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="co"># $ bill_length    &lt;dbl&gt; 39.1, 39.5, 40.3, 36.7, 39.3, 38.9, 39.2, 4…</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a><span class="co"># $ bill_depth     &lt;dbl&gt; 18.7, 17.4, 18.0, 19.3, 20.6, 17.8, 19.6, 1…</span></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co"># $ flipper_length &lt;int&gt; 181, 186, 195, 193, 190, 181, 195, 182, 191…</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="co"># $ body_mass      &lt;int&gt; 3750, 3800, 3250, 3450, 3650, 3625, 4675, 3…</span></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="co"># $ sex            &lt;fct&gt; m, f, f, f, m, f, m, f, m, m, f, f, m, f, m…</span></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a><span class="co"># $ year           &lt;int&gt; 2007, 2007, 2007, 2007, 2007, 2007, 2007, 2…</span></span></code></pre></div>
</div>
<p>There are quite a few variables to choose for illustrating data ellipses in scatterplots. Here I focus on the measures of their bills, <code>bill_length</code> and <code>bill_depth</code> (indicating curvature) and show how to use <code>ggplot2</code> for these plots.</p>
<p>I’ll be using the penguins data quite a lot, so it is useful to set up custom colors like those used in <a href="#fig-penguin-species" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-penguin-species</span></a>. My versions are shown in <a href="#fig-peng-colors" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-colors</span></a> with their color codes. These are shades of:</p>
<ul>
<li><span style="background-color: #FFA500; color: #212529; padding: 2px">Adélie</span>: <span style="background-color: #FFA500; color: #212529; padding: 2px">orange</span>,</li>
<li><span style="background-color: #A020F0; color: #FFFFFF; padding: 2px">Chinstrap</span>: <span style="background-color: #A020F0; color: #FFFFFF; padding: 2px">purple</span>, and</li>
<li><span style="background-color: #00FF00; color: #212529; padding: 2px">Gentoo</span>: <span style="color: green;">green</span>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-colors" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-colors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/peng-colors.png" class="img-fluid quarto-figure quarto-figure-center" style="width:50.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-colors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.18: Color palettes used for penguin species. Each species has a primary hue, which can be used in light, medium or dark versions.
</figcaption>
</figure>
</div>
</div>
</div>
<p>To use these in <code>ggplot2</code> I define a function <code>peng.colors()</code> that allows shades of light, medium and dark and then functions <code>scale_*_penguins()</code> for color and fill.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb23" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>peng.colors <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">shade=</span><span class="fu">c</span>(<span class="st">&quot;medium&quot;</span>, <span class="st">&quot;light&quot;</span>, <span class="st">&quot;dark&quot;</span>)) {</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  shade <span class="ot">=</span> <span class="fu">match.arg</span>(shade)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             light      medium     dark</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  oranges <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;#FDBF6F&quot;</span>, <span class="st">&quot;#F89D38&quot;</span>, <span class="st">&quot;#F37A00&quot;</span>)  <span class="co"># Adelie</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  purples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;#CAB2D6&quot;</span>, <span class="st">&quot;#9A78B8&quot;</span>, <span class="st">&quot;#6A3D9A&quot;</span>)  <span class="co"># Chinstrap</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  greens <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">&quot;#B2DF8A&quot;</span>, <span class="st">&quot;#73C05B&quot;</span>, <span class="st">&quot;#33a02c&quot;</span>)  <span class="co"># Gentoo</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  cols.vec <span class="ot">&lt;-</span> <span class="fu">c</span>(oranges, purples, greens)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  cols.mat <span class="ot">&lt;-</span> </span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">matrix</span>(cols.vec, <span class="dv">3</span>, <span class="dv">3</span>, </span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">byrow =</span> <span class="cn">TRUE</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="at">species =</span> <span class="fu">c</span>(<span class="st">&quot;Adelie&quot;</span>, <span class="st">&quot;Chinstrap&quot;</span>, <span class="st">&quot;Gentoo&quot;</span>),</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>                           <span class="at">shade =</span> <span class="fu">c</span>(<span class="st">&quot;light&quot;</span>, <span class="st">&quot;medium&quot;</span>, <span class="st">&quot;dark&quot;</span>)))</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get shaded colors</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  cols.mat[, shade ]</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="co"># define color and fill scales</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>scale_fill_penguins <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">shade=</span><span class="fu">c</span>(<span class="st">&quot;medium&quot;</span>, <span class="st">&quot;light&quot;</span>, <span class="st">&quot;dark&quot;</span>), ...){</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  shade <span class="ot">=</span> <span class="fu">match.arg</span>(shade)</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">discrete_scale</span>(</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;fill&quot;</span>,<span class="st">&quot;penguins&quot;</span>,</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>     scales<span class="sc">:::</span><span class="fu">manual_pal</span>(<span class="at">values =</span> <span class="fu">peng.colors</span>(shade)), ...)</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>scale_colour_penguins <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">shade=</span><span class="fu">c</span>(<span class="st">&quot;medium&quot;</span>, <span class="st">&quot;light&quot;</span>, <span class="st">&quot;dark&quot;</span>), ...){</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>  shade <span class="ot">=</span> <span class="fu">match.arg</span>(shade)</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">discrete_scale</span>(</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;colour&quot;</span>,<span class="st">&quot;penguins&quot;</span>,</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>    scales<span class="sc">:::</span><span class="fu">manual_pal</span>(<span class="at">values =</span> <span class="fu">peng.colors</span>(shade)), ...)</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>scale_color_penguins <span class="ot">&lt;-</span> scale_colour_penguins</span></code></pre></div>
</details>
</div>
<p>This is used to define a <code>theme_penguins()</code> function that I use to simply change the color and fill scales for plots below. I also define a convenience function <code>legend_inside()</code> to make it less verbose to position a legend <em>inside</em> a plot window, because outside legends reduce resolution within the plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>theme_penguins <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">shade=</span><span class="fu">c</span>(<span class="st">&quot;medium&quot;</span>, <span class="st">&quot;light&quot;</span>, <span class="st">&quot;dark&quot;</span>), </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                           ...) {</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  shade <span class="ot">=</span> <span class="fu">match.arg</span>(shade)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="fu">scale_color_penguins</span>(<span class="at">shade=</span>shade),</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>       <span class="fu">scale_fill_penguins</span>(<span class="at">shade=</span>shade)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>legend_inside <span class="ot">&lt;-</span> <span class="cf">function</span>(position) {          <span class="co"># simplify legend placement</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;inside&quot;</span>,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position.inside =</span> position)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<p>An initial plot using <code>ggplot2</code> shown in <a href="#fig-peng-ggplot1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot1</span></a> uses color and point shape to distinguish the three penguin species. I annotate the plot of points using the linear regression lines, loess smooths to check for non-linearity and 95% data ellipses to show precision of the linear relation.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb25" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng, </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species, <span class="at">shape =</span> species, <span class="at">fill=</span>species)) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;loess&quot;</span>,  <span class="at">formula =</span> y <span class="sc">~</span> x,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">linewidth =</span> <span class="fl">1.5</span>, <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">alpha=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.95</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>(<span class="st">&quot;dark&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.85</span>, <span class="fl">0.15</span>))</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peng-ggplot1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-ggplot1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.19: Penguin bill length and bill depth according to species.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-visual-thinning" class="level3" data-number="4.2.7">
<h3 data-number="4.2.7"><span class="header-section-number">4.2.7</span> Visual thinning</h3>
<p>Overall, the three species occupy different regions of this 2D space and for each species the relation between bill length and depth appears reasonably linear. Given this, we can suppress plotting the data points to get a visual summary of the data using the fitted regression lines and data ellipses, as shown in <a href="#fig-peng-ggplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot2</span></a>.</p>
<p>This idea, of <strong>visual thinning</strong> a graph to focus on what should be seen, becomes increasingly useful as the data becomes more complex. The <code>ggplot2</code> framework encourages this, because we can think of various components as layers, to be included or not. In <a href="#fig-peng-ggplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot2</span></a> I chose to include only the regression line and add data ellipses of 40%, 68% and 95% coverage to highlight the increasing bivariate density around the group means.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb26" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng, </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species, <span class="at">shape =</span> species, <span class="at">fill=</span>species)) <span class="sc">+</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>,  <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.95</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.68</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.40</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>(<span class="st">&quot;dark&quot;</span>) <span class="sc">+</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.85</span>, <span class="fl">0.15</span>))</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peng-ggplot2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-ggplot2-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.20: <strong>Visual thinning</strong>: Suppressing the data points gives a visual summary of the relation between bill length and bill depth using the regression line and data ellipses.
</figcaption>
</figure>
</div>
</div>
</div>
<!-- r child="child/03-bagplot.qmd" -->
</section>
</section>
<section id="sec-bagplots" class="level2" data-number="4.3">
<h2 data-number="4.3"><span class="header-section-number">4.3</span> Bagplots</h2>
<p>If you are concerned about the assumption of bivariate normality entailed by the data ellipse, a very nice non-parametric and robust alternative is a 2D generalization of a boxplot called a <strong>bagplot</strong>, introduced by <span class="citation" data-cites="Rousseeuw-etal:2012">Rousseeuw et al. (<a href="#ref-Rousseeuw-etal:2012" role="doc-biblioref">1999</a>)</span>. The idea is very simple. The bagplot consists of three nested polygons, called the “bag”, the “fence”, and the “loop”:</p>
<ul>
<li><p><strong>bag</strong>: The central 50% box of the boxplot is replaced by a polygon called the “bag”, constructed on the basis of <em>depth</em> of points from the bivariate depth median point. Depth generalizes the univariate concept of rank, but counted from the medians point outward in any direction. The bag contains at most 50% of the most central data points. </p></li>
<li><p><strong>fence</strong>: The univariate fences are replaced by a “fence” polygon, by expanding the bag outward by a factor (<code>coef</code>), usually 3, from the depth median.</p></li>
<li><p><strong>loop</strong>: Points beyond the fence (the “loop”) are potential outliers, sometimes plotted as a convex hull surrounding <em>all</em> of the observations, but better rendered simply as a scatterplot of just those points.</p></li>
</ul>
<p>In this way, the bagplot visualizes the bivariate location (median), spread, correlation, skewness, and tails of the data. Compared with a standard data ellipse, it serves as a visual test of normality and provides a simple way to identify outliers.</p>
<p>Bagplots are implemented in <span style="color: brown;"><strong>gggda</strong></span> <span class="citation" data-cites="R-gggda">(<a href="#ref-R-gggda" role="doc-biblioref">Brunson &amp; Gracey, 2025</a>)</span> in the <code>ggplot2</code> framework as <code>geom_bagplot()</code> with computations done via <code>stat_bagplot()</code>. For the Penguin data, the bagplot version of <a href="#fig-peng-ggplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot2</span></a> is shown in <a href="#fig-peng-bagplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-bagplot</span></a>.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb27" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng, </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species, <span class="at">shape =</span> species, <span class="at">fill=</span>species)) <span class="sc">+</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bagplot</span>(<span class="at">bag.alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">outlier.size =</span> <span class="dv">5</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">fraction =</span> <span class="fl">0.5</span>,    <span class="co"># bag fraction</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">coef =</span> <span class="fl">2.5</span>,        <span class="co"># fence factor</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>(<span class="st">&quot;dark&quot;</span>) <span class="sc">+</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.87</span>, <span class="fl">0.15</span>)) </span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peng-bagplot" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-bagplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-bagplot-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-bagplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.21: <strong>Bagplot</strong>: For each Penguin species the darker inner (bag) polygon reflects the innermost 50% of the data points. The outer (fence) polygon corresponds to points enclosed by a multiple of the bag. Points outside the fence for each species are ploted individually.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Compared with the data ellipses shown in <a href="#fig-peng-ggplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot2</span></a>, the bag polygons have shapes very close to ellipses, giving credence to the assumption of (approximate) normality. Using the factor <code>coef = 2.5</code> causes five points to be flagged as potential outliers. Bivariate skewness is indicated when the bag and fence is not symmetric around the central median in some direction, as is slightly true for all three species.</p>
<p>I discuss other visual tests of multivariate normality in <a href="#sec-multivar-normality" class="quarto-xref"><span class="quarto-unresolved-ref">sec-multivar-normality</span></a> and consider outlier identification for the Penguin data in <a href="#sec-multnorm-penguin" class="quarto-xref"><span class="quarto-unresolved-ref">sec-multnorm-penguin</span></a> below. </p>
</section>
<section id="sec-bivar-density" class="level2" data-number="4.4">
<h2 data-number="4.4"><span class="header-section-number">4.4</span> Non-parametric bivariate density plots</h2>
<p>While I emphasize data ellipses (because I like their beautiful geometry), other visual summaries of the bivariate density are possible and often useful. To see more detail about the “shape” of bivariate data in a non-parametric way you can use one of the methods described below to see a model-free representation of your data.</p>
<p>For a single variable, <code>stats::density()</code> and <code>ggplot2::geom_density()</code> calculate a smoothed estimate of the density using nonparametric kernel methods <span class="citation" data-cites="Silverman:86">(<a href="#ref-Silverman:86" role="doc-biblioref">Silverman, 1986</a>)</span> whose smoothness is controlled by a bandwidth parameter, analogous to the span in a loess smoother. This idea extends to two (and more) variables <span class="citation" data-cites="Scott1992">(<a href="#ref-Scott1992" role="doc-biblioref">Scott, 1992</a>)</span>. For bivariate data, <code>MASS::kde2d()</code> estimates the density on a square <span class="math inline">\(n \times n\)</span> grid over the ranges of the variables.</p>
<p><code>ggplot2</code> provides <code>geom_density_2d()</code> which uses <code>MASS::kde2d()</code> and displays these as contours— horizontal slices of the 3D surface at equally-spaced heights and projects these onto the 2D plane. The <span style="color: brown;"><strong>ggdensity</strong></span> package <span class="citation" data-cites="R-ggdensity">(<a href="#ref-R-ggdensity" role="doc-biblioref">Otto &amp; Kahle, 2023</a>)</span> extends this with <code>geom_hdr()</code>, computing the high density regions that bound given levels of probability and maps these to the <code>alpha</code> transparency aesthetic. A <code>method</code> argument allows you to specify various nonparametric (<code>method ="kde"</code> is the default) and parametric (<code>method ="mvnorm"</code> gives normal data ellipses) ways to estimate the underlying bivariate distribution. </p>
<p><a href="#fig-peng-ggdensity" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggdensity</span></a> shows these side-by-side for comparison. With <code>geom_density_2d()</code> you can specify either the number of contour <code>bins</code> or the width of these bins (<code>binwidth</code>). For <code>geom_hdr()</code>, the <code>probs</code> argument gives a result that is easier to understand.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb28" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdensity)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(peng, </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species)) <span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>,  <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_2d</span>(<span class="at">linewidth =</span> <span class="fl">1.1</span>, <span class="at">bins =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;geom_density_2d&quot;</span>) <span class="sc">+</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>() <span class="sc">+</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.85</span>, <span class="fl">0.15</span>))</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(peng, </span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species, <span class="at">fill =</span> species)) <span class="sc">+</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>,  <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hdr</span>(<span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.95</span>, <span class="fl">0.68</span>, <span class="fl">0.4</span>), <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;ggdensity::geom_hdr&quot;</span>) <span class="sc">+</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>() <span class="sc">+</span></span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peng-ggdensity" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggdensity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-ggdensity-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggdensity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.22: <strong>Bivariate densities</strong> show the contours of the 3D surface representing the frequency in the joint distribution of bill length and bill depth.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Compared with the data ellipse, which is highly smoothed by the Gaussian assumption and the bagplot, which is smoothed by the concept of contours of data depth, the plots in <a href="#fig-peng-ggdensity" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggdensity</span></a> show considerably more detail, and an intriguing suggestion of two peaks for our Chinstrap penguins. </p>
</section>
<section id="sec-simpsons" class="level2" data-number="4.5">
<h2 data-number="4.5"><span class="header-section-number">4.5</span> Simpson’s paradox: marginal and conditional relationships</h2>
<p>Because it provides a visual representation of means, variances, and correlations, the data ellipse is ideally suited as a tool for illustrating and explicating various phenomena that occur in the analysis of linear models. One class of simple, but important, examples concerns the difference between the <em>marginal relationship</em> between variables, ignoring some important factor or covariate, and the <em>conditional</em> relationship, adjusting (controlling) for that variable.</p>
<p>An important example is <strong>Simpson’s paradox</strong> <span class="citation" data-cites="Simpson:51">(<a href="#ref-Simpson:51" role="doc-biblioref">Simpson, 1951</a>)</span> which occurs when the marginal and conditional relationships differ in direction. That is, the overall correlation in a model <code>y ~ x</code> might be negative, while the within-group correlations in separate models for each group <code>y[g] ~ x[g]</code> might be positive, or vice versa. For Flatlanders in 2D space, this is a puzzlement.</p>
<p>We can see this paradox in the plots of bill length against bill depth for the penguin data shown in <a href="#fig-peng-simpsons" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-simpsons</span></a>. Ignoring penguin species, the marginal, total-sample correlation is slightly negative as seen in panel (a). The individual-sample ellipses in panel (b) show that the conditional, within-species correlations are all positive, with approximately equal regression slopes. However the group means have a negative relationship, accounting for the negative marginal correlation when species is ignored.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-simpsons" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-simpsons-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/peng-simpsons.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-simpsons-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.23: Marginal (a), conditional (b), and pooled within-sample (c) relationships of bill length and depth in the Penguins data. Each plot shows the 68% data ellipse and regression line(s) with 95% confidence bands.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The regression line in panel (a) is that for the linear model <code>lm(bill_depth ~ bill_length)</code>, while the separate lines in panel (b) are those for the model <code>lm(bill_depth ~ bill_length * species)</code> which allows a different slope and intercept for each species.</p>
<p>A correct analysis of the (conditional) relationship between these variables, controlling or adjusting for mean differences among species, is based on the pooled within-sample covariance matrix, a weighted average of the individual within-group <span class="math inline">\(\mathbf{S}_i\)</span>,</p>
<p><span class="math display">\[
\mathbf{S}_{\textrm{within}}  =
\sum_{i=1}^g
(n_i - 1) \mathbf{S}_i \, / \, (N - g)
\:\: ,
\]</span></p>
<p>where <span class="math inline">\(N = \sum n_i\)</span>. The result is shown in panel (c) of <a href="#fig-peng-simpsons" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-simpsons</span></a>.</p>
<p>In this graph, the data for each species were first transformed to deviations from the species means on both variables and then translated back to the grand means. You can also see here that the shapes and sizes of the individual data ellipses are roughly comparable, but perhaps not identical. This visual idea of centering groups to a common mean will become important in <a href="#sec-eqcov" class="quarto-xref"><span class="quarto-unresolved-ref">sec-eqcov</span></a> when we want to test the assumption of equality of error covariances in multivariate models.</p>
<p>The <code>ggplot2</code> code for the panels in this figure are shown below. Note that for components that will be the same across panels, you can define elements (e.g., <code>labels</code>, <code>theme_penguins()</code>) once, and then re-use these across several graphs.</p>
<p><strong>TODO</strong> This panel tabset looks fine in HTML but is awkward in PDF</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">(a) Ignoring species</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">(b) By species</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">(c) Within species</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">labs</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">&quot;Bill length (mm)&quot;</span>,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">&quot;Bill depth (mm)&quot;</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="st">&quot;Species&quot;</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">shape =</span> <span class="st">&quot;Species&quot;</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="st">&quot;Species&quot;</span>) </span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>plt1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> peng,</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> bill_length,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> bill_depth)) <span class="sc">+</span></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x, </span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>              <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> <span class="st">&quot;gray50&quot;</span>) <span class="sc">+</span></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">level =</span> <span class="fl">0.68</span>, <span class="at">linewidth =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Ignoring species&quot;</span>) <span class="sc">+</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  labels</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>plt1</span></code></pre></div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>plt2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> peng,</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> bill_length,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> bill_depth,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">color =</span> species,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">shape =</span> species,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill =</span> species)) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">1.5</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x, </span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">level =</span> <span class="fl">0.68</span>, <span class="at">linewidth =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;By species&quot;</span>) <span class="sc">+</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  labels <span class="sc">+</span></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>(<span class="st">&quot;dark&quot;</span>) <span class="sc">+</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.83</span>, <span class="fl">0.16</span>)) </span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>plt2</span></code></pre></div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># center within groups, translate to grand means</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>means <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(peng[, <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>peng.centered <span class="ot">&lt;-</span> peng <span class="sc">|&gt;</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(species) <span class="sc">|&gt;</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">bill_length =</span> means[<span class="dv">1</span>] <span class="sc">+</span> <span class="fu">scale</span>(bill_length, <span class="at">scale =</span> <span class="cn">FALSE</span>),</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">bill_depth  =</span> means[<span class="dv">2</span>] <span class="sc">+</span> <span class="fu">scale</span>(bill_depth, <span class="at">scale =</span> <span class="cn">FALSE</span>))</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>plt3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> peng.centered,</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> bill_length,</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y =</span> bill_depth,</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">color =</span> species,</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">shape =</span> species,</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill =</span> species)) <span class="sc">+</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">1.5</span>,</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x, </span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>              <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">level =</span> <span class="fl">0.68</span>, <span class="at">linewidth =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>  labels <span class="sc">+</span></span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Within species&quot;</span>) <span class="sc">+</span></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>(<span class="st">&quot;dark&quot;</span>) <span class="sc">+</span></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.83</span>, <span class="fl">0.16</span>)) </span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>plt3</span></code></pre></div>
</div>
</div>
</div>
</div>
<p><strong>TODO</strong>: Add stuff on 3D scatterplots, using the <code>R/penguin/peng-3D-rgl.R</code> example. A start on this is in <code>child/03-3D-scat.qmd</code></p>
</section>
<section id="sec-multivar-normality" class="level2" data-number="4.6">
<h2 data-number="4.6"><span class="header-section-number">4.6</span> Multivariate normality and outliers</h2>
<p>The relation of the data ellipsoid for <span class="math inline">\(p\)</span> variables to the <span class="math inline">\(\chi^2_p\)</span> distribution with <span class="math inline">\(p\)</span> degrees of freedom described in <a href="#sec-data-ellipse" class="quarto-xref"><span class="quarto-unresolved-ref">sec-data-ellipse</span></a> is based on the assumption that the data in <span class="math inline">\(\mathbf{y}\)</span> is a sample from a multivariate normal distribution (MVN), with a mean vector <span class="math inline">\(\boldsymbol{\mu}\)</span> and variance-covariance matrix <span class="math inline">\(\boldsymbol{\Sigma}\)</span>, so each one implies the other:</p>
<p><span class="math display">\[
\mathbf{y}_{p \times 1} \sim \mathcal{N}(\boldsymbol{\mu}, \boldsymbol{\Sigma}) \Longleftrightarrow D^2_M (\mathbf{y}) \sim \chi^2_p\; .
\]</span></p>
<p>This fact can be used to assess whether sample data in <span class="math inline">\(\mathbf{y}\)</span> does indeed follow a MVN distribution by plotting the <em>quantiles</em><a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a> of the <em>sorted</em> sample Mahalanobis <span class="math inline">\(D^2\)</span> values (denoted <span class="math inline">\(D^2_{(i)}\)</span>) calculated from <a href="#eq-Dsq" class="quarto-xref">Equation <span class="quarto-unresolved-ref">eq-Dsq</span></a> against the corresponding <span class="math inline">\(\chi^2_p\)</span> quantiles found from <code>qchisq(df = p)</code>. This is called a <span class="math inline">\(\chi^2\)</span> QQ plot.</p>
<p>The essential idea is that the plotted points should then approximately fall along a 45<span class="math inline">\(^o\)</span> line of slope = 1 when the data are MVN (and axes are scaled equally). This also provides a simple method to identify potential outliers as those points which are furthest from the centroid; that is those for which <span class="math inline">\(D^2_{(i)}\)</span> is greater than the <span class="math inline">\(1 - \alpha\)</span> quantile of <span class="math inline">\(\chi^2_p\)</span>.</p>
<p>The topics of assessing multivariate normality and detecting multivariate outliers are larger than I consider here. These topics are also intertwined, because outliers inflate the variance (<span class="math inline">\(\mathbf{S}\)</span>) of the data, making extreme observations appear less extreme. A variety of <em>robust</em> methods described later (<a href="#sec-robust-estimation" class="quarto-xref"><span class="quarto-unresolved-ref">sec-robust-estimation</span></a>) work by down-weighting outliers, which is particularly important in assessing whether the residuals from a multivariate linear model are multivariate normal.</p>
<p>In this section, I focus on the <span class="math inline">\(\chi^2\)</span> QQ plot and graphical methods to relate the the points identified as potential outliers to plots in data space.</p>
<section id="galton-data" class="level3" data-number="4.6.1">
<h3 data-number="4.6.1"><span class="header-section-number">4.6.1</span> Galton data</h3>
<p>Mahalanobis <span class="math inline">\(D^2\)</span> values are calculated by <code>heplots::Mahalanobis()</code>. I illustrate finding the largest <span class="math inline">\(D^2_{(i)}\)</span> for the <code>Galton</code> data as shown below. I’ve used <span class="math inline">\(\alpha = 0.01\)</span>, giving <span class="math inline">\(\chi^2_p (0.99) = 9.21\)</span> as an outlier cutoff, just for the sake of this example. With 928 cases, we would expect about 1% or 9 to have larger squared distances than this cutoff. (Normally, allowing for the fact that we are looking at the largest values in a sample of size <span class="math inline">\(n\)</span>, we would use a much smaller individual significance level, say <span class="math inline">\(\alpha = 0.001\)</span> or smaller still.) Three cases are identified here.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Galton, <span class="at">package =</span> <span class="st">&quot;HistData&quot;</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>DSQ <span class="ot">&lt;-</span> <span class="fu">Mahalanobis</span>(Galton)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.01</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>cutoff <span class="ot">&lt;-</span> (<span class="fu">qchisq</span>(<span class="at">p =</span> <span class="dv">1</span> <span class="sc">-</span> alpha, <span class="at">df =</span> <span class="fu">ncol</span>(Galton))) <span class="sc">|&gt;</span> </span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 9.21</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> <span class="fu">which</span>(DSQ <span class="sc">&gt;</span> cutoff) <span class="sc">|&gt;</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># [1]   1  13 897</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>GaltonD <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Galton, <span class="at">DSQ =</span> DSQ)</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>GaltonD[outliers,]</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="co">#     parent child   DSQ</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 1     70.5  61.7 13.67</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># 13    70.5  63.2  9.45</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="co"># 897   65.5  72.2  9.49</span></span></code></pre></div>
</div>
<p><span class="math inline">\(\chi^2\)</span> QQ plots are constructed by <code>heplots::cqplot()</code>. For the Galton data, the result is shown in <a href="#fig-galton-cqplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-cqplot</span></a>, where I’ve asked for the <code>id.n = 3</code> with the greatest <span class="math inline">\(D^2_{(i)}\)</span> values to be identified with their row numbers in the plot. The function returns (invisibly) the <span class="math inline">\(D^2_{(i)}\)</span> and corresponding <span class="math inline">\(\chi^2_2\)</span> quantile along with the upper-tail <span class="math inline">\(p\)</span>-value.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb33" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">cqplot</span>(Galton, <span class="at">id.n =</span> <span class="dv">3</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>out</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">#       DSQ quantile        p</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1   13.67     15.1 0.000539</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 897  9.49     12.9 0.001616</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 13   9.45     11.8 0.002694</span></span></code></pre></div>
<div class="cell-output-display">
<div id="fig-galton-cqplot" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-galton-cqplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-galton-cqplot-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-galton-cqplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.24: Chi-square QQ plot of Galton’s data on heights of parents and their offspring, with a 95% pointwise confidence envelope
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the typical use of QQ plots, it essential to have something in the nature of a confidence band around the points to be able to appreciate whether, and to what degree the observed data points differ from the reference distribution. For <code>cqplot()</code>, this helps to assess whether the data are reasonably distributed as multivariate normal and also to flag potential outliers.</p>
<p>The pointwise 95% confidence envelope is calculated as <span class="math inline">\(D^2_{(i)} \pm 1.96 \times \text{se} ( D^2_{(i)} )\)</span> where the standard error is calculated <span class="citation" data-cites="Chambers-etal:83">(<a href="#ref-Chambers-etal:83" role="doc-biblioref">Chambers et al., 1983, sec. 8.6</a>)</span> as</p>
<p><span class="math display">\[
\text{se} ( D^2_{(i)} ) = \frac{\hat{b}}{d ( q_i)} \times \sqrt{  p_i (1-p_i) / n }  \:\: .
\]</span> Here, <span class="math inline">\(\hat{b}\)</span> is an estimate of the <em>actual</em> slope of the reference line obtained from from the ratio of the interquartile range of the <span class="math inline">\(D^2\)</span> values to that of the corresponding <span class="math inline">\(\chi^2_p\)</span> distribution. <span class="math inline">\(d(q_i)\)</span> is the density of the chi square distribution at the quantile <span class="math inline">\(q_i\)</span> and <span class="math inline">\(p_i\)</span> is the corresponding percentile.</p>
<p>So, in <a href="#fig-galton-cqplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-cqplot</span></a>, you can see the same three points, with largest <span class="math inline">\(D^2\)</span> identified earlier. The confidence band shows that uncertainty increases a points get further from the mean, but by and large, nearly all the points lie within it. The fact that larger values fall <em>beneath</em> the reference line indicates that the sample <span class="math inline">\(D^2\)</span> are somewhat <em>less concentrated</em> (has a shorter upper tail) than values in the <span class="math inline">\(\chi^2_2\)</span> distribution. </p>
<p>To help understand what we are seeing in this example, it is helpful to view the data in a scatterplot equivalent of <a href="#fig-galton-ellipse-r" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-galton-ellipse-r</span></a>, with the same three observations labeled and made distinctive in the plot. For this plot, because the heights are recorded in whole inches, I draw the data ellipses first without plotting the points and then draw the points using <code>jitter()</code> on top of the data ellipses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">47</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dataEllipse</span>(parent <span class="sc">~</span> child, <span class="at">data =</span> GaltonD,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">levels =</span> <span class="fu">c</span>(<span class="fl">0.68</span>, <span class="fl">0.95</span>),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">add =</span> <span class="cn">FALSE</span>, <span class="at">plot.points =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">center.pch =</span> <span class="st">&quot;+&quot;</span>, <span class="at">center.cex =</span> <span class="dv">3</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex.lab =</span> <span class="fl">1.5</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(GaltonD,{</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">points</span>(<span class="fu">jitter</span>(child), <span class="fu">jitter</span>(parent),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="fu">ifelse</span>(DSQ <span class="sc">&gt;</span> cutoff, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;black&quot;</span>),</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">pch =</span> <span class="fu">ifelse</span>(DSQ <span class="sc">&gt;</span> cutoff, <span class="dv">16</span>, <span class="dv">1</span>),</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">cex =</span> <span class="fu">ifelse</span>(DSQ <span class="sc">&gt;</span> cutoff, <span class="dv">2</span>, <span class="fl">0.8</span>))</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">text</span>(child[outliers], parent[outliers], <span class="at">labels =</span> outliers, <span class="at">pos =</span> <span class="dv">3</span>)</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-galton-outliers" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-galton-outliers-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-galton-outliers-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-galton-outliers-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.25: Scatterplot of Galton’s data showing 68% and 95% data ellipses for the observations on parent and child height. The discrete points have been jittered to avoid overplotting. The three observations identified as having large <span class="math inline">\(D^2\)</span> values are labeled with their observation number and given distinctive size, color and shape.
</figcaption>
</figure>
</div>
</div>
</div>
<p>It is clear from this plot that case 897 is for an exceptionally tall child of quite short parents, while case 1 and 13 are from very short children of tall parents; you could call them poster children for regression toward the mean.</p>
</section>
<section id="sec-multnorm-penguin" class="level3" data-number="4.6.2">
<h3 data-number="4.6.2"><span class="header-section-number">4.6.2</span> Penguin data</h3>
<p>Let’s do a similar analysis to assess multivariate normality and identify possible outliers for the Penguin data. In the <code>cqplot()</code> shown in <a href="#fig-peng-cqplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-cqplot</span></a>, I use the same point symbols and colors as in <a href="#fig-peng-ggplot1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot1</span></a>. For illustration, I again label the three most extreme points.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>clr <span class="ot">&lt;-</span> <span class="fu">peng.colors</span>(<span class="st">&quot;dark&quot;</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>pch <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">19</span>, <span class="dv">17</span>, <span class="dv">15</span>)   <span class="co"># ggplot symbol defaults for a factor</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">cqplot</span>(peng[, <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>], </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">id.n =</span> <span class="dv">3</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">col =</span> clr[peng<span class="sc">$</span>species],</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>   <span class="at">pch =</span> pch[peng<span class="sc">$</span>species],</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>   <span class="at">ref.col =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>   <span class="at">what =</span> <span class="st">&quot;Penguin numeric variables&quot;</span>,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>   <span class="at">cex.lab =</span> <span class="fl">1.25</span>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>out</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co">#      DSQ quantile       p</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 283 27.8     17.6 0.00150</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a><span class="co"># 10  13.3     15.1 0.00450</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a><span class="co"># 35  12.4     13.9 0.00751</span></span></code></pre></div>
<div class="cell-output-display">
<div id="fig-peng-cqplot" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-cqplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-cqplot-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-cqplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.26: Chi-square QQ plot of the Pengiun data. The three cases with the largest <span class="math inline">\(D^2\)</span> values are identified with their case numbers.
</figcaption>
</figure>
</div>
</div>
</div>
<p>One point (case 283) stands out as an extreme multivariate outlier. The other two (10, 35) are well within the confidence envelope.</p>
<p>To relate this to the data, we can plot the data, as was done in <a href="#fig-peng-ggplot1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot1</span></a>, and label the points identified as noteworthy. It is a bit tricky to label points selectively in <code>ggplot2</code> when the criterion for which points to label is complex, or involves variables outside the data frame. Here, I create a logical variable <code>note</code>, which is <code>TRUE</code> for the noteworthy ones. This is then used to subset the data for <code>geom_text()</code> that writes the case id numbers. <a href="#fig-peng-ggplot-out" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot-out</span></a> shows the resulting plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>DSQ <span class="ot">&lt;-</span> <span class="fu">Mahalanobis</span>(peng[, <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>noteworthy <span class="ot">&lt;-</span> <span class="fu">order</span>(DSQ, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="sc">|&gt;</span> <span class="fu">print</span>()</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 283  10  35</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>peng_plot <span class="ot">&lt;-</span> peng <span class="sc">|&gt;</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;id&quot;</span>) <span class="sc">|&gt;</span> </span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">note =</span> id <span class="sc">%in%</span> noteworthy)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng_plot, </span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bill_length, <span class="at">y =</span> bill_depth,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> species, <span class="at">shape =</span> species, <span class="at">fill=</span>species)) <span class="sc">+</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size=</span>note), <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_size_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="dv">4</span>)) <span class="sc">+</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> <span class="fu">subset</span>(peng_plot, note<span class="sc">==</span><span class="cn">TRUE</span>),</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">label =</span> id), <span class="at">fontface =</span> <span class="st">&quot;bold&quot;</span>,</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">nudge_y =</span> .<span class="dv">4</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y <span class="sc">~</span> x,</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>              <span class="at">se=</span><span class="cn">FALSE</span>, <span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.95</span>, <span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>() <span class="sc">+</span></span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">legend_inside</span>(<span class="fu">c</span>(<span class="fl">0.85</span>, <span class="fl">0.15</span>)) </span></code></pre></div>
<div class="cell-output-display">
<div id="fig-peng-ggplot-out" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggplot-out-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-ggplot-out-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:70.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggplot-out-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.27: Plot of bill length and bill depth, with the noteworthy points labeled. Only one (case 283) is a true multivariate outlier.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Two cases (10, 283) appear to be very unusual in this plot, in relation to other members of their species. But only case 283 is a true multivariate outlier, as shown in the cqplot <a href="#fig-peng-cqplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-cqplot</span></a>. We could call this long-billed penguin “Cyrano”.<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> I’ll call case 10, with a very short and curved bill, “Hook Nose”.</p>
<p>Of course, we are only looking at the data in the 2D space of the bill variables, but possible outliers exist in four dimensional Penguin space. Case 35 is well inside the Adélie ellipse, so perhaps it is unusual on one of the other variables. It turns out that multivariate outliers can most often be easily seen as unusual observations in a projection of the data into the space of the <em>smallest</em> principal components. I return to this example in <a href="#sec-outlier-detection" class="quarto-xref"><span class="quarto-unresolved-ref">sec-outlier-detection</span></a>.</p>
<p>It bears noting that for linear models, multivariate normality is <strong>not required</strong> for the response variables or predictors, but rather is an assumption for the <em>residuals</em> from a model. As well, multivariate outliers in the responses may not turn out to be unusual when the predictor variables are taken into account. In this example, a multivariate model would include the effect of <code>species</code>,</p>
<div class="sourceCode" id="cb37" data-code-line-numbers=""><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>peng.mod <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="fu">cbind</span>(bill_length, bill_depth, flipper_length, body_mass) <span class="sc">~</span> species, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">data =</span> peng)</span></code></pre></div>
<p>and there would be cause for concern if the residuals from this model, <code>residuals(peng.mod)</code> were highly non-normal or showed outliers.</p>
<!--
This example also illustrates why outlier detection requires a multivariate rather than univariate approach
and why it is necessary to take the correlations among variables into account with Malahanobis squared distance.
In the code below, I find the observations with the three largest $D^2_{(i)}$ and convert the numeric variables
to standardized $z$ scores, $z_i = (y_i - \bar{y}) / s_y$.


::: {.cell layout-align="center"}

```{.r .cell-code}
DSQ <- Mahalanobis(peng[, 3:6])
noteworthy <- order(DSQ, decreasing = TRUE)[1:3] |> print()
# [1] 283  10  35

peng |>
  tibble::rownames_to_column(var = "id") |> 
  mutate(
    z_BL = c(scale(bill_length)),
    z_BD = c(scale(bill_depth)),
    z_FL = c(scale(flipper_length)),
    z_BM = c(scale(body_mass))
  ) |>
  filter(id %in% noteworthy) |>
  select(id, species, z_BL:z_BM) 
# # A tibble: 3 × 6
#   id    species     z_BL  z_BD   z_FL   z_BM
#   <chr> <fct>      <dbl> <dbl>  <dbl>  <dbl>
# 1 10    Adelie    -1.72  2.00  -0.212  0.240
# 2 35    Adelie    -0.767 0.983 -1.21   0.550
# 3 283   Chinstrap  2.56  0.323 -1.42  -0.630
```
:::

-->
</section>
</section>
<section id="sec-scatmat" class="level2" data-number="4.7">
<h2 data-number="4.7"><span class="header-section-number">4.7</span> Scatterplot matrices</h2>
<p>Going beyond bivariate scatterplots, a <em>pairs</em> plot (or <em>scatterplot matrix</em>) displays all possible <span class="math inline">\(p \times p\)</span> pairs of <span class="math inline">\(p\)</span> variables in a matrix-like display where variables <span class="math inline">\((x_i, x_j)\)</span> are shown in a plot for row <span class="math inline">\(i\)</span>, column <span class="math inline">\(j\)</span>. This idea, due to <span class="citation" data-cites="Hartigan:75b">Hartigan (<a href="#ref-Hartigan:75b" role="doc-biblioref">1975b</a>)</span>, uses small multiple plots, so that the eye can easily scan across a row or down a column to see how a given variable is related to all the others.</p>
<p>The most basic version is provided by <code>pairs()</code> in base R. When one variable is considered as an outcome or response, it is usually helpful to put this in the first row and column. For the <code>Prestige</code> data, in addition to income and education, we also have a measure of % women in each occupational category.</p>
<p>Plotting these together gives <a href="#fig-prestige-pairs" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-prestige-pairs</span></a>. In such plots, the diagonal cells give labels for the variables, but they are also a guide to interpreting what is shown. In each row, say row 2 for <code>income</code>, income is the vertical <span class="math inline">\(y\)</span> variable in plots against other variables. In each column, say column 3 for <code>education</code>, education is the horizontal <span class="math inline">\(x\)</span> variable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(<span class="sc">~</span> prestige <span class="sc">+</span> income <span class="sc">+</span> education <span class="sc">+</span> women,</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">data=</span>Prestige)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-prestige-pairs" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-prestige-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-prestige-pairs-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-prestige-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.28: Scatterplot matrix of the variables in the Prestige dataset produced by <code>pairs()</code>
</figcaption>
</figure>
</div>
</div>
</div>
<p>The plots in the first row show what we have seen before for the relations between prestige and income and education, adding to those the plot of prestige vs. % women. Plots in the first column show the same data, but with <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> interchanged.</p>
<p>But this basic <code>pairs()</code> plot is very limited. A more feature-rich version is provided by <code>car::scatterplotMatrix()</code> which can add the regression lines, loess smooths and data ellipses for each pair, as shown in <a href="#fig-prestige-spm1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-prestige-spm1</span></a>.</p>
<p>The diagonal panels show density curves for the distribution of each variable; for example, <code>education</code> appears to be multi-modal and that of <code>women</code> shows that most of the occupations have a low percentage of women.</p>
<p>The combination of the regression line with the loess smoothed curve, but without their confidence envelopes, provides about the right amount of detail to take in at a glance where the relations are non-linear. We’ve already seen (<a href="#fig-Prestige-scatterplot-income1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Prestige-scatterplot-income1</span></a>) the non-linear relation between prestige and income (row 1, column 2) when occupational type is ignored. But all relations with income in column 2 are non-linear, reinforcing our idea (<a href="#sec-log-scale" class="quarto-xref"><span class="quarto-unresolved-ref">sec-log-scale</span></a>) that effects of income should be assessed on a log scale.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplotMatrix</span>(<span class="sc">~</span> prestige <span class="sc">+</span> income <span class="sc">+</span> education <span class="sc">+</span> women,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data=</span>Prestige,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">method=</span>lm, <span class="at">lty=</span><span class="dv">1</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">&quot;black&quot;</span>),</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth=</span><span class="fu">list</span>(<span class="at">smoother=</span>loessLine, <span class="at">spread=</span><span class="cn">FALSE</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">lty.smooth=</span><span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>, <span class="at">col.smooth=</span><span class="st">&quot;red&quot;</span>),</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse=</span><span class="fu">list</span>(<span class="at">levels=</span><span class="fl">0.68</span>, <span class="at">fill.alpha=</span><span class="fl">0.1</span>))</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-prestige-spm1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-prestige-spm1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-prestige-spm1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-prestige-spm1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.29: Scatterplot matrix of the variables in the Prestige dataset from <code>car::scatterplotMatrix()</code>.
</figcaption>
</figure>
</div>
</div>
</div>
<p><code>scatterplotMatrix()</code> can also label points using the <code>id =</code> argument (though this can get messy) and can stratify the observations by a grouping variable with different symbols and colors. For example, <a href="#fig-prestige-spm2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-prestige-spm2</span></a> uses the syntax <code>~ prestige + education + income + women | type</code> to provide separate regression lines, smoothed curves and data ellipses for the three types of occupations. (The default colors are somewhat garish, so I use <code>scales::hue_pal()</code> to mimic the discrete color scale used in <code>ggplot2</code>).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplotMatrix</span>(<span class="sc">~</span> prestige <span class="sc">+</span> income <span class="sc">+</span> education <span class="sc">+</span> women <span class="sc">|</span> type,</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> Prestige,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">hue_pal</span>()(<span class="dv">3</span>),</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">15</span><span class="sc">:</span><span class="dv">17</span>,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth=</span><span class="fu">list</span>(<span class="at">smoother=</span>loessLine, <span class="at">spread=</span><span class="cn">FALSE</span>,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">lty.smooth=</span><span class="dv">1</span>, <span class="at">lwd.smooth=</span><span class="dv">3</span>, <span class="at">col.smooth=</span><span class="st">&quot;black&quot;</span>),</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse=</span><span class="fu">list</span>(<span class="at">levels=</span><span class="fl">0.68</span>, <span class="at">fill.alpha=</span><span class="fl">0.1</span>))</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-prestige-spm2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-prestige-spm2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-prestige-spm2-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-prestige-spm2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.30: Scatterplot matrix of the variables in the Prestige dataset from <code>car::scatterplotMatrix()</code>, stratified by type of occupation.
</figcaption>
</figure>
</div>
</div>
</div>
<p>It is now easy to see why education is multi-modal: blue collar, white collar and professional occupations have largely non-overlapping years of education. As well, the distribution of % women is much higher in the white collar category.</p>
<p>For the <code>penguins</code> data, given what we’ve seen before in <a href="#fig-peng-ggplot1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot1</span></a> and <a href="#fig-peng-ggplot2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggplot2</span></a>, we may wish to suppress details of the points (<code>plot.points = FALSE</code>) and loess smooths (<code>smooth = FALSE</code>) to focus attention on the similarity of regression lines and data ellipses for the three penguin species. In <a href="#fig-peng-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-spm</span></a>, I’ve chosen to show boxplots rather than density curves in the diagonal panels in order to highlight differences in the means and interquartile ranges of the species, and to show 68% and 95% data ellipses in the off-diagonal panels.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplotMatrix</span>(<span class="sc">~</span> bill_length <span class="sc">+</span> bill_depth <span class="sc">+</span> flipper_length <span class="sc">+</span> body_mass <span class="sc">|</span> species,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> peng, </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">peng.colors</span>(<span class="st">&quot;medium&quot;</span>), </span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend=</span><span class="cn">FALSE</span>,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fu">c</span>(<span class="fl">0.68</span>, <span class="fl">0.95</span>), </span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill.alpha =</span> <span class="fl">0.1</span>),</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">regLine =</span> <span class="fu">list</span>(<span class="at">lwd=</span><span class="dv">3</span>),</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">diagonal =</span> <span class="fu">list</span>(<span class="at">method =</span> <span class="st">&quot;boxplot&quot;</span>),</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">smooth =</span> <span class="cn">FALSE</span>,</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot.points =</span> <span class="cn">FALSE</span>,</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex.labels=</span><span class="dv">1</span>) </span></code></pre></div>
<div class="cell-output-display">
<div id="fig-peng-spm" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-spm-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.31: Scatterplot matrix of the quantitative variables in the penguins dataset, stratified by species.
</figcaption>
</figure>
</div>
</div>
</div>
<!--# I added cex.labels=1 in the code above because the diag labels were on top of the boxplots, I think it looks a bit better now, but still slightly obscured-->
<p>It can be seen that the species are widely separated in most of the bivariate plots. As well, the regression lines for species have similar slopes and the data ellipses have similar size and shape in most of the plots. From the boxplots, we can also see that <span style="color: orange;">Adelie</span> penguins have shorter bill lengths than the others, while <span style="color: green;">Gentoo</span> penguins have smaller bill depth, but longer flippers and are heavier than <span style="color: purple;">Chinstrap</span> and <span style="color: orange;">Adelie</span> penguins.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Looking ahead">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Looking ahead
</div>
</div>
<div class="callout-body-container callout-body">
<p><a href="#fig-peng-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-spm</span></a> provides a reasonably complete visual summary of the data in relation to multivariate models that ask “do the species differ in their means on these body size measures?” This corresponds to the MANOVA model,</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>peng.mod <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="fu">cbind</span>(bill_length, bill_depth, flipper_length, body_mass) <span class="sc">~</span> species, </span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">data=</span>peng)</span></code></pre></div>
</div>
<p>Hypothesis-error (HE) plots, described in <a href="#sec-vis-mlm" class="quarto-xref"><span class="quarto-unresolved-ref">sec-vis-mlm</span></a> provide a better summary of the evidence for the MANOVA test of differences among means on all variables together. These give an <span class="math inline">\(\mathbf{H}\)</span> ellipse reflecting the differences among means, to be compared with an <span class="math inline">\(\mathbf{E}\)</span> ellipse reflecting within-group variation and a visual test of significance.</p>
<p>A related question is “how well are the penguin species distinguished by these body size measures?” Here, the relevant model is linear discriminant analysis (LDA), where <code>species</code> plays the role of the response in the model,</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>peng.lda <span class="ot">&lt;-</span> MASS<span class="sc">:</span><span class="fu">lda</span>( species <span class="sc">~</span> <span class="fu">cbind</span>(bill_length, bill_depth, flipper_length, body_mass), </span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">data=</span>peng)</span></code></pre></div>
</div>
<p>Both MANOVA and LDA depend on the assumption that the variances and correlations between the variables are the same for all groups. This assumption can be tested and visualized using the methods in <a href="#sec-eqcov" class="quarto-xref"><span class="quarto-unresolved-ref">sec-eqcov</span></a>.</p>
</div>
</div>
<section id="visual-thinning" class="level3" data-number="4.7.1">
<h3 data-number="4.7.1"><span class="header-section-number">4.7.1</span> Visual thinning</h3>
<p>What can you do if there are even more variables than in these examples? If what you want is a high-level, zoomed-out display summarizing the pairwise relations more strongly, you can apply the idea of visual thinning to show only the most important features.</p>
<p>This example uses data on the rate of various crimes in the 50 U.S. states from the United States Statistical Abstracts, 1970, used by <span class="citation" data-cites="Hartigan:75">Hartigan (<a href="#ref-Hartigan:75" role="doc-biblioref">1975a</a>)</span> and <span class="citation" data-cites="Friendly:91">Friendly (<a href="#ref-Friendly:91" role="doc-biblioref">1991</a>)</span>. These are ordered in the dataset roughly by seriousness of crime or from crimes of violence to property crimes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(crime, <span class="at">package =</span> <span class="st">&quot;ggbiplot&quot;</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(crime)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;data.frame&#39;: 50 obs. of  10 variables:</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ state   : chr  &quot;Alabama&quot; &quot;Alaska&quot; &quot;Arizona&quot; &quot;Arkansas&quot; ...</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ murder  : num  14.2 10.8 9.5 8.8 11.5 6.3 4.2 6 10.2 11.7 ...</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ rape    : num  25.2 51.6 34.2 27.6 49.4 42 16.8 24.9 39.6 31.1 ...</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ robbery : num  96.8 96.8 138.2 83.2 287 ...</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ assault : num  278 284 312 203 358 ...</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ burglary: num  1136 1332 2346 973 2139 ...</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ larceny : num  1882 3370 4467 1862 3500 ...</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ auto    : num  281 753 440 183 664 ...</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ st      : chr  &quot;AL&quot; &quot;AK&quot; &quot;AZ&quot; &quot;AR&quot; ...</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a><span class="co">#  $ region  : Factor w/ 4 levels &quot;Northeast&quot;,&quot;South&quot;,..: 2 4 4 2 4 4 1 2 2 2 ...</span></span></code></pre></div>
</div>
<p><a href="#fig-crime-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-spm</span></a> displays the scatterplot matrix for these seven variables, using only the regression line and data ellipse to show the linear relation and the loess smooth to show potential non-linearity. <!--# I believe you used non-linear, hyphenated earlier so I added the - --> To make this even more schematic, the axis tick marks and labels are also removed using the <code>par()</code> settings <code>xaxt = "n", yaxt = "n"</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>crime <span class="sc">|&gt;</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">|&gt;</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scatterplotMatrix</span>(</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.points =</span> <span class="cn">FALSE</span>,</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ellipse =</span> <span class="fu">list</span>(<span class="at">levels =</span> <span class="fl">0.68</span>, <span class="at">fill=</span><span class="cn">FALSE</span>),</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> <span class="fu">list</span>(<span class="at">spread =</span> <span class="cn">FALSE</span>, </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">lwd.smooth=</span><span class="dv">2</span>, <span class="at">lty.smooth =</span> <span class="dv">1</span>, </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>                  <span class="at">col.smooth =</span> <span class="st">&quot;red&quot;</span>),</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.labels =</span> <span class="dv">2</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="at">yaxt =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-crime-spm" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-crime-spm-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.32: <strong>Visual thinning</strong>: Scatterplot matrix of the crime data, showing only high-level summaries of the linear and nonlinear relations betgween each pair of variables.
</figcaption>
</figure>
</div>
</div>
</div>
<p>We can see that all pairwise correlations are positive, pairs closer to the main diagonal tend to be more highly correlated and in most cases the nonparametric smooth doesn’t differ much from the linear regression line. Exceptions to this appear mainly in the columns for <code>robbery</code> and <code>auto</code> (auto theft).</p>
</section>
</section>
<section id="sec-corrgram" class="level2" data-number="4.8">
<h2 data-number="4.8"><span class="header-section-number">4.8</span> Corrgrams</h2>
<p><strong>TODO</strong>: Perhaps split the chapter here</p>
<p>What if you want to summarize the data even further simple visual thinning. For example with many variables you might want to show only the value of the correlation for each pair of variables, but do so in a way to help see patterns in the correlations that would be invisible in just a table.</p>
<p>A <strong>corrgram</strong> <span class="citation" data-cites="Friendly:02:corrgram">(<a href="#ref-Friendly:02:corrgram" role="doc-biblioref">Friendly, 2002</a>)</span> is a visual display of a correlation matrix, where the correlation can be rendered in a variety of ways to show the direction and magnitude: circular “pac-man” (or pie) symbols, ellipses, colored vars or shaded rectangles, as shown in <a href="#fig-corrgram-renderings" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-corrgram-renderings</span></a>.</p>
<p>Another aspect is that of <strong>effect ordering</strong> <span class="citation" data-cites="FriendlyKwan:03:effect">(<a href="#ref-FriendlyKwan:03:effect" role="doc-biblioref">Friendly &amp; Kwan, 2003</a>)</span>, ordering the levels of factors and variables in graphic displays to make important features most apparent. For variables, this means that we can arrange the variables in a matrix-like display in such a way as to make the pattern of relationships easiest to see. Methods to achieve this include using principal components and cluster analysis to put the most related variables together as described in <a href="#sec-pca-biplot" class="quarto-xref"><span class="quarto-unresolved-ref">sec-pca-biplot</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-corrgram-renderings" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-corrgram-renderings-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/corrgram-renderings.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-corrgram-renderings-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.33: <strong>Corrgrams</strong>: Some renderings for the value of a correlation in a corrgram display, conveying sign and magnitude in different ways.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In R, these diagrams can be created using the <span style="color: brown;"><strong>corrgram</strong></span> <span class="citation" data-cites="R-corrgram">(<a href="#ref-R-corrgram" role="doc-biblioref">Wright, 2021</a>)</span> and <span style="color: brown;"><strong>corrplot</strong></span> <span class="citation" data-cites="R-corrplot">(<a href="#ref-R-corrplot" role="doc-biblioref">Wei &amp; Simko, 2024</a>)</span> packages, with different features. <code>corrgram::corrgram()</code> is closest to <span class="citation" data-cites="Friendly:02:corrgram">Friendly (<a href="#ref-Friendly:02:corrgram" role="doc-biblioref">2002</a>)</span>, in that it allows different rendering functions for the lower, upper and diagonal panels as illustrated in <a href="#fig-corrgram-renderings" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-corrgram-renderings</span></a>. For example, a corrgram similar to <a href="#fig-crime-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-spm</span></a> can be produced as follows (not shown here):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>crime <span class="sc">|&gt;</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">|&gt;</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">corrgram</span>(<span class="at">lower.panel =</span> panel.ellipse,</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">upper.panel =</span> panel.ellipse,</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">diag.panel =</span> panel.density)</span></code></pre></div>
</div>
<p>With the <span style="color: brown;"><strong>corrplot</strong></span> package, <code>corrplot()</code> provides the rendering methods <code>c("circle", "square", "ellipse", "number", "shade", "color", "pie")</code>, but only one can be used at a time. The function <code>corrplot.mixed()</code> allows different options to be selected for the lower and upper triangles. The iconic rendering shape is colored with a gradient in relation to the correlation value. For comparison, <a href="#fig-crime-corrplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-corrplot</span></a> uses ellipses below the diagonal and filled pie charts below the diagonal using a gradient of the fill color in both cases.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb47" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>crime.cor <span class="ot">&lt;-</span> crime <span class="sc">|&gt;</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">|&gt;</span> </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>()</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">corrplot.mixed</span>(crime.cor,</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">lower =</span> <span class="st">&quot;ellipse&quot;</span>,</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>   <span class="at">upper =</span> <span class="st">&quot;pie&quot;</span>,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>   <span class="at">tl.col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>   <span class="at">tl.srt =</span> <span class="dv">0</span>,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>   <span class="at">tl.cex =</span> <span class="fl">1.25</span>,</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>   <span class="at">addCoef.col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>   <span class="at">addCoefasPercent =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-crime-corrplot" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-corrplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-crime-corrplot-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-corrplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.34: Mixed corrplot of the <code>crime</code> data, showing the correlation between each pair of variables with an ellipse (lower) and a pie chart symbol (upper), all shaded in proportion to the correlation value, also shown numerically.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The combination of renderings shown in <a href="#fig-crime-corrplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-corrplot</span></a> is instructive. Small differences among correlation values are easier to see with the pie symbols than with the ellipses; for example, compare the values for murder with larceny and auto theft in row 1, columns 6-7 with those in column 1, rows 6-7, where the former are easier to distinguish. The shading color adds another visual cue.</p>
<p>The variables in <a href="#fig-crime-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-spm</span></a> and <a href="#fig-crime-corrplot" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-corrplot</span></a> are arranged by their order in the dataset, which is not often the most useful. A better idea is to arrange the variables so that the most highly correlated variables are adjacent.</p>
<p>A general method described in <a href="#sec-var-order" class="quarto-xref"><span class="quarto-unresolved-ref">sec-var-order</span></a> orders the variables according to the angles of the first two eigenvectors from a principal components analysis (PCA) around a unit circle. The function <code>corrMatOrder()</code> provides several methods (<code>order = c("AOE", "FPC", "hclust", "alphabet")</code>) for doing this, and PCA ordering is <code>order = "AOE"</code>. Murder and auto theft are still first and last, but some of the intermediate crimes have been rearranged.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb48" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">corrMatOrder</span>(crime.cor, <span class="at">order =</span> <span class="st">&quot;AOE&quot;</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(crime.cor)[ord]</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;murder&quot;   &quot;assault&quot;  &quot;rape&quot;     &quot;robbery&quot;  &quot;burglary&quot;</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [6] &quot;larceny&quot;  &quot;auto&quot;</span></span></code></pre></div>
</div>
<!-- other orders: FPC (PC1), hclust (with various clustering methods), alphabet, original -->
<!-- calculated with corrMatOrder() -->
<p>Using this ordering in <code>corrplot()</code> produces <a href="#fig-crime-corrplot-AOE" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-corrplot-AOE</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb49" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">corrplot.mixed</span>(crime.cor,</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">order =</span> <span class="st">&quot;AOE&quot;</span>, </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lower =</span> <span class="st">&quot;ellipse&quot;</span>,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">upper =</span> <span class="st">&quot;ellipse&quot;</span>,</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">tl.col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">tl.srt =</span> <span class="dv">0</span>,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">tl.cex =</span> <span class="fl">1.25</span>,</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">addCoef.col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">addCoefasPercent =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-crime-corrplot-AOE" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-corrplot-AOE-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-crime-corrplot-AOE-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-corrplot-AOE-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.35: Corrplot of the <code>crime</code> data with the variables reordered according to the angles of variable eigenvectors. Correlations are rendered with ellipses shaded in proportion to their magnitude.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In this case, where the correlations among the crime variables are all positive, the effect of variable re-ordering is subtle, but note that there is now a slightly pronounced pattern of highest correlations near the diagonal, and decreasing away from the diagonal. <a href="#fig-mtcars-corrplot-varorder" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-mtcars-corrplot-varorder</span></a> and <a href="#fig-mtcars-corrplot-pcaorder" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-mtcars-corrplot-pcaorder</span></a> in <a href="#sec-var-order" class="quarto-xref"><span class="quarto-unresolved-ref">sec-var-order</span></a> provide a more dramatic example of variable ordering using this method.</p>
<p>Variations of corrgrams are worthy replacements for a numeric table of correlations, which are often presented in publications only for archival value. Including the numeric value (rounded here, for presentation purposes), makes this an attractive alternative to boring tables of correlations.</p>
</section>
<section id="sec-ggpairs" class="level2" data-number="4.9">
<h2 data-number="4.9"><span class="header-section-number">4.9</span> Generalized pairs plots</h2>
<p>When a dataset contains one or more discrete variables, the traditional pairs plot cannot cope, because the discrete categories would plot as many overlaid points. This cannot be represented using only color and/or point symbols in a meaningful scatterplot.</p>
<p>But the associations between categorical variables in a frequency table <em>can</em> be shown in <em>mosaic displays</em> <span class="citation" data-cites="Friendly:94a">(<a href="#ref-Friendly:94a" role="doc-biblioref">Friendly, 1994</a>)</span>, using an array of tiles whose areas are depict the cell frequencies. For an <span class="math inline">\(n\)</span>-way frequency, an analog of the scatterplot matrix uses mosaic plots for each pair of variables. The <span style="color: brown;"><strong>vcd</strong></span> package <span class="citation" data-cites="R-vcd">(<a href="#ref-R-vcd" role="doc-biblioref">Meyer et al., 2024</a>)</span> implements very general <code>pairs()</code> methods for <code>"table"</code> objects and <span style="color: brown;"><strong>vcdExtra</strong></span> <span class="citation" data-cites="R-vcdExtra">(<a href="#ref-R-vcdExtra" role="doc-biblioref">Friendly, 2025</a>)</span> extends this to wide classes of loglinear models <span class="citation" data-cites="Friendly:99:EMD">(<a href="#ref-Friendly:99:EMD" role="doc-biblioref">Friendly, 1999</a>)</span> See <span class="citation" data-cites="Friendly:99:EMD">Friendly (<a href="#ref-Friendly:99:EMD" role="doc-biblioref">1999</a>)</span> and my book <em>Discrete Data Analysis with R</em> <span class="citation" data-cites="FriendlyMeyer:2016:DDAR">(<a href="#ref-FriendlyMeyer:2016:DDAR" role="doc-biblioref">Friendly &amp; Meyer, 2016</a>)</span> for mosaic plots and matrices.</p>
<p>For example, we can tabulate the distributions of penguin species by sex and the island where they were observed using <code>xtabs()</code>. <code>ftable()</code> prints this three-way table more compactly. (In this example, and what follows in the chapter, I’ve changed the labels for sex from (“f”, “m”) to (“Female”, “Male”)).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb50" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use better labels for sex</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>peng <span class="ot">&lt;-</span> peng <span class="sc">|&gt;</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">factor</span>(sex, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;Female&quot;</span>, <span class="st">&quot;Male&quot;</span>)))</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>peng.table <span class="ot">&lt;-</span> <span class="fu">xtabs</span>(<span class="sc">~</span> species <span class="sc">+</span> sex <span class="sc">+</span> island, <span class="at">data =</span> peng)</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ftable</span>(peng.table)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                  island Biscoe Dream Torgersen</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="co"># species   sex                                 </span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Adelie    Female            22    27        24</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a><span class="co">#           Male              22    28        23</span></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Chinstrap Female             0    34         0</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="co">#           Male               0    34         0</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gentoo    Female            58     0         0</span></span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a><span class="co">#           Male              61     0         0</span></span></code></pre></div>
</div>
<p>We can see immediately that the penguin species differ by island: only Adélie were observed on all three islands; Biscoe Island had no Chinstraps and Dream Island had no Gentoos.</p>
<p><code>vcd::pairs()</code> produces all pairwise mosaic plots, as shown in <a href="#fig-peng-mosaic" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-mosaic</span></a>. The diagonal panels show the one-way frequencies by width of the divided bars. Each off-diagonal panel shows the bivariate counts, breaking down each column variable by splitting the bars in proportion to a second variable. Consequently, the frequency of each cell is represented by its’ area. The purpose is to show the <strong>pattern of association</strong> between each pair, and so, the tiles in the mosaic are shaded according to the signed standardized residual, <span class="math inline">\(d_{ij} = (n_{ij} - \hat{n}_{ij}) / \sqrt{\hat{n}_{ij}}\)</span> in a simple <span class="math inline">\(\chi^2 = \Sigma_{ij} \; d_{ij}^2\)</span> test for association— <span style="color: blue;">blue</span> where the observed frequency <span class="math inline">\(n_{ij}\)</span> is significantly greater than expected <span class="math inline">\(\hat{n}_{ij}\)</span> under independence, and <span style="color: red;">red</span> where it is less than expected. The tiles are unshaded when <span class="math inline">\(| d_{ij} | &lt; 2\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb51" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vcd)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(peng.table, <span class="at">shade =</span> <span class="cn">TRUE</span>,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">lower_panel_args =</span> <span class="fu">list</span>(<span class="at">labeling =</span> <span class="fu">labeling_values</span>()),</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">upper_panel_args =</span> <span class="fu">list</span>(<span class="at">labeling =</span> <span class="fu">labeling_values</span>()))</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-peng-mosaic" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-mosaic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-mosaic-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-mosaic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.36: Mosaic pairs plot for the combinations of species, sex and island. Diagnonal plots show the marginal frequency of each variable by the width of each rectangle. Off-diagonal mosaic plots subdivide by the conditional frequency of the second variable, shown numerically in the tiles. The tiles are shaded according to the departures from independence.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The shading patterns in cells (1,3) and (3,1) of <a href="#fig-peng-mosaic" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-mosaic</span></a> show what we’ve seen before in the table of frequencies: The distribution of species varies across island because on each island one or more species did not occur. Row 2 and column 2 show that sex is nearly exactly proportional among species and islands, indicating independence, <span class="math inline">\(\text{sex} \perp \{\text{species}, \text{island}\}\)</span>. More importantly, mosaic pairs plots can show, at a glance, all (bivariate) associations among multivariate categorical variables.</p>
<p>The next step, by John Emerson and others <span class="citation" data-cites="Emerson-etal:2013">(<a href="#ref-Emerson-etal:2013" role="doc-biblioref">Emerson et al., 2013</a>)</span> was to recognize that combinations of continuous and discrete, categorical variables could be plotted in different ways.</p>
<ul>
<li>Two continuous variables can be shown as a standard scatterplot of points and/or bivariate density contours, or simply by numeric summaries such as a correlation value;</li>
<li>A pair of one continuous and one categorical variable can be shown as side-by-side boxplots or violin plots, histograms or density plots;</li>
<li>Two categorical variables could be shown in a mosaic plot or by grouped bar plots.</li>
</ul>
<p>In the ggplot2 framework, these displays are implemented using the <code>ggpairs()</code> function from the <span style="color: brown;"><strong>GGally</strong></span> package <span class="citation" data-cites="R-GGally">(<a href="#ref-R-GGally" role="doc-biblioref">Schloerke et al., 2025</a>)</span>. This allows different plot types to be shown in the lower and upper triangles and in the diagonal cells of the plot matrix. As well, aesthetics such as color and shape can be used within the plots to distinguish groups directly. As illustrated below, you can define custom functions to control exactly what is plotted in any panel.</p>
<p>The basic, default plot shows scatterplots for pairs of continuous variables in the lower triangle and the values of correlations in the upper triangle. A combination of a discrete and continuous variables is plotted as histograms in the lower triangle and boxplots in the upper triangle. <a href="#fig-peng-ggpairs1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpairs1</span></a> includes <code>sex</code> to illustrate the combinations.</p>
<!-- fig.code: R/peng/peng-ggally.R -->
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb52" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(peng, <span class="at">columns=</span><span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, <span class="dv">7</span>),</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">color=</span>species, <span class="at">alpha=</span><span class="fl">0.5</span>),</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">progress =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>() <span class="sc">+</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="sc">-</span><span class="dv">45</span>))</span></code></pre></div>
</details>
</div>
<!-- this figure now frozen because it takes too long and caching creates git problems -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-ggpairs1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggpairs1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-peng-ggpairs1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggpairs1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.37: Basic <code>ggpairs()</code> plot of penguin size variables and sex, stratified by species.
</figcaption>
</figure>
</div>
</div>
</div>
<p>To my eye, printing the values of correlations in the upper triangle is often a waste of graphic space. But in this example the correlations show something peculiar and interesting if you look closely: In all pairs among the penguin size measurements, there are positive correlations within each species, as we can see in <a href="#fig-peng-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-spm</span></a>. Yet, in three of these panels, the overall correlation ignoring species is negative. For example, the overall correlation between bill depth and flipper length is <span class="math inline">\(r = -0.579\)</span> in row 2, column 3; the scatterplot in the diagonally opposite cell, row 3, column 2 shows the data. These cases, of differing signs for an overall correlation, ignoring a group variable and the within group correlations are examples of <strong>Simpson’s Paradox</strong>, explored later in <a href="#sec-marginal-conditional" class="quarto-xref"><span class="quarto-unresolved-ref">sec-marginal-conditional</span></a>.</p>
<p>The last row and column, for <code>sex</code> in <a href="#fig-peng-ggpairs1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpairs1</span></a>, provides an initial glance at the issue of sex differences among penguin species that motivated the collection of these data. We can go further by also examining differences among species and island, but first we need to understand how to display exactly what we want for each pairwise plot.</p>
<p><code>ggpairs()</code> is extremely general in that for each of the <code>lower</code>, <code>upper</code> and <code>diag</code> sections you can assign any of a large number of built-in functions (of the form <code>ggally_NAME</code>), or your own custom function for what is plotted, depending on the types of variables in each plot.</p>
<ul>
<li><code>continuous</code>: both X and Y are continuous variables, supply this as the <code>NAME</code> part of a <code>ggally_NAME()</code> function or the name of a custom function;</li>
<li><code>combo</code>: one X of and Y variable is discrete while the other is continuous, using the same convention;</li>
<li><code>discrete</code>: both X and Y are discrete variables.</li>
</ul>
<p>The defaults, which were used in <a href="#fig-peng-ggpairs1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpairs1</span></a>, are:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb53" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>upper <span class="ot">=</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="st">&quot;cor&quot;</span>,          <span class="co"># correlation values</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">combo =</span> <span class="st">&quot;box_no_facet&quot;</span>,      <span class="co"># boxplots </span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">discrete =</span> <span class="st">&quot;count&quot;</span>)          <span class="co"># rectangles ~ count</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>lower <span class="ot">=</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="st">&quot;points&quot;</span>,       <span class="co"># just data points</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">combo =</span> <span class="st">&quot;facethist&quot;</span>,         <span class="co"># faceted histograms</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">discrete =</span> <span class="st">&quot;facetbar&quot;</span>)       <span class="co"># faceted bar plots</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>diag  <span class="ot">=</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="st">&quot;densityDiag&quot;</span>,  <span class="co"># density plots</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">discrete =</span> <span class="st">&quot;barDiag&quot;</span>)        <span class="co"># bar plots</span></span></code></pre></div>
</div>
<p>Thus, <code>ggpairs()</code> uses <code>ggally_cor()</code> to print the correlation values for pairs of continuous variables in the upper triangle, and uses <code>ggally_points()</code> to plot scatterplots of points in the lower portion. The diagonal panels as shown as density plots (<code>ggally_densityDiag()</code>) for continuous variables but as bar plots (<code>ggally_barDiag()</code>) for discrete factors.</p>
<p>See the vignette, <a href="https://ggobi.github.io/ggally/articles/ggally_plots.html">ggally_plots</a> for an illustrated list of available high-level plots. For our purpose here, which is to illustrate enhanced displays, note that for scatterplots of continuous variables, there are two functions which plot the points and also add a smoother, <code>_lm</code> or <code>_loess</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb54" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>(<span class="fu">getNamespace</span>(<span class="st">&quot;GGally&quot;</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_subset</span>(<span class="st">&quot;^ggally_smooth_&quot;</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;ggally_smooth_lm&quot;    &quot;ggally_smooth_loess&quot;</span></span></code></pre></div>
</div>
<p>A customized display for scatterplots of continuous variables can be any function that takes <code>data</code> and <code>mapping</code> arguments and returns a <code>"ggplot"</code> object. The <code>mapping</code> argument supplies the aesthetics, e.g., <code>aes(color=species, alpha=0.5)</code>, but only if you wish to override what is supplied in the <code>ggpairs()</code> call.</p>
<p>Here is a function, <code>my_panel()</code> that plots the data points, regression line and loess smooth:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb55" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>my_panel <span class="ot">&lt;-</span> <span class="cf">function</span>(data, mapping, ...){</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> data, <span class="at">mapping =</span> mapping) <span class="sc">+</span> </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">formula =</span> y <span class="sc">~</span> x, <span class="at">se =</span> <span class="cn">FALSE</span>, ...) <span class="sc">+</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method=</span>loess, <span class="at">formula =</span> y <span class="sc">~</span> x, <span class="at">se =</span> <span class="cn">FALSE</span>, ...)</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  p</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<p>For this example, I want only simple summaries of for the scatterplots, so I don’t want to plot the data points, but do want to add the regression line and a data ellipse.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb56" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>my_panel1 <span class="ot">&lt;-</span> <span class="cf">function</span>(data, mapping, ...){</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> data, <span class="at">mapping =</span> mapping) <span class="sc">+</span> </span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">formula =</span> y <span class="sc">~</span> x, <span class="at">se =</span> <span class="cn">FALSE</span>, ...) <span class="sc">+</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">stat_ellipse</span>(<span class="at">geom =</span> <span class="st">&quot;polygon&quot;</span>, <span class="at">level =</span> <span class="fl">0.68</span>, ...)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  p</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<p>Then, to show what can be done, <a href="#fig-peng-ggpairs7" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpairs7</span></a> uses <code>my_panel1()</code> for the scatterplots in the 4 x 4 block of plots in the upper left. The combination of the continuous body size measures and the discrete factors <code>species</code>, <code>island</code> and <code>sex</code> are shown in upper triangle by boxplots but by faceted histograms in the lower portion. The factors are shown as rectangles with area proportional to count (poor-man’s mosaic plots) above the diagonal and as faceted bar plots below.</p>
<!-- fig.code: R/peng/peng-ggally.R -->
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb57" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(peng, <span class="at">columns=</span><span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span>),</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">color=</span>species, <span class="at">fill =</span> species, <span class="at">alpha=</span><span class="fl">0.2</span>),</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">lower =</span> <span class="fu">list</span>(<span class="at">continuous =</span> my_panel1),</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">upper =</span> <span class="fu">list</span>(<span class="at">continuous =</span> my_panel1),</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">progress =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_penguins</span>() <span class="sc">+</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="sc">-</span><span class="dv">45</span>))</span></code></pre></div>
</details>
</div>
<!-- this figure now frozen because it takes too long and caching creates git problems -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-ggpairs7" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggpairs7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/fig-peng-ggpairs7-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggpairs7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.38: Customized <code>ggpairs()</code> plot of penguin size variables, together with species, island and sex.
</figcaption>
</figure>
</div>
</div>
</div>
<p>There is certainly a lot going on in <a href="#fig-peng-ggpairs7" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpairs7</span></a>, but it does show a high-level overview of all the variables (except <code>year</code>) in the penguins dataset. It is probably easiest to “read” this figure by focusing on the four blocks for the combinations of 4 continuous and 3 categorical measures. In the upper left block, visual thinning of the scatterplots, showing only the data ellipses and regression lines gives a simple view as it did in <a href="#fig-peng-spm" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-spm</span></a>.</p>
</section>
<section id="sec-parcoord" class="level2" data-number="4.10">
<h2 data-number="4.10"><span class="header-section-number">4.10</span> Parallel coordinate plots</h2>
<p>As we have seen above, scatterplot matrices and generalized pairs plots extend data visualization to multivariate data, but these variables share one 2D space, so resolution decreases as the number of variable increase. You need a very large screen or sheet of paper to see more than, say 5-6 variables with any clarity.</p>
<p>Parallel coordinate plots are an attractive alternative, with which we can visualize an arbitrary number of variables to get a visual summary of a potentially high-dimensional dataset, and perhaps recognize outliers and clusters in the data in a different way. In these plots, each variable is shown on a separate, parallel axis. A multivariate observation is then plotted by connecting their respective values on each axis with lines across all the axes.</p>
<p>The geometry of parallel coordinates is interesting, because what is a point in <span class="math inline">\(n\)</span>-dimensional (Euclidean) <em>data</em> space becomes a line in the <em>projective</em> parallel coordinate space with <span class="math inline">\(n\)</span> axes, and vice-versa: lines in parallel coordinate space correspond to points in data space. Thus, a collection of points in data space map to lines that intersect in a point in projective space. What this does is to map <span class="math inline">\(n\)</span>-dimensional relations into 2D patterns we can see in a parallel coordinates plot.</p>
<div class="callout callout-style-default callout-note callout-titled" title="History Corner: Who invented parallel coordinates?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
History Corner: Who invented parallel coordinates?
</div>
</div>
<div class="callout-body-container callout-body">
<blockquote>
<p><em>Those who don’t know history are doomed to plagiarize it</em> —The author</p>
</blockquote>
<p>The theory of projective geometry originated with the French mathematician Maurice d’Ocagne <span class="citation" data-cites="Ocagne:1885">(<a href="#ref-Ocagne:1885" role="doc-biblioref">1885</a>)</span> who sought a way to provide graphic calculation of mathematical functions with alignment diagrams or <em>nomograms</em> using parallel axes with different scales. A three-variable equation, for example, could be solved using three parallel axes, where known values could be marked on their scales, a line drawn between them, and an unknown read on its scale at the point where the line intersects that scale.</p>
<p>Henry Gannet (1880), in work preceding the <em>Statistical Atlas of the United States</em> for the 1890 Census <span class="citation" data-cites="Gannett:1898">(<a href="#ref-Gannett:1898" role="doc-biblioref">Gannett, 1898</a>)</span>, is widely credited with being the first to use parallel coordinates plots to show data, in his case, to show the <a href="https://www.davidrumsey.com/luna/servlet/detail/RUMSEY~8~1~32803~1152181">rank ordering of US states</a> by 10 measures including population, occupations, wealth, manufacturing, agriculture and so on.</p>
<p>However, both d’Ocagne and Gannet were far preceded in this by Andre-Michel Guerry <span class="citation" data-cites="Guerry:1833">(<a href="#ref-Guerry:1833" role="doc-biblioref">1833</a>)</span> who used this method to show how the rank order of various crimes changed with age of the accused. See <span class="citation" data-cites="Friendly2022">Friendly (<a href="#ref-Friendly2022" role="doc-biblioref">2022</a>)</span>, Figure 7 for his version and for an appreciation of the remarkable contributions of this amateur statistician to the history of data visualization.</p>
<!-- **TODO**: Revise the _History_ section of the Wikipedia page for [Parallel coordinates](https://en.wikipedia.org/wiki/Parallel_coordinates). -->
<p>The use of parallel coordinates for display of multidimensional data was rediscovered by Alfred Inselberg <span class="citation" data-cites="Inselberg:1985">(<a href="#ref-Inselberg:1985" role="doc-biblioref">1985</a>)</span> and extended by Edward Wegman <span class="citation" data-cites="Wegman:1990">(<a href="#ref-Wegman:1990" role="doc-biblioref">1990</a>)</span>, neither of whom recognized the earlier history. Somewhat earlier, David Andrews <span class="citation" data-cites="Andrews:72">(<a href="#ref-Andrews:72" role="doc-biblioref">1972</a>)</span> proposed mapping multivariate observations to smooth Fourrier functions composed of alternating <span class="math inline">\(\sin()\)</span> and <span class="math inline">\(\cos()\)</span> terms. And in my book, <em>SAS System for Statistical Graphics</em> <span class="citation" data-cites="Friendly:91">(<a href="#ref-Friendly:91" role="doc-biblioref">Friendly, 1991</a>)</span>, I implemented what I called <a href="https://blogs.sas.com/content/iml/2022/11/14/profile-plots-sas.html"><em>profile plots</em></a> without knowing their earlier history as parallel coordinate plots.</p>
</div>
</div>
<p>Parallel coordinate plots present a challenge for graphic developers, in that they require a different way to think about plot construction for multiple variables, which can be quantitative, as in the original idea, or categorical factors, all to be shown along parallel axes.</p>
<p>Here, I use the <span style="color: brown;"><strong>ggpcp</strong></span> package <span class="citation" data-cites="R-ggpcp">(<a href="#ref-R-ggpcp" role="doc-biblioref">Hofmann et al., 2022</a>)</span>, best described in <span class="citation" data-cites="VanderPlas2023">VanderPlas et al. (<a href="#ref-VanderPlas2023" role="doc-biblioref">2023</a>)</span>, who also review the modern history.<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a> This takes some getting used to, because they develop <code>pcp_*()</code> extensions of the <code>ggplot2</code> grammar of graphics framework to allow:</p>
<ul>
<li><code>pcp_select()</code>: selections of the variables to be plotted and their horizontal order on parallel axes,</li>
<li><code>pcp_scale()</code>: methods for scaling of the variables to each axis,</li>
<li><code>pcp_arrange()</code>: methods for breaking ties in factor variables to space them out.</li>
</ul>
<p>Then, it provides <code>geom_pcp_*()</code> functions to control the display of axes with appropriate aesthetics, labels for categorical factors and so forth. <a href="#fig-peng-ggpcp1" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-ggpcp1</span></a> illustrates this type of display, using sex and species in addition to the quantitative variables for the penguin data.</p>
<!-- fig.code: R/peng/peng-ggpcp.R -->
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb58" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">|&gt;</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_select</span>(bill_length<span class="sc">:</span>body_mass, sex, species) <span class="sc">|&gt;</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_scale</span>(<span class="at">method =</span> <span class="st">&quot;uniminmax&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_arrange</span>() <span class="sc">|&gt;</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes_pcp</span>()) <span class="sc">+</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp_axes</span>() <span class="sc">+</span></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp</span>(<span class="fu">aes</span>(<span class="at">colour =</span> species), <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">overplot =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp_labels</span>() <span class="sc">+</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">peng.colors</span>()) <span class="sc">+</span></span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y =</span> <span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(), <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
</details>
</div>
<!-- CHEATING HERE because ggcpc plots take so long & caching causes problems -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-ggpcp1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggpcp1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/fig-peng-ggpcp1-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggpcp1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.39: Parallel coordinates plot of penguin size variables, together with sex and species.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Rearranging the order of variables and the ordering of factor levels can make a difference in what we can see in such plots. For a simple example (following <span class="citation" data-cites="VanderPlas2023">VanderPlas et al. (<a href="#ref-VanderPlas2023" role="doc-biblioref">2023</a>)</span>), we reorder the levels of species and islands to make it clearer which species occur on each island.</p>
<div class="cell" data-layout-align="center">
<details open class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb59" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>peng1 <span class="ot">&lt;-</span> peng <span class="sc">|&gt;</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">species =</span> <span class="fu">factor</span>(species, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;Chinstrap&quot;</span>, <span class="st">&quot;Adelie&quot;</span>, <span class="st">&quot;Gentoo&quot;</span>))) <span class="sc">|&gt;</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">island =</span> <span class="fu">factor</span>(island, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;Dream&quot;</span>, <span class="st">&quot;Torgersen&quot;</span>, <span class="st">&quot;Biscoe&quot;</span>)))</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>peng1 <span class="sc">|&gt;</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_select</span>(species, island, bill_length<span class="sc">:</span>body_mass) <span class="sc">|&gt;</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_scale</span>() <span class="sc">|&gt;</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pcp_arrange</span>(<span class="at">method =</span> <span class="st">&quot;from-left&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes_pcp</span>()) <span class="sc">+</span></span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp_axes</span>() <span class="sc">+</span></span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp</span>(<span class="fu">aes</span>(<span class="at">colour =</span> species), <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">overplot =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp_boxes</span>(<span class="at">fill =</span> <span class="st">&quot;white&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pcp_labels</span>() <span class="sc">+</span></span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">peng.colors</span>()[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">3</span>)]) <span class="sc">+</span></span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y =</span> <span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
</details>
</div>
<!-- cHEATING HERE because ggcpc plots take so long -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-ggpcp2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-ggpcp2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/fig-peng-ggpcp2-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-ggpcp2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.40: Parallel coordinates plot of penguin size variables, with the levels of species and island reordered.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The order of variables in this plot emphasizes the relation between penguin species and the island where they were observed and then shows the values of the quantitative body size measurements.</p>
<p>More generally, quantitative variables can, and probably should, be ordered to place the most highly correlated variables adjacently to minimize the degree of crossing lines from one variable to the next <span class="citation" data-cites="MartiLaguna2003">(<a href="#ref-MartiLaguna2003" role="doc-biblioref">Martí &amp; Laguna, 2003</a>)</span>. When variables are highly <em>negatively</em> correlated (such as <code>bill_depth</code> and <code>flipper_length</code> here), crossings can be reduced simply by reversing the scale of one of the variables, e.g., by plotting <code>-bill_depth</code>. </p>
</section>
<section id="sec-animated-tours" class="level2" data-number="4.11">
<h2 data-number="4.11"><span class="header-section-number">4.11</span> Animated tours</h2>
<p>In the mid 17<span class="math inline">\(^{th}\)</span> to early 19<span class="math inline">\(^{th}\)</span>-century the <strong>Grand Tour</strong> became a coming-of-age custom for young Europeans (mainly British nobility and landed gentry) of sufficient rank and means to undertake a journey to the principal sites of Europe (Paris, Geneva, Rome, Athens, …) to complete their education by learning something of the cultural legacies in history, art, and music from antiquity to the Renaissance. Thereby, they could gain a wider appreciation of history and be prepared to play a role in polite society or in their chosen endeavors.</p>
<p>Travels in high-dimensional data space might be less thrilling than a journey from London through Paris and Millan to Rome. Yet, in both cases it is useful to think of the path taken, and what might be seen along the way. But there are different kinds of tours. We might simply take a meandering tour, exploring all the way, or want to plan a tour to see the most interesting sites in travel or have a tour guided by an expert. Similarly in data space, we might travel randomly to see what we can find or be guided to find interesting features such as clusters, outliers or non-linear relations in data. </p>
<p>Following the demonstration in PRIM-9 (<a href="#sec-discoveries" class="quarto-xref"><span class="quarto-unresolved-ref">sec-discoveries</span></a>) of exploring multidimensional data space by rotation <span class="citation" data-cites="Asimov:85">Asimov (<a href="#ref-Asimov:85" role="doc-biblioref">1985</a>)</span> developed the idea of the <em>grand tour</em>, a computer method for viewing multivariate statistical data via orthogonal projections onto an animated sequence of low-dimensional subspaces, like a movie. In contrast to a scatterplot matrix which shows a static view of a data cloud projected onto all pairwise variable axes, a statistical tour is like the view of an eye moving smoothly in high-dimensional space, capturing what it sees from a given location onto the 2-d plane of the computer screen.</p>
<p>More generally, statistical tours are a type of dynamic projections onto orthogonal axes (called a <em>basis</em>) that embed data in a <em>smaller</em> <span class="math inline">\(p\)</span>-dimensional space into a <span class="math inline">\(d\)</span>-dimensional viewing subspace with <span class="math inline">\(d &lt; p\)</span>. Typically, <span class="math inline">\(d=2\)</span>, where the result can displayed as scatterplots, together with vectors representing the projections of the data variables in this space. But the projected data can be rendered in 1D as densities or histograms, or in more dimensions (<span class="math inline">\(d &gt; 2\)</span>) as glyphs, or even as parallel coordinate plots. The essential idea is that we can define, and animate, a <em>tour path</em> as a smooth sequence of such projections over small changes to the projection basis, which gives the orientation of the data in the viewing space.</p>
<section id="sec-projections" class="level3" data-number="4.11.1">
<h3 data-number="4.11.1"><span class="header-section-number">4.11.1</span> Projections</h3>
<p>The idea of a projection is fundamental to touring methods and other visualizations of high-D data, so it is useful to understand what a projection is. Quite simply, you can think of a projection as the shadow of an object or cloud of points.</p>
<p>This is nicely illustrated by the cover image (<a href="#fig-cover-GBE2" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-cover-GBE2</span></a>) used for Douglas Hofstadter’s <span class="citation" data-cites="Hofstadter1979">(<a href="#ref-Hofstadter1979" role="doc-biblioref">1979</a>)</span> <em>Gödel, Bach and Escher</em> which shows two 3D solid shapes illuminated by light sources so their shadows form the letters G, B and E projected onto the planes formed by pairs of the three coordinate axes. The set of three 2D views is essentially the same that we see in a scatterplot matrix, where a 3D dataset is portrayed by the set of <em>shadows</em> of the points on planes formed by pairs of coordinate axes.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-cover-GBE2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-cover-GBE2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/Cover-GBE.png" class="img-fluid quarto-figure quarto-figure-center" style="width:40.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-cover-GBE2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.41: The cover image from <span class="citation" data-cites="Hofstadter1979">Hofstadter (<a href="#ref-Hofstadter1979" role="doc-biblioref">1979</a>)</span> illustrates how projections are shadows of an object cast by a light from a given direction.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the simplest case, a data point <span class="math inline">\(\mathbf{x} = (x_1, x_2)\)</span> in two dimensions can be represented geometrically as a vector from the origin as shown in <a href="#fig-projection" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-projection</span></a>. This point can be projected on any one-dimensional axis <span class="math inline">\(\mathbf{p}\)</span> by dropping a line perpendicular to <span class="math inline">\(\mathbf{p}\)</span>, which is the idea of a shadow. Mathematically, this is calculated as the product <span class="math inline">\(\mathbf{x}^\mathsf{T} \mathbf{p} = x_1 p_1 + x_2 p_2\)</span> and suitably normalized to give the correct length. …</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-projection" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-projection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/projection.png" class="img-fluid quarto-figure quarto-figure-center" style="width:40.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-projection-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.42: Projection of a point <strong>x</strong> onto a direction or axis <strong>p</strong>.
</figcaption>
</figure>
</div>
</div>
</div>
<p>More generally, a projection of an <span class="math inline">\((n \times p)\)</span> data matrix <span class="math inline">\(\mathbf{X}\)</span> representing <span class="math inline">\(n\)</span> observations in <span class="math inline">\(p\)</span> dimensions onto a <span class="math inline">\(d\)</span>-dimensional viewing space <span class="math inline">\(\mathbf{Y}_{n \times d}\)</span> is represented by a <span class="math inline">\(p \times d\)</span> projection matrix <span class="math inline">\(\mathbf{P}\)</span> as <span class="math inline">\(\mathbf{Y} = \mathbf{X} \mathbf{P}\)</span>, where the columns of <span class="math inline">\(\mathbf{P}\)</span> are orthogonal and of unit length,i.e., <span class="math inline">\(\mathbf{P}^\mathsf{T} \mathbf{P} = \mathbf{I}_{(d \times d)}\)</span>.</p>
<p>For example, to project a data matrix <span class="math inline">\(\mathbf{X}\)</span> in three dimensions onto a 2D plane, we would multiply it by a <span class="math inline">\((3 \times 2)\)</span> orthonormal matrix <span class="math inline">\(\mathbf{P}\)</span>. The matrix <span class="math inline">\(\mathbf{P}_1\)</span> below simply selects the first two columns of <span class="math inline">\(\mathbf{X}\)</span>.<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a></p>
<p><span class="math display">\[
\mathbf{X} =
\begin{bmatrix}
    0 &amp; 0 &amp; 0 \\
    0 &amp; 0 &amp; 10 \\
    0 &amp; 10 &amp; 0 \\
    0 &amp; 10 &amp; 10 \\
    10 &amp; 0 &amp; 0 \\
    10 &amp; 0 &amp; 10 \\
    10 &amp; 10 &amp; 0 \\
    10 &amp; 10 &amp; 10 \\
\end{bmatrix}_{8 \times 3}
;\;
\mathbf{P_1} =
\begin{bmatrix}
    1 &amp; 0 \\
    0 &amp; 1 \\
    0 &amp; 0 \\
\end{bmatrix}_{3 \times 2}
\;\Rightarrow\quad
\mathbf{Y} = \mathbf{X} \; \mathbf{P_1} =
\begin{bmatrix}
    0 &amp; 0 \\
    0 &amp; 0 \\
    0 &amp; 10 \\
    0 &amp; 10 \\
    10 &amp; 0 \\
    10 &amp; 0 \\
    10 &amp; 10 \\
    10 &amp; 10 \\
\end{bmatrix}_{8 \times 2}
\]</span> An oblique projection using all three dimensions is given by <span class="math inline">\(\mathbf{P_2}\)</span> below. This produces a new 2D view in <span class="math inline">\(\mathbf{Y}\)</span>:</p>
<p><span class="math display">\[
\mathbf{P_2} =
\begin{bmatrix}
    0.71 &amp; -0.42 \\
    0.71 &amp; 0.42 \\
    0 &amp; 0.84 \\
\end{bmatrix}_{3 \times 2}
\quad\Rightarrow\quad
\mathbf{Y} = \mathbf{X} \; \mathbf{P_2} =
\begin{bmatrix}
    0 &amp; 0 \\
    0 &amp; 8.4 \\
    7.1 &amp; 4.2 \\
    7.1 &amp; 12.6 \\
    7.1 &amp; -4.2 \\
    7.1 &amp; 4.2 \\
    14.2 &amp; 0 \\
    14.2 &amp; 8.4 \\
\end{bmatrix}
\]</span></p>
<p>The columns in <span class="math inline">\(\mathbf{Y}\)</span> are simply the linear combinations of those of <span class="math inline">\(\mathbf{X}\)</span> using the weights in each column of <span class="math inline">\(\mathbf{P_2}\)</span></p>
<p><span class="math display">\[\begin{aligned}
\mathbf{y}_1 &amp; = &amp; 0.71 \mathbf{x}_1 + 0.71 \mathbf{x}_2 + 0 \mathbf{x}_3\\
\mathbf{y}_2 &amp; = &amp; -0.42 \mathbf{x}_1 + 0.42 \mathbf{x}_2 + 0.84 \mathbf{x}_3 \\
\end{aligned}\]</span></p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb60" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>vals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">x1 =</span> vals, <span class="at">x2=</span>vals, <span class="at">x3=</span>vals) <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co"># project on just x1, x2 plane</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>P1 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(<span class="fu">diag</span>(<span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>Y1 <span class="ot">&lt;-</span> X <span class="sc">%*%</span> P1</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="co"># oblique projection</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>P2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.71</span>, <span class="fl">0.71</span>, <span class="dv">0</span>, <span class="sc">-</span><span class="fl">0.42</span>, .<span class="dv">42</span>, <span class="fl">0.84</span>), <span class="at">ncol=</span><span class="dv">2</span>)</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>Y2 <span class="ot">&lt;-</span> X <span class="sc">%*%</span> P2</span></code></pre></div>
</details>
</div>
<!-- fig.code: R/projections.R -->
<p>In this example, the matrix <span class="math inline">\(\mathbf{X}\)</span> consists of 8 points at the vertices of a cube of size 10, as shown in <a href="#fig-proj-combined" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-combined</span></a> (a). The projections <span class="math inline">\(\mathbf{Y}_1 = \mathbf{P}_1 \mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}_2 = \mathbf{P}_2 \mathbf{X}\)</span> are shown in panels (b) and (c). To make it easier to relate the points in different views, shapes and colors are assigned so that each point has a unique combination of these attributes.<!-- [^pch] --></p>
<!--
The unicode characters cause problems in LaTeX with `pdflatex`; they seem to work with `xelatex`.
[^pch]: Plot shapes given by `pch = 15:18` correspond to: 
■ (&#25F9;) filled square (15);
● (&#25CF;) filled circle (16); 
▲ (&#25B2;) filled triangle point-up (17);
◆ (&#25C8;) filled diamond (18).
-->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb61" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>pch <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">15</span><span class="sc">:</span><span class="dv">18</span>, <span class="at">times =</span> <span class="dv">2</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;darkgreen&quot;</span>, <span class="st">&quot;brown&quot;</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">rep</span>(colors, <span class="at">each =</span> <span class="dv">2</span>)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(X, pch, col)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   x1 x2 x3 pch       col</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 1  0  0  0  15       red</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 10  0  0  16       red</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 3  0 10  0  17      blue</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 4 10 10  0  18      blue</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a><span class="co"># 5  0  0 10  15 darkgreen</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a><span class="co"># 6 10  0 10  16 darkgreen</span></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a><span class="co"># 7  0 10 10  17     brown</span></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 8 10 10 10  18     brown</span></span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-proj-combined" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-proj-combined-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/proj-combined.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-proj-combined-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.43: <strong>Projection example</strong>: (a) The 8 points in <strong>X</strong> form a cube of size 10; (b) the projection by <strong>P1</strong> is the view ignoring <strong>x3</strong> (two points coincide at each vertex); (c) the projection by <strong>P2</strong> is an oblique view.
</figcaption>
</figure>
</div>
</div>
</div>
<p>But, if we are traveling in the projection space of <span class="math inline">\(\mathbf{Y}\)</span>, we need some signposts to tell us how the new dimensions relate to those of <span class="math inline">\(\mathbf{X}\)</span>. The answer is provided simply by plotting the rows of <span class="math inline">\(\mathbf{P}\)</span> as vectors, as shown in <a href="#fig-proj-vectors" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-vectors</span></a>. In these plots, each row of <span class="math inline">\(\mathbf{P}_1\)</span> and <span class="math inline">\(\mathbf{P}_2\)</span> appears as a vector from the origin. It’s direction shows the contribution each of <span class="math inline">\(\mathbf{x}_1, \mathbf{x}_2, \mathbf{x}_3\)</span> make to the new coordinates <span class="math inline">\(\mathbf{y}_1\)</span> and <span class="math inline">\(\mathbf{y}_2\)</span>.</p>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode" id="cb62" data-code-line-numbers=""><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(matlib)     <span class="co"># circle(), vectors()</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>xlim <span class="ot">&lt;-</span> ylim <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.1</span>, <span class="fl">1.1</span>)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>axes.x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="cn">NA</span>, <span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>axes.y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="cn">NA</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>labs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">expression</span>(x[<span class="dv">1</span>]), <span class="fu">expression</span>(x[<span class="dv">2</span>]), <span class="fu">expression</span>(x[<span class="dv">3</span>]))</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xlim, ylim, <span class="at">type =</span> <span class="st">&quot;n&quot;</span>, <span class="at">asp=</span><span class="dv">1</span>,</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">expression</span>(y[<span class="dv">1</span>]), <span class="at">ylab =</span> <span class="fu">expression</span>(y[<span class="dv">2</span>]),</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex.lab =</span> <span class="fl">1.8</span>)</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a><span class="fu">circle</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;skyblue&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>))</span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(axes.x, axes.y, <span class="at">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a><span class="fu">vectors</span>(P1, <span class="at">labels =</span> labs, <span class="at">cex.lab =</span> <span class="fl">1.8</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">pos.lab =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xlim, ylim, <span class="at">type =</span> <span class="st">&quot;n&quot;</span>, <span class="at">asp=</span><span class="dv">1</span>,</span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">expression</span>(y[<span class="dv">1</span>]), <span class="at">ylab =</span> <span class="fu">expression</span>(y[<span class="dv">2</span>]),</span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex.lab =</span> <span class="fl">1.8</span>)</span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a><span class="fu">circle</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;skyblue&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>))</span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(axes.x, axes.y, <span class="at">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a><span class="fu">vectors</span>(P2, <span class="at">labels =</span> labs, <span class="at">cex.lab =</span> <span class="fl">1.8</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span></code></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout-ncol="2" data-layout-align="center">

</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-proj-vectors" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-proj-vectors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/proj-vectors-labels.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-proj-vectors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.44: <strong>Variable vectors</strong>: Data variables viewed as vectors in the space of their projections. The angles of the <strong>x</strong> vectors with respect to the <strong>y</strong> coordinate axes show their relative contributions to each. The lengths of the <strong>x</strong> vectors show the relative degree to which they are represented in the space of <strong>y</strong>s. Left: the <strong>P1</strong> projection; right: the <strong>P2</strong> projection.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In <span class="math inline">\(\mathbf{P}_1\)</span>, the projected variable <span class="math inline">\(\mathbf{y}_1\)</span> is related only to <span class="math inline">\(\mathbf{x}_1\)</span>, while <span class="math inline">\(\mathbf{y}_2\)</span> is related only to <span class="math inline">\(\mathbf{x}_2\)</span>. <span class="math inline">\(\mathbf{x}_3\)</span> makes no contribution, and appears at the origin. However in the projection given by <span class="math inline">\(\mathbf{P}_2\)</span>, <span class="math inline">\(\mathbf{x}_1\)</span> and <span class="math inline">\(\mathbf{x}_2\)</span> make the same contribution to <span class="math inline">\(\mathbf{y}_1\)</span>, while <span class="math inline">\(\mathbf{x}_3\)</span> has no contribution to that horizontal axis. The vertical axis, <span class="math inline">\(\mathbf{y}_2\)</span> here is completely aligned with <span class="math inline">\(\mathbf{x}_3\)</span>; <span class="math inline">\(\mathbf{x}_1\)</span> and <span class="math inline">\(\mathbf{x}_2\)</span> have vertical components that are half of that for <span class="math inline">\(\mathbf{x}_3\)</span> in absolute value.</p>
<p>` #### Vector lengths In <a href="#fig-proj-vectors" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-vectors</span></a>, the <strong>lengths</strong> of the <span class="math inline">\(\mathbf{x}\)</span> vectors reflect the relative degree to which each variable is represented in the space of the projection, and this is important for interpretation. For the <span class="math inline">\(\mathbf{P}_1\)</span> projection, <span class="math inline">\(\mathbf{x}_3\)</span> is of length 0, while <span class="math inline">\(\mathbf{x}_1\)</span> and <span class="math inline">\(\mathbf{x}_2\)</span> fill the unit circle. In the projection given by <span class="math inline">\(\mathbf{P}_2\)</span>, all three <span class="math inline">\(\mathbf{x}\)</span> are approximately the same length.</p>
<p>In algebra, the length of a vector <span class="math inline">\(\mathbf{x}\)</span> is <span class="math inline">\(||\mathbf{x}|| = (\mathbf{x}^\mathsf{T} \mathbf{x})^{1/2} = \sqrt{\Sigma x_i^2}\)</span>, the Euclidean distance of the tip of the vector from the origin. In R, we calculate the lengths of row vectors in a projection matrix by transposing and using <code>matlib::len()</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb63" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>P1 <span class="sc">|&gt;</span> <span class="fu">t</span>() <span class="sc">|&gt;</span> matlib<span class="sc">::</span><span class="fu">len</span>()</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 1 1 0</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>P2 <span class="sc">|&gt;</span> <span class="fu">t</span>() <span class="sc">|&gt;</span> matlib<span class="sc">::</span><span class="fu">len</span>()</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 0.825 0.825 0.840</span></span></code></pre></div>
</div>
<section id="joint-views" class="level4" data-number="4.11.1.1">
<h4 data-number="4.11.1.1"><span class="header-section-number">4.11.1.1</span> Joint-views</h4>
<p>To interpret such projections, we want to see <strong>both</strong> the projected data and the signposts that tell us where we are in relation to the original variables. To do this, we can overlay the variable vectors represented by the rows of the projection matrix <span class="math inline">\(\mathbf{P}\)</span> onto plots like <a href="#fig-proj-combined" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-combined</span></a> (b) and <a href="#fig-proj-combined" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-combined</span></a> (c) to see how the axes in a projection relate to those in the data. To place these together on the same plot, we can either center the columns of <span class="math inline">\(\mathbf{Y}\)</span> at their means or shift the the columns of <span class="math inline">\(\mathbf{P}\)</span> to <code>colMeans(Y)</code>. It is only the directions of the vectors that matters, so we are free to scale their lengths by any convenient factor.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode" id="cb64" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>Y2s <span class="ot">&lt;-</span> <span class="fu">scale</span>(Y2, <span class="at">scale=</span><span class="cn">FALSE</span>)       <span class="co"># center Y2</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Y2s, <span class="at">cex =</span> <span class="dv">3</span>, </span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">asp =</span> <span class="dv">1</span>,</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> pch, <span class="at">col =</span> col,</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">expression</span>(y[<span class="dv">1</span>]), <span class="at">ylab =</span> <span class="fu">expression</span>(y[<span class="dv">2</span>]),</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>), <span class="at">cex.lab =</span> <span class="fl">1.8</span>)</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="dv">7</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>vecs <span class="ot">&lt;-</span> (r<span class="sc">*</span><span class="fu">diag</span>(<span class="dv">3</span>) <span class="sc">%*%</span> P2)</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="fu">vectors</span>(vecs, <span class="at">labels =</span> labs, <span class="at">cex.lab =</span> <span class="fl">1.8</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="fu">vectors</span>(<span class="sc">-</span>vecs, <span class="at">labels =</span> <span class="cn">NULL</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">angle =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">&quot;gray&quot;</span>)</span></code></pre></div>
</details>
</div>
<p>The plot in <a href="#fig-proj-P2-vec" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-P2-vec</span></a> illustrates this, centering <span class="math inline">\(\mathbf{Y}\)</span>, and multiplying the vectors in <span class="math inline">\(\mathbf{P}\)</span> by 7. To check your understanding, try to see if you can relate what is shown in this plot to the 3D plot in <a href="#fig-proj-combined" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-proj-combined</span></a> (a).</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-proj-P2-vec" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-proj-P2-vec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/proj-P2-vec.png" class="img-fluid quarto-figure quarto-figure-center" style="width:50.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-proj-P2-vec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.45: The <strong>P2</strong> projection of the data showing vectors for the original variables in the space of <strong>Y</strong>.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The idea of viewing low-dimensional projections of data together with vectors representing the contributions of the original variables to the dimensions shown in a display is also the basis of <strong>biplot</strong> techniques (<a href="#sec-biplot" class="quarto-xref"><span class="quarto-unresolved-ref">sec-biplot</span></a>) we will use in relation to principal components analysis.</p>
</section>
</section>
<section id="touring-methods" class="level3" data-number="4.11.2">
<h3 data-number="4.11.2"><span class="header-section-number">4.11.2</span> Touring methods</h3>
<p>The trick of statistical touring methods is to generate a smooth sequence of interpolated projections <span class="math inline">\(\mathbf{P}_{(t)}\)</span> indexed by time <span class="math inline">\(t\)</span>, <span class="math inline">\(\mathbf{P}_{(1)}, \mathbf{P}_{(2)}, \mathbf{P}_{(3)}, \dots, \mathbf{P}_{(T)}\)</span>. This gives a path of views <span class="math inline">\(\mathbf{Y}_{(t)} = \mathbf{X} \mathbf{P}_{(t)}\)</span>, that can be animated in successive frames, as shown schematically in <a href="#fig-peng-tourr-diagram" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-tourr-diagram</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-tourr-diagram" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tourr-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/peng-tourr-diagram.png" class="img-fluid quarto-figure quarto-figure-center" style="width:80.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tourr-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.46: <strong>Interpolations</strong>: Illustration of a grand tour of interpolations of projection planes showing 2D scatterplots of the Penguin dataset. The seqeunce of views moves smoothly from an initial frame <strong>P(1)</strong> to a final frame <strong>P(T)</strong> where the penguin species are widely separated.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Asimov’s <span class="citation" data-cites="Asimov:85">(<a href="#ref-Asimov:85" role="doc-biblioref">1985</a>)</span> original idea of the grand tour was that of a random path, picking orthogonal projections <span class="math inline">\(\mathbf{P}_{(i)}\)</span> at random. Given enough time, the grand tour gives a space-filling path and would eventually show every possible projection of the data. But it does so smoothly, by interpolating from one projection to the next. In the travel analogy, the path by road from London to Paris might go smoothly through Kent to Dover, thence via Amiens and Beauvais before reaching Paris. By air, the tour would follow a smoother <em>geodesic</em> path, and this is what the grand tour does. The sense in watching an animation of a statistical grand tour is that of continuous motion. The grand tour algorithm is described in detail by <span class="citation" data-cites="Buja-etal-2005">Buja et al. (<a href="#ref-Buja-etal-2005" role="doc-biblioref">2005</a>)</span> and <span class="citation" data-cites="Cook-etal-2008">Cook et al. (<a href="#ref-Cook-etal-2008" role="doc-biblioref">2008</a>)</span>. </p>
<!-- Projection pursuit @Cook-etal-1995 -> guided tour -->
<section id="guided-tours" class="level4" data-number="4.11.2.1">
<h4 data-number="4.11.2.1"><span class="header-section-number">4.11.2.1</span> Guided tours</h4>
<p>The next big idea was that rather than traveling randomly in projection space one could take a <em>guided tour</em>, following a path that leads to “interesting projections”, such as those that reveal clusters, gaps in data space or outliers. This idea, called <em>projection pursuit</em> <span class="citation" data-cites="Cook-etal-1995">(<a href="#ref-Cook-etal-1995" role="doc-biblioref">Cook et al., 1995</a>)</span>, works by defining a measure of interestingness of a data projection. In a guided tour, the next projection is chosen to increase that index, so over time the projection moves toward one that is maximizes that index.</p>
<p>In the time since <span class="citation" data-cites="Asimov:85">Asimov (<a href="#ref-Asimov:85" role="doc-biblioref">1985</a>)</span>, there have been many implementations of touring visualization methods. XGobi <span class="citation" data-cites="Swayne-etal-1998">(<a href="#ref-Swayne-etal-1998" role="doc-biblioref">Swayne et al., 1998</a>)</span> for X-Windows displays on Linux systems provided a test-bed for dynamic, interactive graphic methods; it’s successor, GGobi <span class="citation" data-cites="Swayne-etal-2003 CookSwayne:2007">(<a href="#ref-CookSwayne:2007" role="doc-biblioref">Cook &amp; Swayne, 2007</a>; <a href="#ref-Swayne-etal-2003" role="doc-biblioref">Swayne et al., 2003</a>)</span> extended the range of touring methods to include a wider variety of projection pursuit indices.</p>
</section>
<section id="tourr-package" class="level4" data-number="4.11.2.2">
<h4 data-number="4.11.2.2"><span class="header-section-number">4.11.2.2</span> <code>tourr</code> package</h4>
<p>The current state of art is best captured in the <span style="color: brown;"><strong>tourr</strong></span> package for R <span class="citation" data-cites="Wickham-etal-2011 R-tourr">(<a href="#ref-Wickham-etal-2011" role="doc-biblioref">Wickham et al., 2011</a>; <a href="#ref-R-tourr" role="doc-biblioref">Wickham &amp; Cook, 2025</a>)</span>. It defines a tour to consist of three components:</p>
<ul>
<li><strong>data</strong>: An <span class="math inline">\((n \times p)\)</span> numerical data matrix to be viewed.</li>
<li><strong>path</strong>: A tour path function that produces a smoothed sequence of projection matrices <span class="math inline">\(\mathbf{P}_{(p \times d)}\)</span> in <span class="math inline">\(d\)</span>. dimensions, for example <code>grand_tour(d = 2)</code> or <code>guided_tour(index = holes)</code>.</li>
<li><strong>display</strong>: A function that renders the projected data, for example <code>display_xy()</code> for a scatterplot, <code>display_depth()</code> for a 3D plot with simulated depth, or <code>display_pcp()</code> for a parallel coordinates plots</li>
</ul>
<p>This very nicely separates the aspects of a tour, and allows one to think of and define new tour path methods and display methods. The package defines two general tour functions: <code>animate()</code> produces a real-time animation on a display device and <code>render()</code> saves image frames to disk, such as a <code>.gif</code> file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb65" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(data, tour_path, display_method)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">render</span>(data, tour_path, display_method)</span></code></pre></div>
</div>
<p>The <span style="color: brown;"><strong>tourr</strong></span> package provides a wide range of tour path methods and display methods:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb66" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tour path methods</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">&quot;_tour$&quot;</span>, <span class="fu">lsf.str</span>(<span class="st">&quot;package:tourr&quot;</span>), <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;dependence_tour&quot;     &quot;frozen_guided_tour&quot; </span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  [3] &quot;frozen_tour&quot;         &quot;grand_tour&quot;         </span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  [5] &quot;guided_anomaly_tour&quot; &quot;guided_section_tour&quot;</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  [7] &quot;guided_tour&quot;         &quot;little_tour&quot;        </span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;local_tour&quot;          &quot;new_tour&quot;           </span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a><span class="co"># [11] &quot;planned_tour&quot;        &quot;planned2_tour&quot;      </span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;radial_tour&quot;</span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="co"># display methods</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">&quot;display_&quot;</span>, <span class="fu">lsf.str</span>(<span class="st">&quot;package:tourr&quot;</span>), <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;display_andrews&quot;   &quot;display_density2d&quot; &quot;display_depth&quot;    </span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a><span class="co">#  [4] &quot;display_dist&quot;      &quot;display_faces&quot;     &quot;display_groupxy&quot;  </span></span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a><span class="co">#  [7] &quot;display_idx&quot;       &quot;display_image&quot;     &quot;display_pca&quot;      </span></span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a><span class="co"># [10] &quot;display_pcp&quot;       &quot;display_sage&quot;      &quot;display_scatmat&quot;  </span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;display_slice&quot;     &quot;display_stars&quot;     &quot;display_stereo&quot;   </span></span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a><span class="co"># [16] &quot;display_trails&quot;    &quot;display_xy&quot;</span></span></code></pre></div>
</div>
<p>Tour path methods take a variety of optional arguments to specify the detailed behavior of the method. For example, most allow you to specify the number of dimension (<code>d =</code>) of the projections. The <code>guided_tour()</code> is of particular interest here.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb67" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">args</span>(guided_tour)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="co"># function (index_f, d = 2, cooling = 0.99, max.tries = 25, max.i = Inf, </span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="co">#     search_f = search_geodesic, n_jellies = 30, n_sample = 100, </span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co">#     alpha = 0.5, ...) </span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co"># NULL</span></span></code></pre></div>
</div>
<p>In this, <code>index_f</code> specifies a function that the method tries to optimize on its path and package defines four indices:</p>
<ul>
<li>Holes (<code>holes()</code>): This is sensitive to projections with separated clusters of points, with few points near the origin</li>
<li>Central mass (<code>cmass()</code>): Sensitive to projections with lots of points in the center, but perhaps with some outliers</li>
<li>Linear discriminant analysis (<code>lda_pp()</code>): For data with a grouping factor, optimizes a measure of separation of the group means as in MANOVA or linear discriminant analysis.</li>
<li>PDA analysis (<code>pda_pp()</code>): A penalized version of <code>lda_pp()</code> for cases of large <span class="math inline">\(p\)</span> relative to sample size <span class="math inline">\(n\)</span> <span class="citation" data-cites="LeeCook-2009">(<a href="#ref-LeeCook-2009" role="doc-biblioref">E.-K. Lee &amp; Cook, 2009</a>)</span>.</li>
</ul>
<p>In addition, there is now a <code>guided_anomaly_tour()</code> that looks for the best projection of observations that are outside the data ellipsoid, finding a view showing observations with large Mahalanobis distances from the centroid.</p>
</section>
<section id="penguin-tours" class="level4 unnumbered">
<h4 class="unnumbered">Penguin tours</h4>
<p>Penguins are a traveling species. They make yearly travels inland to breeding sites in early spring, repeating the patterns of their ancestors. Near the beginning of summer, adult penguins and their chicks return to the sea and spend the rest of the summer feeding there, but they have different strategies to survive the polar winter <span class="citation" data-cites="Black-etal-2018">(<a href="#ref-Black-etal-2018" role="doc-biblioref">Black et al., 2018</a>)</span>. The data scientists among them might wonder about the relations among among their cousins of different species on various islands and take a data-centric tour of their 4D measurements of body size with trusty R tools.</p>
<!-- Cite: Black et al Time-lapse imagery of Ade´lie penguins, https://doi.org/10.1371/journal.pone.0193532 -->
<p>For example, using the Penguins dataset, the following calls produce grand tours in 2, 3, and 4 dimensions. The 2D tour is displayed as a scatterplot, the 3D tour using simulated depth as shown by variation in point size and transparency, and the 4D tour is shown using a parallel coordinate plot.</p>
<!-- **TODO**: ~~This doesn't render well in @fig-peng-tour-demo ~~ -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb68" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(peng, <span class="at">package =</span> <span class="st">&quot;heplots&quot;</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>peng_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(peng[,<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(peng_scaled) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;BL&quot;</span>, <span class="st">&quot;BD&quot;</span>, <span class="st">&quot;FL&quot;</span>, <span class="st">&quot;BM&quot;</span>)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(peng_scaled, <span class="fu">grand_tour</span>(<span class="at">d =</span> <span class="dv">2</span>), <span class="fu">display_xy</span>())</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(peng_scaled, <span class="fu">grand_tour</span>(<span class="at">d =</span> <span class="dv">3</span>), <span class="fu">display_depth</span>())</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(peng_scaled, <span class="fu">grand_tour</span>(<span class="at">d =</span> <span class="dv">4</span>), <span class="fu">display_pcp</span>())</span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-tour-demo" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tour-demo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-tour-demo-all.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tour-demo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.47: Snapshops of grand tours of the penguin dataset in 2, 3, and 4 dimensions using different <code>display_*()</code> methods.
</figcaption>
</figure>
</div>
</div>
</div>
<!-- 
#| layout-ncol: 3
#| out-width: "33%"
#| fig-subcap: 
#|  - "2D, scatterplot"
#|  - "3D, simulated depth"
#|  - "4D, parallel coordinates plot"
knitr::include_graphics("images/tours/peng-tour-demo1.png")
knitr::include_graphics("images/tours/peng-tour-demo2.png")
knitr::include_graphics("images/tours/peng-tour-demo3.png")
-->
<p>To illustrate, I’ll start with a grand tour designed to explore this 4D space of penguins. I’ll abbreviate the variables to two characters, “BL” = <code>bill_length</code>, “BD” = <code>bill_depth</code>, “FL” = <code>flipper_length</code>, and “BM” = <code>body_mass</code> and identify the penguin species using point shape (<code>pch</code>) and color (<code>col</code>).</p>
<p>As you watch this pay attention to the separation of the species and any other interesting features. What do you see?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb69" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(peng, <span class="at">package =</span> <span class="st">&quot;heplots&quot;</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>peng_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(peng[,<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(peng_scaled) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;BL&quot;</span>, <span class="st">&quot;BD&quot;</span>, <span class="st">&quot;FL&quot;</span>, <span class="st">&quot;BM&quot;</span>)</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>pch <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">17</span>)[peng<span class="sc">$</span>species] </span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>cex <span class="ot">=</span> <span class="fl">1.2</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(peng_scaled,</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">tour_path =</span> <span class="fu">grand_tour</span>(<span class="at">d=</span><span class="dv">2</span>),</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">display_xy</span>(<span class="at">col =</span> peng<span class="sc">$</span>species,</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">palette =</span> <span class="fu">peng.colors</span>(<span class="st">&quot;dark&quot;</span>),</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pch =</span> pch, <span class="at">cex =</span> cex,</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>                   <span class="at">axis.col =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a>                   <span class="at">axis.text.col =</span> <span class="st">&quot;black&quot;</span>, </span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>                   <span class="at">axis.lwd =</span> <span class="fl">1.5</span>))</span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-tour-grand" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tour-grand-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-tourr-grand.gif" class="img-fluid quarto-figure quarto-figure-center" style="width:75.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tour-grand-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.48: Animation of a grand tour of the Penguin data.
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-peng-tour-grand-frames" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-peng-tour-grand-frames</span></a> shows three frames from this movie. The first (a) is the initial frame that shows the projection in the plane of bill depth and bill length. The variable vectors indicate that bill length differentiates Adélie penguins from the others. In frame (b), the three species are widely separated, with bill depth distinguishing Gentoo from the others. In frame (c) the three species are largely mixed, but two points stand out as outliers, with exceptionally long bills compared to the rest.</p>
<div id="fig-peng-tour-grand-frames" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tour-grand-frames-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-peng-tour-grand-frames" style="flex-basis: 33.3%;justify-content: center;">
<div id="fig-peng-tour-grand-frames-1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig">
<div aria-describedby="fig-peng-tour-grand-frames-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-grand-tour1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" data-ref-parent="fig-peng-tour-grand-frames" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-peng-tour-grand-frames-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Initial frame
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-peng-tour-grand-frames" style="flex-basis: 33.3%;justify-content: center;">
<div id="fig-peng-tour-grand-frames-2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig">
<div aria-describedby="fig-peng-tour-grand-frames-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-grand-tour2.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" data-ref-parent="fig-peng-tour-grand-frames" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-peng-tour-grand-frames-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Clusters
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-peng-tour-grand-frames" style="flex-basis: 33.3%;justify-content: center;">
<div id="fig-peng-tour-grand-frames-3" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig">
<div aria-describedby="fig-peng-tour-grand-frames-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-grand-tour3.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" data-ref-parent="fig-peng-tour-grand-frames" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-peng-tour-grand-frames-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) Outliers
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tour-grand-frames-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.49: Three frames from the grand tour of the Penguin data. (a) The initial frame is the projection showing only BD and BL, where bill length conveniently separates Adélie from the other two species. (b) A frame that shows the three species more widely separated. (c) A frame that shows two outliers with very large bills.
</figcaption>
</figure>
</div>
<!-- ::: {#fig-peng-tour-grand-frames layout-ncol=3} -->
<!-- ![Initial frame](images/tours/peng-grand-tour1.png){#fig-peng-grand-tour1 width=140} -->
<!-- ![Clusters](images/tours/peng-grand-tour2.png){#fig-peng-grand-tour2 width=140} -->
<!-- ![Outliers](images/tours/peng-grand-tour3.png){#fig-peng-grand-tour3 width=140} -->
<!-- Three frames from the grand tour of the Penguin data. (a) The initial frame is the projection showing only -->
<!-- BD and BL, where bill length conveniently separates Adelie from the other two species. -->
<!-- (b) A frame that shows the three species more widely separated. (c) A frame that shows two outliers. -->
<!-- ::: -->
<p>Let’s take the penguins on a <em>guided tour</em>, trying to find views that show the greatest separations among their species; that is, a guided tour, optimizing the <code>lda_pp()</code> index of group separation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb70" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">animate</span>(peng_scaled, </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">guided_tour</span>(<span class="fu">lda_pp</span>(peng<span class="sc">$</span>species)),</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">display_xy</span>(<span class="at">col =</span> peng<span class="sc">$</span>species,</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">palette =</span> <span class="fu">peng.colors</span>(<span class="st">&quot;dark&quot;</span>),</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pch =</span> pch,</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cex =</span> cex)</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-peng-tour-lda" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tour-lda-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-tourr-lda.gif" class="img-fluid quarto-figure quarto-figure-center" style="width:75.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tour-lda-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.50: Animation of a guided tour of the Penguin data, using a tour criterion designed to find an optimal separation among the penguin species. The animation shows three loops of the sequence of projections and stops when the LDA criterion cannot be improved.
</figcaption>
</figure>
</div>
</div>
</div>
<!--
**TODO**: ~~I'm trying to balance what will/can be shown in the HTML version vs. the printed PDF.
Needs text here specifically for the PDF version.~~
-->
<div id="fig-peng-tour-guided" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-peng-tour-guided-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-peng-tour-guided" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-peng-tour-guided-1" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig">
<div aria-describedby="fig-peng-tour-guided-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-tour-lda-final.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" data-ref-parent="fig-peng-tour-guided" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-peng-tour-guided-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Optimizing <code>lda_pp()</code>
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-peng-tour-guided" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-peng-tour-guided-2" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig">
<div aria-describedby="fig-peng-tour-guided-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/tours/peng-tour-anomalies-final.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" data-ref-parent="fig-peng-tour-guided" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-peng-tour-guided-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Optimizing <code>anomaly_index()</code>
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peng-tour-guided-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.51: <strong>Guided tours</strong>: These figures show the final frame in the animations of guided tours designed to find the projection that optimize an index. (a) The <code>lda_pp()</code> criterion optimizes the separation of the means for species relative to within-group variation. (b) The <code>anomalies_index()</code> optimizes the average Mahalanobis distance of points from the centroid
</figcaption>
</figure>
</div>
<p>These examples are intended to highlight what is possible with dynamic graphics for exploring high-dimensional data visually. <span class="citation" data-cites="CookLaa-mulgar">Cook &amp; Laa (<a href="#ref-CookLaa-mulgar" role="doc-biblioref">2024</a>)</span> extend the discussion of these methods from <span class="citation" data-cites="CookSwayne:2007">Cook &amp; Swayne (<a href="#ref-CookSwayne:2007" role="doc-biblioref">2007</a>)</span> (which used <code>Ggobi</code>) to the <span style="color: brown;"><strong>tourr</strong></span> package. They illustrate dimension reduction, various cluster analysis methods, trees and random forests and some machine-learning techniques.</p>
<!-- **TODO**: Clean up these package references. I'm linking to the pkgdown sites, b/c I don't have them installed and can't easily get a `citation()`.  -->
<p>Ideally, we should be able interact with a tour,</p>
<ul>
<li>pausing when we see something interesting and saving the view for later analysis;</li>
<li>selecting or highlighting unusual points,</li>
<li>changing tour methods or variables displayed on the fly, and so forth.</li>
</ul>
<p>Some packages that provide these capabilities are: <a href="https://casperhart.github.io/detourr/index.html"><strong>detourr</strong></a><span class="citation" data-cites="R-detourr">(<a href="#ref-R-detourr" role="doc-biblioref">Hart &amp; Wang, 2022</a>)</span> <a href="https://logarithmic.net/langevitour/index.html"><strong>langevitour</strong></a><span class="citation" data-cites="R-langevitour Harrison2023">(<a href="#ref-R-langevitour" role="doc-biblioref">Harrison, 2023a</a>, <a href="#ref-Harrison2023" role="doc-biblioref">2023b</a>)</span> and <span style="color: brown;"><strong>liminal</strong></span> <span class="citation" data-cites="R-liminal">(<a href="#ref-R-liminal" role="doc-biblioref">S. Lee, 2025</a>)</span> <!-- [**liminal**](https://sa-lee.github.io/liminal/)[@R-liminal] --> The <span style="color: brown;"><strong>loon</strong></span> package <span class="citation" data-cites="R-loon">(<a href="#ref-R-loon" role="doc-biblioref">Waddell &amp; Oldford, 2025</a>)</span> is a general toolkit that enables highly interactive data visualization. It provides a <strong>loon.tour</strong> package <span class="citation" data-cites="R-loon-tour">(<a href="#ref-R-loon-tour" role="doc-biblioref">Xu &amp; Oldford, 2021</a>)</span> for using touring methods within the <code>loon</code> environment.</p>
</section>
</section>
</section>
<section id="sec-network" class="level2" data-number="4.12">
<h2 data-number="4.12"><span class="header-section-number">4.12</span> Network diagrams</h2>
<p>A major theme throughout this chapter has been to understand how to extend data visualization from simple bivariate scatterplots to increasingly more complex situations with larger datasets. With a moderate number of variables, techniques such as smoothing, summarizing with data ellipses and fitted curves, and visual thinning can be used to tame “big <span class="math inline">\(N\)</span>” datasets with thousands of cases.</p>
<p>However “big <span class="math inline">\(p\)</span>” datasets, with more than a moderate number (<span class="math inline">\(p\)</span>) of variables still remain a challenge. It is hard to see how the more advanced methods (corrgrams, parallel coordinate) described earlier could cope with <span class="math inline">\(p = 20, 50, 100, 500, \dots\)</span> variables. At some point, each of these begins to break down for the purpose of visualizing associations among many variables. We are forced to thin the information presented in graphs more and more as the number of variables increases.</p>
<p>It turns out that there is a way to increase the number of variables displayed dramatically, if we are mainly interested in the pairwise correlations for reasonably normally distributed data. A graphical <strong>network diagram</strong> portrays variables by <em>nodes</em> (vertices), connected by (weighted) <em>edges</em> whose properties reflect the strength of connections between pairs, such as a correlation. Such diagrams can reveal properties not readily seen by other means.</p>
<p>As an example consider <a href="#fig-big5-qgraph-rodrigues" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-big5-qgraph-rodrigues</span></a>, which portrays the correlations among 25 self-report items reflecting 5 factors (the “Big Five”) considered in personality psychology to represent the dominant aspects of all of personality. These factors are easily remembered by the acronum <strong>OCEAN</strong>: <strong>O</strong>penness, <strong>C</strong>onscientiousness, <strong>E</strong>xtraversion, <strong>A</strong>greeableness and <strong>N</strong>euroticism. The dataset, <code>bfi</code>, contains data from an online sample of <span class="math inline">\(n=2800\)</span> with 5 items for each scale.</p>
<p>In this figure (taken from <a href="https://bit.ly/3A6kvq5">Rodrigues (2021)</a>), the item nodes are labeled according to the OCEAN factor they are assumed to measure. For 25 items, there are <span class="math inline">\(25 \times 24 / 2 = 300\)</span> correlations, way too much to see. A clearer picture arises when we reduce the number of edges shown according to some criterion. Here, edges are drawn <em>only</em> between nodes where the correlation is considered important by a method (“glasso” = graphical LASSO) designed to make the graph optimally sparse.</p>
<!-- Actually, the psych::bfi data is:
25 personality self report items taken from the International Personality Item Pool (ipip.ori.org) were included as part of the Synthetic Aperture Personality Assessment (SAPA) web based personality assessment project. The data from 2800 subjects are included here as a demonstration set for scale construction, factor analysis, and Item Response Theory analysis. Three additional demographic variables (sex, education, and age) are also included.
-->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-big5-qgraph-rodrigues" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-big5-qgraph-rodrigues-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/big5-qgraph-rodrigues.png" class="img-fluid quarto-figure quarto-figure-center" style="width:90.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-big5-qgraph-rodrigues-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.52: Network diagram of the correlations among 25 items from a Big-Five personality scale, 5 items for each scale. The magnitude of a correlation is shown by the thickness and transparency of the edge between two item nodes. The sign of a correlation is shown by edge color and style: solid blue for positive and dashed red for negative. <em>Source</em>: <a href="https://bit.ly/3A6kvq5">Rodrigues (2021)</a>
</figcaption>
</figure>
</div>
</div>
</div>
<p>The edges shown in <a href="#fig-big5-qgraph-rodrigues" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-big5-qgraph-rodrigues</span></a> reflect the Pearson correlation between a given pair of items by the visual attributes of color and line style: magnitude is shown by both the thickness and transparency of the edge; the sign of the correlation is shown by color and line type: solid <span style="color: blue;">blue</span> for positive correlations and dashed <span style="color: red;">red</span> for negative ones.</p>
<p>According to some theories, the five personality factors should be largely non-overlapping, so there should not be many edges connecting items of one factor with those of another. Yet, there are quite a few cross-factor connections in <a href="#fig-big5-qgraph-rodrigues" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-big5-qgraph-rodrigues</span></a>, so perhaps the theory is wrong, or, more likely, the 25 items are not good representatives of these underlying dimensions. The network diagram shown here is a visual tool for thought and refinement. See <span class="citation" data-cites="Costantini2015">Costantini et al. (<a href="#ref-Costantini2015" role="doc-biblioref">2015</a>)</span> for a tutorial on network analysis of personality data in R.</p>
<p>Network diagrams stem from mathematical graph theory <span class="citation" data-cites="BondyMurty2008 West2001">(<a href="#ref-BondyMurty2008" role="doc-biblioref">Bondy &amp; Murty, 2008</a>; <a href="#ref-West2001" role="doc-biblioref">West, 2001</a>)</span> of the abstract properties of nodes and edges used to represent pairwise relationships. These can be used to model many types of relations and processes in physical, biological, social and other sciences, where such properties as connectedness, centrality, cliques of connected nodes and so forth provide a vocabulary used to understand and explain complex systems.</p>
<p>For one example, <span class="citation" data-cites="Grandjean2016">Grandjean (<a href="#ref-Grandjean2016" role="doc-biblioref">2016</a>)</span> used network analysis to study the connections among 2500 Twitter users (nodes) who identified as belonging to a “digital humanities” community from the relations (edges) of who follows whom. Grandjean also used these methods to study the relationships among <a href="http://www.martingrandjean.ch/network-visualization-shakespeare/">characters in Shakespeare’s tragedies</a> in terms of the characters (nodes) and edges representing how often they appeared in the same scene.</p>
<p>The wide applicability of these ideas has led to what is now called <em>network science</em> <span class="citation" data-cites="Barabasi2016network">(<a href="#ref-Barabasi2016network" role="doc-biblioref">Barab’asi, 2016</a>)</span> encompassing computer networks, biological networks, cognitive and semantic networks, and social networks. Recent developments in psychology led to a framework of <em>network psychometrics</em> <span class="citation" data-cites="IsvoranuEpskamp2022">(<a href="#ref-IsvoranuEpskamp2022" role="doc-biblioref">Isvoranu et al., 2022</a>)</span>, where, for example, symptoms of psychopathology (phobias, anxiety, substance abuse) can be conceptualized as an interconnected network of clusters and studied for possible causal relations <span class="citation" data-cites="Robinaugh2019">(<a href="#ref-Robinaugh2019" role="doc-biblioref">Robinaugh et al., 2019</a>)</span>.</p>
<p>Because a network diagram can potentially reflect hundreds of variables, various <a href="https://en.wikipedia.org/wiki/Graph_drawing">graph layout algorithms</a> have been developed to automatically position the nodes so as to generate aesthetically pleasing network visualizations that emphasize important structural properties, like clusters and central nodes, while minimizing visual clutter (many crossing lines) to promote understandability and usability.</p>
<p>There are quite a few R packages for constructing network diagrams, both static and dynamic / interactive, and these differ considerably in how the information required for a graph is structured as R objects, and the flexibility to produce attractive graphs. Among these, <span style="color: brown;"><strong>igraph</strong></span> <span class="citation" data-cites="R-igraph">(<a href="#ref-R-igraph" role="doc-biblioref">Csárdi et al., 2024</a>)</span> structures the data as a dataset of vertices and edges with properties</p>
<p>-&gt; packages: qgraph, …</p>
<section id="crime-data" class="level3" data-number="4.12.1">
<h3 data-number="4.12.1"><span class="header-section-number">4.12.1</span> Crime data</h3>
<p>For the present purposes, let’s see what network diagrams can tell us about the crime data analyzed earlier. Here, I first reorder the variables as in <a href="#fig-crime-corrplot-AOE" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-corrplot-AOE</span></a>. In the call to <code>qgraph()</code>, the argument <code>minimum = "sig"</code> says to show only the edges for significant correlations (at <span class="math inline">\(\alpha = 0.01\)</span> here). In <a href="#fig-crime-cor" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-cor</span></a>, the variable nodes are positioned around a circle (<code>layout = "circle"</code>), which is the default.</p>
<!-- fig.code: R/crime/crime-network.R -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb71" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qgraph)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">corrMatOrder</span>(crime.cor, <span class="at">order =</span> <span class="st">&quot;AOE&quot;</span>)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(crime.cor)[ord]</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;murder&quot;   &quot;assault&quot;  &quot;rape&quot;     &quot;robbery&quot;  &quot;burglary&quot;</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="co"># [6] &quot;larceny&quot;  &quot;auto&quot;</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>crime.cor <span class="ot">&lt;-</span> crime.cor[ord, ord]</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="co"># &quot;association graph&quot;: network of correlations</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a><span class="fu">qgraph</span>(crime.cor, </span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;Crime data:</span><span class="sc">\n</span><span class="st">correlations&quot;</span>, <span class="at">title.cex =</span> <span class="fl">1.5</span>,</span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">graph =</span> <span class="st">&quot;cor&quot;</span>,</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">layout =</span> <span class="st">&quot;circle&quot;</span>,</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">minimum =</span> <span class="st">&quot;sig&quot;</span>, <span class="at">sampleSize =</span> <span class="fu">nrow</span>(crime), <span class="at">alpha =</span> <span class="fl">0.01</span>,</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">grey</span>(.<span class="dv">9</span>), <span class="at">vsize =</span> <span class="dv">12</span>,</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> <span class="fu">rownames</span>(crime.cor),</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">posCol =</span> <span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-crime-cor" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-cor-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-crime-cor-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:60.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-cor-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.53: Network diagram depicting the correlations among the crime variables. Only edges for correlations that are significant at the <span class="math inline">\(\alpha = 0.01\)</span> level are displayed.
</figcaption>
</figure>
</div>
</div>
</div>
<!--

::: {.cell layout-align="center"}

```{.r .cell-code}
knitr::include_graphics("images/crime-cor.png")
```

::: {.cell-output-display}
![Network diagram depicting the correlations among the crime variables.](images/crime-cor.png){#fig-crime-cor-image fig-align='center' width=80%}
:::
:::

-->
<p>In this figure, you can see the group of property crimes (auto theft, larceny, burglary) at the left separated from the violent crimes against persons at the right.</p>
</section>
<section id="sec-partial-cor" class="level3" data-number="4.12.2">
<h3 data-number="4.12.2"><span class="header-section-number">4.12.2</span> Partial correlations</h3>
<p>Among the more important statistical applications of network graph theory is the idea that you can also use them to study the the <em>partial</em> (conditional) associations among variables with the contributions of all other variables removed in what are called Graphical Gaussian Models (GGMs) <span class="citation" data-cites="Lauritzen1996 Hojsgaard2012graphical">(<a href="#ref-Hojsgaard2012graphical" role="doc-biblioref">Højsgaard et al., 2012</a>; <a href="#ref-Lauritzen1996" role="doc-biblioref">Lauritzen, 1996</a>)</span>. In a network diagram of these partial associations, </p>
<ul>
<li><p>The edges between nodes represent the <em>partial correlations</em> between those variables.</p></li>
<li><p>The absence of an edge between two nodes indicates their variables are <em>conditionally independent</em>, given the other variables.</p></li>
</ul>
<p>So, whereas a network diagram of correlations shows <em>marginal associations</em> ignoring other variables, one of partial correlations allows you to visualize the <em>direct</em> relationship between each pair of variables, removing the indirect effects that might be mediated through all other variables.</p>
<p>For a set of variables <span class="math inline">\(X = \{x_1, x_2, \dots, x_p \}\)</span>, the partial correlation between <span class="math inline">\(x_i\)</span> and <span class="math inline">\(x_i\)</span>, controlling for all other variables <span class="math inline">\(Z = X \setminus \{x_i, x_j\} = x_\text{others}\)</span> is equivalent to the correlation between the <em>residuals</em> of the linear regressions of <span class="math inline">\(x_i\)</span> on all other <span class="math inline">\(\mathbf{Z}\)</span> and <span class="math inline">\(x_j\)</span> on <span class="math inline">\(\mathbf{Z}\)</span>. (The notation <span class="math inline">\(X \setminus \{x_i, x_j\}\)</span> is read as “<span class="math inline">\(X\)</span> without the set <span class="math inline">\(\{x_i, x_j\}\)</span>”).</p>
<p>Mathematically, let <span class="math inline">\(\hat{x}_i\)</span> and <span class="math inline">\(\hat{x}_j\)</span> be the predicted values from the linear regressions of <span class="math inline">\(x_i\)</span> on <span class="math inline">\(\mathbf{Z}\)</span> and of <span class="math inline">\(x_j\)</span> on <span class="math inline">\(\mathbf{Z}\)</span>, respectively. The partial correlation <span class="math inline">\(p_{ij}\)</span> between <span class="math inline">\(x_i\)</span> and <span class="math inline">\(x_j\)</span> controlling for <span class="math inline">\(\mathbf{Z}\)</span> is given by:</p>
<p><span id="eq-parcor"><span class="math display">\[
p_{x_i,x_j|\mathbf{Z}} = r( x_i, x_j \mid \text{others}) = \text{cor}[ (x_i - \hat{x}_i),\; (x_j - \hat{x}_j)]
\tag{4.3}\]</span></span></p>
<p>But, rather than running all these linear regressions, they can all be computed from the inverse of the correlation matrix <span class="citation" data-cites="Whittaker1990">(<a href="#ref-Whittaker1990" role="doc-biblioref">Whittaker, 1990</a>, Ch. 5)</span>, a relation first noted by <span class="citation" data-cites="Dempster1972">Dempster (<a href="#ref-Dempster1972" role="doc-biblioref">1972</a>)</span>. Let <span class="math inline">\(\mathbf{R}\)</span> be the correlation matrix of the variables. Then, the matrix <span class="math inline">\(\mathbf{P}\)</span> of partial correlations can be obtained from the negative inverse, <span class="math inline">\(-\mathbf{R}^{-1}\)</span>, standardized to a correlation matrix by dividing by the square root of product of its diagonal elements,</p>
<p><span class="math display">\[
P_{ij} = - \frac{R^{-1}_{ij}}{\sqrt{(R^{-1}_{ii} \cdot R^{-1}_{jj})}} \:\: .
\]</span></p>
<!-- Detrius from earlier descriptions ...
$$
\mathbf{P} = -\mathbf{R}^{-1} \comma
$$
where $\mathbf{R}^{-1}$ is the inverse of the correlation matrix $\mathbf{R}$.

Let $\mathbf{S}$ be the covariance matrix of the variable, and $\mathbf{R}$ be the corresponding
correlation matrix.
Then the matrix of partial variances and covariances, $\mathbf{T}$ is

$$
\mathbf{T} = -\mathbf{S}^{-1} \period
$$
The diagonal elements 
$R_{ij} = \frac{S_{ij}}{\sqrt{S_{ii} S_{jj}}}$
-->
<p>The practical implications of this are:</p>
<ul>
<li><p>If a partial correlation is close to zero, it suggests the relationship between two variables is primarily <em>mediated</em> through other variables.</p></li>
<li><p>Non-zero partial correlations indicate a direct relationship that persists after controlling for other variables.</p></li>
</ul>
<p><a href="#fig-crime-partial-spring" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-partial-spring</span></a> shows the partial correlation network for the crime data, using the <code>qgraph()</code> argument <code>graph = "pcor"</code> To provide a more interpretable result, the argument <code>layout = "spring"</code> positions the nodes using a force-embedded algorithm where edges act like springs, pulling connected nodes together and unconnected nodes repel each other, pushing them apart.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb72" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qgraph</span>(crime.cor, </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Crime data:</span><span class="sc">\n</span><span class="st">partial correlations&quot;</span>, <span class="at">title.cex =</span> <span class="fl">1.5</span>,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">graph =</span> <span class="st">&quot;pcor&quot;</span>,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">layout =</span> <span class="st">&quot;spring&quot;</span>, <span class="at">repulsion =</span> <span class="fl">1.2</span>,</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">minimum =</span> <span class="st">&quot;sig&quot;</span>, <span class="at">sampleSize =</span> <span class="fu">nrow</span>(crime), <span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="fu">grey</span>(.<span class="dv">9</span>), <span class="at">vsize =</span> <span class="dv">14</span>,</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">labels =</span> <span class="fu">rownames</span>(crime.cor),</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">edge.labels =</span> <span class="cn">TRUE</span>, <span class="at">edge.label.cex =</span> <span class="fl">1.7</span>,</span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">posCol =</span> <span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<div class="cell-output-display">
<div id="fig-crime-partial-spring" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-partial-spring-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch04/fig-crime-partial-spring-1.png" class="img-fluid quarto-figure quarto-figure-center" style="width:60.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-partial-spring-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.54: Network diagram of partial correlations among the crime variables, controlling for all others. Variable nodes have been positioned by a “spring” layout method …
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-crime-partial-spring" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-partial-spring</span></a> shows that, once all other crime variables are controlled for each pair, there remain only a few partial correlations at the <span class="math inline">\(\alpha = 0.05\)</span> level. Of these, only the largest three in absolute value are significant at <span class="math inline">\(\alpha = 0.01\)</span>.</p>
<p>Thus, once all other variables are taken into account, what remains is mainly a strong positive association between burglary and larceny and a moderate one between auto theft and robbery. There also remains a moderate negative correlation between murder and larceny. The spring layout makes it clear that, with suppression of weak edges, auto theft and robbery form a cluster separated from the other variables.</p>
</section>
<section id="sec-pvPlot" class="level3" data-number="4.12.3">
<h3 data-number="4.12.3"><span class="header-section-number">4.12.3</span> Visualizing partial correlations</h3>
<p>Just as you can visualize <em>marginal</em> association between variables in a scatterplot, you can also visualize <em>conditional</em> association. A <strong>partial variables plot</strong> is simply a scatterplot of the partial residuals <span class="math inline">\(e_i = (x_i - \hat{x}_i)\)</span> from a regression of <span class="math inline">\(x_i\)</span> on the other variables <span class="math inline">\(Z\)</span> against those <span class="math inline">\(e_j = (x_j - \hat{x}_j)\)</span> for another variable <span class="math inline">\(x_j\)</span>.</p>
<p>In this, you can use all the bells and whistles of standard scatterplots (regression lines, smooths, data ellipses, …) to listen more attentively to the story partial association has to tell. The function <code>pvPlot()</code> calculates the partial residuals and then calls <code>car::dataEllipse()</code> for display. The five most “unusual” observations by Mahalanobis <span class="math inline">\(D^2\)</span> are identified with their abbreviated state labels. <a href="#fig-crime-pvPlots" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-crime-pvPlots</span></a> shows these plots for the variable pairs with the two largest partial correlations.</p>
<!-- **TODO**: I'd like to put these two figures side-by-side and integrate the code chunks, but `car::scatterplot()` doesn't seem to allow this. -->
<p><strong>TODO</strong>: Put <code>pvPlot</code> into <code>heplots</code> package or <code>car</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb73" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/pvPlot.R&quot;</span>)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="co"># select numeric, make `st` into rownames</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>crime.num <span class="ot">&lt;-</span> crime <span class="sc">|&gt;</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">&quot;st&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">where</span>(is.numeric))</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pvPlot</span>(crime.num, <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">&quot;burglary&quot;</span>, <span class="st">&quot;larceny&quot;</span>), </span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">5</span>),</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">cex.lab =</span> <span class="fl">1.5</span>)</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a><span class="fu">pvPlot</span>(crime.num, <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">&quot;robbery&quot;</span>, <span class="st">&quot;auto&quot;</span>),</span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">id =</span> <span class="fu">list</span>(<span class="at">n=</span><span class="dv">5</span>),</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">cex.lab =</span> <span class="fl">1.5</span>)</span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-crime-pvPlots" class="quarto-float quarto-figure quarto-figure-center" data-fig-align="center">
<figure class="quarto-float quarto-float-fig">
<div aria-describedby="fig-crime-pvPlots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/crime-pvPlot-1-2.png" class="img-fluid quarto-figure quarto-figure-center" style="width:100.0%" />
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crime-pvPlots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure 4.55: Partial variables plots for burglary and larceny (left) and for robbery and auto theft (right) in the network diagram for partial correlations of the crime variables.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the pvPlot for burglary and larceny, you can see that the high partial correlation is largely driven by the extreme points at the left and and right sides. Once all other variables are taken into account, Arizona (AZ) and Hawaii (HI) have larger incidence of both crimes, while Arkansas (AK) are smaller on both.</p>
<p>In the pvPlot for robbery and auto theft, New York stands out as an influential, high-leverage point (see <a href="#sec-leverage" class="quarto-xref"><span class="quarto-unresolved-ref">sec-leverage</span></a>); Massachusetts (MA) is noteworthy because auto theft in that state is considerably higher than what would be predicted from all other variables.</p>
<!---
## Multivariate thinking and visualization

**TODO**: These are just initial notes on a chapter summary, and pointing the way to dimension reduction
methods in @sec-pca-biplot.

This chapter has covered a lot of ground. We started with simple scatterplots and how to enhance them
with graphical summaries and annotations ...

**The two curses**

Multivariate data is often said to suffer from the **curse of dimensionality** [@Bellman1957],
meaning that that as the dimensionality of data increases, the volume of the space increases so fast that the available data become sparse, so that the amount of data needed often grows exponentially with the dimensionality.

But, there is another curse here, the **curse of TWO-dimensionality**,
meaning that as the dimensionality of data increases, what we can display and understand from
a 2D image decreases rapidly with the number of observations and number of groups represented.
-->
</section>
</section>
<section id="what-have-we-learned" class="level2" data-number="4.13">
<h2 data-number="4.13"><span class="header-section-number">4.13</span> What have we learned?</h2>
<p>This chapter provides a comprehensive toolkit for visualizing multivariate relationships, transforming the humble scatterplot into a powerful engine for data exploration and discovery.</p>
<ul>
<li><p><strong>Enhance your plots</strong>: The basic scatterplot becomes far more informative when we add <strong>smoothers</strong> (regression lines, loess curves), <strong>stratifiers</strong> (grouping by color, shape, or panels), and <strong>data ellipses</strong> that capture correlation, variance, and regression relationships in a single elegant geometric form. These annotations help turn static dots into informative stories about relationships in your data.</p></li>
<li><p><strong>Data ellipses are a visualization Multi-Tool</strong>: These simple geometric summaries encode a remarkable amount of information—means, standard deviations, correlation, regression slopes, and confidence regions—all in one visual package. When your data are roughly bivariate normal, the ellipse becomes a sufficient summary, telling you everything you need to know about the relationship between two variables. Relaxing the normality assumption, robust methods like the bagplot can serve nearly as well.</p></li>
<li><p><strong>Simpson’s Paradox lurks everywhere</strong>: The grouping variables you ignore can completely reverse the relationships you see, once you properly account for the grouping variables. Marginal correlations can be negative while all within-group correlations—conditional on the grouping variable–are positive (or vice versa). This fundamental lesson reminds us that <strong>context matters</strong>—always consider what variables you might be overlooking.</p></li>
<li><p><strong>Scatterplot matrices scale gracefully</strong>: When you have multiple variables, pairs plots let you see all pairwise relationships simultaneously. Add <strong>visual thinning</strong> (removing points, keeping only smoothers and ellipses) and <strong>effect ordering</strong> (arranging variables by similarity) to handle even larger numbers of variables while maintaining interpretability.</p></li>
<li><p><strong>Generalized pairs plots bridge data types</strong>: Modern extensions handle mixtures of continuous and categorical variables elegantly, using appropriate plot types for each combination—scatterplots for continuous pairs, box plots for continuous-categorical combinations, and mosaic plots for categorical pairs. This unified framework means no variable gets left behind.</p></li>
<li><p><strong>Parallel coordinates reveal high-dimensional patterns</strong>: When scatterplot matrices reach their limits, parallel coordinate plots let you visualize dozens of variables simultaneously. Each observation becomes a connected line across parallel axes, revealing clusters, outliers, and multivariate patterns that would be invisible in traditional 2D projections.</p></li>
<li><p><strong>Tours provide dynamic exploration</strong>: Animated statistical tours take you on guided journeys through high-dimensional data space, smoothly rotating through different 2D projections. Whether taking a random grand tour or a guided tour optimized for specific features (clusters, outliers, group separation), these methods reveal structures hidden in static displays.</p></li>
<li><p><strong>Network diagrams conquer “big p” data</strong>: When correlations become too numerous to display traditionally, network graphs show variables as nodes connected by edges representing associations. Partial correlation networks reveal <strong>direct</strong> relationships between variables with all other influences removed—the difference between marginal and conditional independence becomes visually apparent.</p></li>
<li><p><strong>Visual thinning is a superpower</strong>: As datasets grow in size and complexity, strategic removal of visual elements (points, axes, labels) while retaining essential summaries (trends, ellipses, connections) lets you maintain clarity and insight. Less can indeed be more when every remaining element carries a greater share of graphic information.</p></li>
</ul>
<p>Having understood these lessons of multivariate thinking and visualization in <em>data space</em>, you are now prepared to take the next step in <a href="#sec-pca-biplot" class="quarto-xref"><span class="quarto-unresolved-ref">sec-pca-biplot</span></a> to consider how to enhance your understanding of multivariate data using <strong>multivariate juicers</strong> which project high-dimensional data into a lower-dimensional space in which important effects can be more easily seen.</p>
</section>
<section id="exercises" class="level2" data-number="4.14">
<h2 data-number="4.14"><span class="header-section-number">4.14</span> Exercises</h2>
<div id="exr-salary-smooth" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.1</strong></span> Using the <code>Salaries</code> dataset, create one or more plots to compare different smoothing methods for the relationship between <code>yrs.since.phd</code> and <code>salary</code> shown in <a href="#fig-Salaries-discipline" class="quarto-xref">Figure <span class="quarto-unresolved-ref">fig-Salaries-discipline</span></a>. Include linear regression, quadratic polynomial, and loess smoothers,</p>
<div class="sourceCode" id="cb74" data-code-line-numbers=""><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Salaries, <span class="at">package =</span> <span class="st">&quot;carData&quot;</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Your code here</span></span></code></pre></div>
</div>
<div id="exr-salary-smooth2" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.2</strong></span> One alternative to a loess smooth, which allows a <code>span</code> argument to control the degree of smoothing is a <strong>natural spline</strong>, that can be used in <code>geom_smooth()</code> using the argument <code>formula = y ~ splines::ns(x, df=)</code>, where <code>df</code> is the equivalent number of degrees of freedom for the spline smoother. Re-do <a href="#exr-salary-smooth" class="quarto-xref">Exercise <span class="quarto-unresolved-ref">exr-salary-smooth</span></a>, but trying out this smoothing method for several values of <code>df</code>.</p>
</div>
<!-- **Package summary** -->
<!-- ## References {.unnumbered} -->
<div id="quarto-navigation-envelope" class="hidden">
<p><span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyLXRpdGxl">Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXItdGl0bGU=">Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uZXh0"><span class="chapter-number">5</span>  <span class="chapter-title">Dimension Reduction</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1wcmV2"><span class="chapter-number">3</span>  <span class="chapter-title">Getting Started</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi9pbmRleC5odG1sUHJlZmFjZQ==">Preface</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wMC1BdXRob3IuaHRtbEF1dGhvcg==">Author</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tMQ==">Orienting Ideas</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wMS1QcmVsdWRlLmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjE8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+V2FybS11cC1FeGVyY2lzZXM8L3NwYW4+"><span class="chapter-number">1</span>  <span class="chapter-title">Warm-up Exercises</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wMi1pbnRyby5odG1sPHNwYW4tY2xhc3M9J2NoYXB0ZXItbnVtYmVyJz4yPC9zcGFuPi0tPHNwYW4tY2xhc3M9J2NoYXB0ZXItdGl0bGUnPkludHJvZHVjdGlvbjwvc3Bhbj4="><span class="chapter-number">2</span>  <span class="chapter-title">Introduction</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wMy1nZXR0aW5nX3N0YXJ0ZWQuaHRtbDxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+Mzwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5HZXR0aW5nLVN0YXJ0ZWQ8L3NwYW4+"><span class="chapter-number">3</span>  <span class="chapter-title">Getting Started</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tMg==">Exploratory Methods</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wNC1tdWx0aXZhcmlhdGVfcGxvdHMuaHRtbDxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+NDwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5QbG90cy1vZi1NdWx0aXZhcmlhdGUtRGF0YTwvc3Bhbj4="><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wNS1wY2EtYmlwbG90Lmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjU8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+RGltZW5zaW9uLVJlZHVjdGlvbjwvc3Bhbj4="><span class="chapter-number">5</span>  <span class="chapter-title">Dimension Reduction</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tMw==">Univariate Linear Models</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wNi1saW5lYXJfbW9kZWxzLmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjY8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+T3ZlcnZpZXctb2YtTGluZWFyLW1vZGVsczwvc3Bhbj4="><span class="chapter-number">6</span>  <span class="chapter-title">Overview of Linear models</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wNy1saW5lYXJfbW9kZWxzLXBsb3RzLmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjc8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+UGxvdHMtZm9yLVVuaXZhcmlhdGUtUmVzcG9uc2UtTW9kZWxzPC9zcGFuPg=="><span class="chapter-number">7</span>  <span class="chapter-title">Plots for Univariate Response Models</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wOC1saW4tbW9kLXRvcGljcy5odG1sPHNwYW4tY2xhc3M9J2NoYXB0ZXItbnVtYmVyJz44PC9zcGFuPi0tPHNwYW4tY2xhc3M9J2NoYXB0ZXItdGl0bGUnPlRvcGljcy1pbi1MaW5lYXItTW9kZWxzPC9zcGFuPg=="><span class="chapter-number">8</span>  <span class="chapter-title">Topics in Linear Models</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8wOS1jb2xsaW5lYXJpdHktcmlkZ2UuaHRtbDxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+OTwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5Db2xsaW5lYXJpdHktJi1SaWRnZS1SZWdyZXNzaW9uPC9zcGFuPg=="><span class="chapter-number">9</span>  <span class="chapter-title">Collinearity &amp; Ridge Regression</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tNA==">Multivariate Linear Models</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xMC1ob3RlbGxpbmcuaHRtbDxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+MTA8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+SG90ZWxsaW5nJ3MtJFReMiQ8L3NwYW4+"><span class="chapter-number">10</span>  <span class="chapter-title">Hotelling’s <span class="math inline">\(T^2\)</span></span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xMS1tbG0tcmV2aWV3Lmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjExPC9zcGFuPi0tPHNwYW4tY2xhc3M9J2NoYXB0ZXItdGl0bGUnPk11bHRpdmFyaWF0ZS1MaW5lYXItTW9kZWxzPC9zcGFuPg=="><span class="chapter-number">11</span>  <span class="chapter-title">Multivariate Linear Models</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xMi1tbG0tdml6Lmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPjEyPC9zcGFuPi0tPHNwYW4tY2xhc3M9J2NoYXB0ZXItdGl0bGUnPlZpc3VhbGl6aW5nLU11bHRpdmFyaWF0ZS1Nb2RlbHM8L3NwYW4+"><span class="chapter-number">12</span>  <span class="chapter-title">Visualizing Multivariate Models</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xMy1lcWNvdi5odG1sPHNwYW4tY2xhc3M9J2NoYXB0ZXItbnVtYmVyJz4xMzwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5WaXN1YWxpemluZy1FcXVhbGl0eS1vZi1Db3ZhcmlhbmNlLU1hdHJpY2VzPC9zcGFuPg=="><span class="chapter-number">13</span>  <span class="chapter-title">Visualizing Equality of Covariance Matrices</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xNC1pbmZsLXJvYnVzdC5odG1sPHNwYW4tY2xhc3M9J2NoYXB0ZXItbnVtYmVyJz4xNDwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5NdWx0aXZpYXRlLUluZmx1ZW5jZS1hbmQtUm9idXN0LUVzdGltYXRpb248L3NwYW4+"><span class="chapter-number">14</span>  <span class="chapter-title">Multiviate Influence and Robust Estimation</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8xNS1jYXNlLXN0dWRpZXMuaHRtbDxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+MTU8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+Q2FzZS1TdHVkaWVzPC9zcGFuPg=="><span class="chapter-number">15</span>  <span class="chapter-title">Case Studies</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tNQ==">End Matter</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi85MS1jb2xvcGhvbi5odG1sQ29sb3Bob24=">Colophon</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi85NS1yZWZlcmVuY2VzLmh0bWxSZWZlcmVuY2Vz">References</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOnF1YXJ0by1zaWRlYmFyLXNlY3Rpb24tNg==">Appendices</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyOi8yMS1kaXNjcmltLmh0bWw8c3Bhbi1jbGFzcz0nY2hhcHRlci1udW1iZXInPkE8L3NwYW4+LS08c3Bhbi1jbGFzcz0nY2hhcHRlci10aXRsZSc+RGlzY3JpbWluYW50LWFuYWx5c2lzPC9zcGFuPg=="><span class="chapter-number">A</span>  <span class="chapter-title">Discriminant analysis</span></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWJyZWFkY3J1bWJzLUV4cGxvcmF0b3J5LU1ldGhvZHM=">Exploratory Methods</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWJyZWFkY3J1bWJzLTxzcGFuLWNsYXNzPSdjaGFwdGVyLW51bWJlcic+NDwvc3Bhbj4tLTxzcGFuLWNsYXNzPSdjaGFwdGVyLXRpdGxlJz5QbG90cy1vZi1NdWx0aXZhcmlhdGUtRGF0YTwvc3Bhbj4="><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span></p>
</div>
<div id="quarto-meta-markdown" class="hidden">
<p><span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW1ldGF0aXRsZQ=="><span id="sec-multivariate_plots" class="quarto-section-identifier"><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span> – Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLXR3aXR0ZXJjYXJkdGl0bGU="><span id="sec-multivariate_plots" class="quarto-section-identifier"><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span> – Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW9nY2FyZHRpdGxl"><span id="sec-multivariate_plots" class="quarto-section-identifier"><span class="chapter-number">4</span>  <span class="chapter-title">Plots of Multivariate Data</span></span> – Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW1ldGFzaXRlbmFtZQ==">Visualizing Multivariate Data and Models in R</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLXR3aXR0ZXJjYXJkZGVzYw=="></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW9nY2FyZGRkZXNj"></span></p>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-Andrews:72" class="csl-entry" role="listitem">
Andrews, D. F. (1972). Plots of high dimensional data. <em>Biometrics</em>, <em>28</em>, 123–136.
</div>
<div id="ref-Asimov:85" class="csl-entry" role="listitem">
Asimov, D. (1985). Grand tour. <em>SIAM Journal of Scientific and Statistical Computing</em>, <em>6</em>(1), 128–143.
</div>
<div id="ref-Barabasi2016network" class="csl-entry" role="listitem">
Barab’asi, A.-L. (2016). <em>Network science</em>. Cambridge University Press.
</div>
<div id="ref-Becker:1996:VDC" class="csl-entry" role="listitem">
Becker, R. A., Cleveland, W. S., &amp; Shyu, M.-J. (1996). The visual design and control of trellis display. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>(2), 123–155.
</div>
<div id="ref-Black-etal-2018" class="csl-entry" role="listitem">
Black, C., Southwell, C., Emmerson, L., Lunn, D., &amp; Hart, T. (2018). Time-lapse imagery of <span>Adélie</span> penguins reveals differential winter strategies and breeding site occupation. <em>PLOS ONE</em>, <em>13</em>(3), e0193532. <a href="https://doi.org/10.1371/journal.pone.0193532">https://doi.org/10.1371/journal.pone.0193532</a>
</div>
<div id="ref-Blishen-etal-1987" class="csl-entry" role="listitem">
Blishen, B., Carroll, W., &amp; Moore, C. (1987). The 1981 socioeconomic index for occupations in canada. <em>Canadian Review of Sociology/Revue Canadienne de Sociologie</em>, <em>24</em>(4), 465–488. <a href="https://doi.org/10.1111/j.1755-618x.1987.tb00639.x">https://doi.org/10.1111/j.1755-618x.1987.tb00639.x</a>
</div>
<div id="ref-BondyMurty2008" class="csl-entry" role="listitem">
Bondy, J. A., &amp; Murty, U. S. R. (2008). <em>Graph theory</em>. Springer.
</div>
<div id="ref-R-gggda" class="csl-entry" role="listitem">
Brunson, J. C., &amp; Gracey, J. (2025). <em>Gggda: A ’ggplot2’ extension for geometric data analysis</em>. <a href="https://github.com/corybrunson/gggda">https://github.com/corybrunson/gggda</a>
</div>
<div id="ref-Buja-etal-2005" class="csl-entry" role="listitem">
Buja, A., Cook, D., Asimov, D., &amp; Hurley, C. (2005). Computational methods for high-dimensional rotations in data visualization. In J. S. CR Rao EJ Wegman (Ed.), <em>Handbook of statistics</em> (pp. 391–413). Elsevier. <a href="https://doi.org/10.1016/s0169-7161(04)24014-7">https://doi.org/10.1016/s0169-7161(04)24014-7</a>
</div>
<div id="ref-Ocagne:1885" class="csl-entry" role="listitem">
cagne, M. (1885). <em>Coordonn<span>é</span>es parall<span>è</span>les et axiales: M<span>é</span>thode de transformation g<span>é</span>om<span>é</span>trique et proc<span>é</span>d<span>é</span> nouveau de calcul graphique d<span>é</span>duits de la consid<span>é</span>ration des coordonn<span>é</span>es parall<span>è</span>lles</em>. Gauthier-Villars. <a href="http://historical.library.cornell.edu/cgi-bin/cul.math/docviewer?did=00620001&amp;seq=3">http://historical.library.cornell.edu/cgi-bin/cul.math/docviewer?did=00620001&amp;seq=3</a>
</div>
<div id="ref-Chambers-etal:83" class="csl-entry" role="listitem">
Chambers, J. M., Cleveland, W. S., Kleiner, B., &amp; Tukey, P. A. (1983). <em>Graphical methods for data analysis</em>. Wadsworth.
</div>
<div id="ref-ChambersHastie1991" class="csl-entry" role="listitem">
Chambers, J. M., &amp; Hastie, T. J. (1991). <em>Statistical models in s</em> (p. 624). Chapman &amp; Hall/CRC.
</div>
<div id="ref-Cleveland:79" class="csl-entry" role="listitem">
Cleveland, W. S. (1979). Robust locally weighted regression and smoothing scatterplots. <em>Journal of the American Statistical Association</em>, <em>74</em>, 829–836.
</div>
<div id="ref-Cleveland:85" class="csl-entry" role="listitem">
Cleveland, W. S. (1985). <em>The elements of graphing data</em>. Wadsworth Advanced Books.
</div>
<div id="ref-ClevelandDevlin:88" class="csl-entry" role="listitem">
Cleveland, W. S., &amp; Devlin, S. J. (1988). Locally weighted regression: An approach to regression analysis by local fitting. <em>Journal of the American Statistical Association</em>, <em>83</em>, 596–610.
</div>
<div id="ref-ClevelandMcGill:84b" class="csl-entry" role="listitem">
Cleveland, W. S., &amp; McGill, R. (1984). Graphical perception: Theory, experimentation and application to the development of graphical methods. <em>Journal of the American Statistical Association</em>, <em>79</em>, 531–554.
</div>
<div id="ref-ClevelandMcGill:85" class="csl-entry" role="listitem">
Cleveland, W. S., &amp; McGill, R. (1985). Graphical perception and graphical methods for analyzing scientific data. <em>Science</em>, <em>229</em>, 828–833.
</div>
<div id="ref-Cook-etal-1995" class="csl-entry" role="listitem">
Cook, D., Buja, A., Cabrera, J., &amp; Hurley, C. (1995). Grand tour and projection pursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(3), 155. <a href="https://doi.org/10.2307/1390844">https://doi.org/10.2307/1390844</a>
</div>
<div id="ref-Cook-etal-2008" class="csl-entry" role="listitem">
Cook, D., Buja, A., Lee, E.-K., &amp; Wickham, H. (2008). Grand tours, projection pursuit guided tours, and manual controls. In <em>Handbook of data visualization</em> (pp. 295–314). Springer Berlin Heidelberg. <a href="https://doi.org/10.1007/978-3-540-33037-0_13">https://doi.org/10.1007/978-3-540-33037-0_13</a>
</div>
<div id="ref-CookLaa-mulgar" class="csl-entry" role="listitem">
Cook, D., &amp; Laa, U. (2024). <em>Interactively exploring high-dimensional data and models in <span>R</span></em>. Online. <a href="https://dicook.github.io/mulgar_book/">https://dicook.github.io/mulgar_book/</a>
</div>
<div id="ref-CookSwayne:2007" class="csl-entry" role="listitem">
Cook, D., &amp; Swayne, D. F. (2007). <em>Interactive and dynamic graphics for data analysis : With <span>R</span> and <span>GGobi</span></em>. Springer. <a href="http://www.ggobi.org/book/">http://www.ggobi.org/book/</a>
</div>
<div id="ref-Costantini2015" class="csl-entry" role="listitem">
Costantini, G., Epskamp, S., Borsboom, D., Perugini, M., Mõttus, R., Waldorp, L. J., &amp; Cramer, A. O. J. (2015). State of the <span class="nocase">aRt</span> personality research: A tutorial on network analysis of personality data in <span>R</span>. <em>Journal of Research in Personality</em>, <em>54</em>, 13–29. <a href="https://doi.org/10.1016/j.jrp.2014.07.003">https://doi.org/10.1016/j.jrp.2014.07.003</a>
</div>
<div id="ref-R-igraph" class="csl-entry" role="listitem">
Csárdi, G., Nepusz, T., Traag, V., Horvát, S., Zanini, F., Noom, D., &amp; Müller, K. (2024). <em><span class="nocase">igraph</span>: Network analysis and visualization in r</em>. <a href="https://doi.org/10.5281/zenodo.7682609">https://doi.org/10.5281/zenodo.7682609</a>
</div>
<div id="ref-Dempster:69" class="csl-entry" role="listitem">
Dempster, A. P. (1969). <em>Elements of continuous multivariate analysis</em>. Addison-Wesley.
</div>
<div id="ref-Dempster1972" class="csl-entry" role="listitem">
Dempster, A. P. (1972). Covariance selection. <em>Biometrics</em>, <em>28</em>(1), 157–175.
</div>
<div id="ref-Emerson-etal:2013" class="csl-entry" role="listitem">
Emerson, J. W., Green, W. A., Schloerke, B., Crowley, J., Cook, D., Hofmann, H., &amp; Wickham, H. (2013). The generalized pairs plot. <em>Journal of Computational and Graphical Statistics</em>, <em>22</em>(1), 79–91. <a href="https://doi.org/10.1080/10618600.2012.694762">https://doi.org/10.1080/10618600.2012.694762</a>
</div>
<div id="ref-Fox:2016:ARA" class="csl-entry" role="listitem">
Fox, J. (2016). <em>Applied regression analysis and generalized linear models</em> (Third edition.). SAGE.
</div>
<div id="ref-R-car" class="csl-entry" role="listitem">
Fox, J., &amp; Weisberg, S. (2019). <em>An <span>R</span> companion to applied regression</em> (Third). Sage. <a href="https://www.john-fox.ca/Companion/">https://www.john-fox.ca/Companion/</a>
</div>
<div id="ref-Friendly:91" class="csl-entry" role="listitem">
Friendly, M. (1991). <em><span>SAS System</span> for statistical graphics</em> (1st ed.). SAS Institute. <a href="http://www.sas. com/service/doc/pubcat/uspubcat/ind_files/56143.html">http://www.sas. com/service/doc/pubcat/uspubcat/ind_files/56143.html</a>
</div>
<div id="ref-Friendly:94a" class="csl-entry" role="listitem">
Friendly, M. (1994). Mosaic displays for multi-way contingency tables. <em>Journal of the American Statistical Association</em>, <em>89</em>, 190–200. <a href="http://www.jstor.org/stable/2291215">http://www.jstor.org/stable/2291215</a>
</div>
<div id="ref-Friendly:99:EMD" class="csl-entry" role="listitem">
Friendly, M. (1999). Extending mosaic displays: Marginal, conditional, and partial views of categorical data. <em>Journal of Computational and Graphical Statistics</em>, <em>8</em>(3), 373–395. <a href="http://datavis.ca/papers/drew/drew.pdf">http://datavis.ca/papers/drew/drew.pdf</a>
</div>
<div id="ref-Friendly:02:corrgram" class="csl-entry" role="listitem">
Friendly, M. (2002). Corrgrams: Exploratory displays for correlation matrices. <em>The American Statistician</em>, <em>56</em>(4), 316–324. <a href="https://doi.org/10.1198/000313002533">https://doi.org/10.1198/000313002533</a>
</div>
<div id="ref-Friendly2022" class="csl-entry" role="listitem">
Friendly, M. (2022). The life and works of andr<span>é</span>-michel guerry, revisited. <em>Sociological Spectrum</em>, <em>42</em>(4-6), 233–259. <a href="https://doi.org/10.1080/02732173.2022.2078450">https://doi.org/10.1080/02732173.2022.2078450</a>
</div>
<div id="ref-R-vcdExtra" class="csl-entry" role="listitem">
Friendly, M. (2025). <em>vcdExtra: ’Vcd’ extensions and additions</em>. <a href="https://doi.org/10.32614/CRAN.package.vcdExtra">https://doi.org/10.32614/CRAN.package.vcdExtra</a>
</div>
<div id="ref-FriendlyKwan:03:effect" class="csl-entry" role="listitem">
Friendly, M., &amp; Kwan, E. (2003). Effect ordering for data displays. <em>Computational Statistics and Data Analysis</em>, <em>43</em>(4), 509–539. <a href="https://doi.org/10.1016/S0167-9473(02)00290-6">https://doi.org/10.1016/S0167-9473(02)00290-6</a>
</div>
<div id="ref-FriendlyMeyer:2016:DDAR" class="csl-entry" role="listitem">
Friendly, M., &amp; Meyer, D. (2016). <em>Discrete data analysis with <span>R</span>: Visualization and modeling techniques for categorical and count data</em>. Chapman &amp; Hall/CRC.
</div>
<div id="ref-Friendly-etal:ellipses:2013" class="csl-entry" role="listitem">
Friendly, M., Monette, G., &amp; Fox, J. (2013). Elliptical insights: Understanding statistical methods through elliptical geometry. <em>Statistical Science</em>, <em>28</em>(1), 1–39. <a href="https://doi.org/10.1214/12-STS402">https://doi.org/10.1214/12-STS402</a>
</div>
<div id="ref-Galton:1886" class="csl-entry" role="listitem">
Galton, F. (1886). Regression towards mediocrity in hereditary stature. <em>Journal of the Anthropological Institute</em>, <em>15</em>, 246–263. <a href="http://www.jstor.org/cgi-bin/jstor/viewitem/09595295/dm995266/99p0374f/0">http://www.jstor.org/cgi-bin/jstor/viewitem/09595295/dm995266/99p0374f/0</a>
</div>
<div id="ref-Gannett:1898" class="csl-entry" role="listitem">
Gannett, H. (1898). <em>Statistical atlas of the united states, eleventh (1890) census</em>. U.S. Government Printing Office.
</div>
<div id="ref-Gorman2014" class="csl-entry" role="listitem">
Gorman, K. B., Williams, T. D., &amp; Fraser, W. R. (2014). Ecological sexual dimorphism and environmental variability within a community of antarctic penguins (genus pygoscelis). <em><span>PLoS</span> <span>ONE</span></em>, <em>9</em>(3), e90081. <a href="https://doi.org/10.1371/journal.pone.0090081">https://doi.org/10.1371/journal.pone.0090081</a>
</div>
<div id="ref-Grandjean2016" class="csl-entry" role="listitem">
Grandjean, M. (2016). A social network analysis of <span>T</span>witter: <span>M</span>apping the digital humanities community. <em>Cogent Arts &amp;Amp; Humanities</em>, <em>3</em>(1), 1171458. <a href="https://doi.org/10.1080/23311983.2016.1171458">https://doi.org/10.1080/23311983.2016.1171458</a>
</div>
<div id="ref-Guerry:1833" class="csl-entry" role="listitem">
Guerry, A.-M. (1833). <em>Essai sur la statistique morale de la <span>France</span></em>. Crochard.
</div>
<div id="ref-Harrison2023" class="csl-entry" role="listitem">
Harrison, P. (2023b). Langevitour: Smooth interactive touring of high dimensions, demonstrated with scRNA-seq data. <em>The R Journal</em>, <em>15</em>(2), 206–219. <a href="https://doi.org/10.32614/RJ-2023-046">https://doi.org/10.32614/RJ-2023-046</a>
</div>
<div id="ref-R-langevitour" class="csl-entry" role="listitem">
Harrison, P. (2023a). Langevitour: Smooth interactive touring of high dimensions, demonstrated with scRNA-seq data. <em>The R Journal</em>, <em>15</em>(2), 206–219. <a href="https://doi.org/10.32614/RJ-2023-046">https://doi.org/10.32614/RJ-2023-046</a>
</div>
<div id="ref-R-detourr" class="csl-entry" role="listitem">
Hart, C., &amp; Wang, E. (2022). <em>Detourr: Portable and performant tour animations</em>. <a href="https://CRAN.R-project.org/package=detourr">https://CRAN.R-project.org/package=detourr</a>
</div>
<div id="ref-Hartigan:75" class="csl-entry" role="listitem">
Hartigan, J. A. (1975a). <em>Clustering algorithms</em>. John Wiley; Sons.
</div>
<div id="ref-Hartigan:75b" class="csl-entry" role="listitem">
Hartigan, J. A. (1975b). Printer graphics for clustering. <em>Journal of Statistical Computing and Simulation</em>, <em>4</em>, 187–213.
</div>
<div id="ref-R-ggpcp" class="csl-entry" role="listitem">
Hofmann, H., VanderPlas, S., &amp; Ge, Y. (2022). <em>Ggpcp: Parallel coordinate plots in the ’ggplot2’ framework</em>. <a href="https://doi.org/10.32614/CRAN.package.ggpcp">https://doi.org/10.32614/CRAN.package.ggpcp</a>
</div>
<div id="ref-Hofstadter1979" class="csl-entry" role="listitem">
Hofstadter, D. R. (1979). <em>Gödel, escher, bach: An eternal golden braid</em>. Basic Books.
</div>
<div id="ref-Hojsgaard2012graphical" class="csl-entry" role="listitem">
Højsgaard, S., Edwards, D., &amp; Lauritzen, S. (2012). <em>Graphical models with <span>R</span></em>. Springer Science &amp; Business Media.
</div>
<div id="ref-R-palmerpenguins" class="csl-entry" role="listitem">
Horst, A. M., Hill, A. P., &amp; Gorman, K. B. (2020). <em>Palmerpenguins: Palmer archipelago (antarctica) penguin data</em>. <a href="https://doi.org/10.5281/zenodo.3960218">https://doi.org/10.5281/zenodo.3960218</a>
</div>
<div id="ref-Inselberg:1985" class="csl-entry" role="listitem">
Inselberg, A. (1985). The plane with parallel coordinates. <em>The Visual Computer</em>, <em>1</em>, 69–91.
</div>
<div id="ref-IsvoranuEpskamp2022" class="csl-entry" role="listitem">
Isvoranu, A.-M., Epskamp, S., Waldorp, L. J., &amp; Borsboom, D. (2022). <em>Network psychometrics with r: A guide for behavioral and social scientists</em>. Routledge. <a href="https://doi.org/10.4324/9781003111238">https://doi.org/10.4324/9781003111238</a>
</div>
<div id="ref-Lauritzen1996" class="csl-entry" role="listitem">
Lauritzen, S. L. (1996). <em>Graphical models</em>. Oxford University Press.
</div>
<div id="ref-LeeCook-2009" class="csl-entry" role="listitem">
Lee, E.-K., &amp; Cook, D. (2009). A projection pursuit index for large p small n data. <em>Statistics and Computing</em>, <em>20</em>(3), 381–392. <a href="https://doi.org/10.1007/s11222-009-9131-1">https://doi.org/10.1007/s11222-009-9131-1</a>
</div>
<div id="ref-R-liminal" class="csl-entry" role="listitem">
Lee, S. (2025). <em>Liminal: Multivariate data visualization with tours and embeddings</em>. <a href="https://github.com/sa-lee/liminal">https://github.com/sa-lee/liminal</a>
</div>
<div id="ref-MartiLaguna2003" class="csl-entry" role="listitem">
Martí, R., &amp; Laguna, M. (2003). Heuristics and meta-heuristics for 2-layer straight line crossing minimization. <em>Discrete Applied Mathematics</em>, <em>127</em>(3), 665–678.
</div>
<div id="ref-R-vcd" class="csl-entry" role="listitem">
Meyer, D., Zeileis, A., Hornik, K., &amp; Friendly, M. (2024). <em>Vcd: Visualizing categorical data</em>. <a href="https://doi.org/10.32614/CRAN.package.vcd">https://doi.org/10.32614/CRAN.package.vcd</a>
</div>
<div id="ref-Monette:90" class="csl-entry" role="listitem">
Monette, G. (1990). Geometry of multiple regression and interactive 3-<span>D</span> graphics. In J. Fox &amp; S. Long (Eds.), <em>Modern methods of data analysis</em> (pp. 209–256). SAGE Publications.
</div>
<div id="ref-R-ggdensity" class="csl-entry" role="listitem">
Otto, J., &amp; Kahle, D. (2023). <em>Ggdensity: Interpretable bivariate density visualization with ’ggplot2’</em>. <a href="https://doi.org/10.32614/CRAN.package.ggdensity">https://doi.org/10.32614/CRAN.package.ggdensity</a>
</div>
<div id="ref-Pearson:1901" class="csl-entry" role="listitem">
Pearson, K. (1901). On lines and planes of closest fit to systems of points in space. <em>Philosophical Magazine</em>, <em>6</em>(2), 559–572.
</div>
<div id="ref-PineoPorter2008" class="csl-entry" role="listitem">
Pineo, P. O., &amp; Porter, J. (2008). Occupational prestige in canada. <em>Canadian Review of Sociology</em>, <em>4</em>(1), 24–40. <a href="https://doi.org/10.1111/j.1755-618x.1967.tb00472.x">https://doi.org/10.1111/j.1755-618x.1967.tb00472.x</a>
</div>
<div id="ref-Robinaugh2019" class="csl-entry" role="listitem">
Robinaugh, D. J., Hoekstra, R. H. A., Toner, E. R., &amp; Borsboom, D. (2019). The network approach to psychopathology: A review of the literature 2008–2018 and an agenda for future research. <em>Psychological Medicine</em>, <em>50</em>(3), 353–366. <a href="https://doi.org/10.1017/s0033291719003404">https://doi.org/10.1017/s0033291719003404</a>
</div>
<div id="ref-Rousseeuw-etal:2012" class="csl-entry" role="listitem">
Rousseeuw, P. J., Ruits, I., &amp; Tukey, J. W. (1999). The bagplot: A bivariate boxplot. <em>The American Statistician</em>, <em>53</em>(4), 382–387.
</div>
<div id="ref-R-lattice" class="csl-entry" role="listitem">
Sarkar, D. (2008). <em>Lattice: Multivariate data visualization with r</em>. Springer. <a href="http://lmdvr.r-forge.r-project.org">http://lmdvr.r-forge.r-project.org</a>
</div>
<div id="ref-R-GGally" class="csl-entry" role="listitem">
Schloerke, B., Cook, D., Larmarange, J., Briatte, F., Marbach, M., Thoen, E., Elberg, A., &amp; Crowley, J. (2025). <em>GGally: Extension to ’ggplot2’</em>. <a href="https://doi.org/10.32614/CRAN.package.GGally">https://doi.org/10.32614/CRAN.package.GGally</a>
</div>
<div id="ref-Scott1992" class="csl-entry" role="listitem">
Scott, D. W. (1992). <em>Multivariate density estimation: Theory, practice, and visualization</em>. Wiley.
</div>
<div id="ref-Silverman:86" class="csl-entry" role="listitem">
Silverman, B. W. (1986). <em>Density estimation for statistics and data analysis</em>. Chapman &amp; Hall.
</div>
<div id="ref-Simpson:51" class="csl-entry" role="listitem">
Simpson, E. H. (1951). The interpretation of interaction in contingency tables. <em>Journal of the Royal Statistical Society, Series B</em>, <em>30</em>, 238–241.
</div>
<div id="ref-Swayne-etal-1998" class="csl-entry" role="listitem">
Swayne, D. F., Cook, D., &amp; Buja, A. (1998). XGobi: Interactive dynamic data visualization in the x window system. <em>Journal of Computational and Graphical Statistics</em>, <em>7</em>(1), 113–130. <a href="https://doi.org/10.1080/10618600.1998.10474764">https://doi.org/10.1080/10618600.1998.10474764</a>
</div>
<div id="ref-Swayne-etal-2003" class="csl-entry" role="listitem">
Swayne, D. F., Lang, D. T., Buja, A., &amp; Cook, D. (2003). <span>GGobi</span>: Evolving from <span>XGobi</span> into an extensible framework for interactive data visualization. <em>Computational Statistics &amp;Amp; Data Analysis</em>, <em>43</em>(4), 423–444. <a href="https://doi.org/10.1016/s0167-9473(02)00286-4">https://doi.org/10.1016/s0167-9473(02)00286-4</a>
</div>
<div id="ref-Tukey:77" class="csl-entry" role="listitem">
Tukey, J. W. (1977). <em>Exploratory data analysis</em>. Addison Wesley.
</div>
<div id="ref-VanderPlas2023" class="csl-entry" role="listitem">
VanderPlas, S., Ge, Y., Unwin, A., &amp; Hofmann, H. (2023). Penguins go parallel: A grammar of graphics framework for generalized parallel coordinate plots. <em>Journal of Computational and Graphical Statistics</em>, 1–16. <a href="https://doi.org/10.1080/10618600.2023.2195462">https://doi.org/10.1080/10618600.2023.2195462</a>
</div>
<div id="ref-R-loon" class="csl-entry" role="listitem">
Waddell, A., &amp; Oldford, R. W. (2025). <em>Loon: Interactive statistical data visualization</em>. <a href="https://doi.org/10.32614/CRAN.package.loon">https://doi.org/10.32614/CRAN.package.loon</a>
</div>
<div id="ref-Wegman:1990" class="csl-entry" role="listitem">
Wegman, E. J. (1990). Hyperdimensional data analysis using parallel coordinates. <em>Journal of the American Statistical Association</em>, <em>85</em>(411), 664–675.
</div>
<div id="ref-R-corrplot" class="csl-entry" role="listitem">
Wei, T., &amp; Simko, V. (2024). <em>R package ’corrplot’: Visualization of a correlation matrix</em>. <a href="https://github.com/taiyun/corrplot">https://github.com/taiyun/corrplot</a>
</div>
<div id="ref-West2001" class="csl-entry" role="listitem">
West, D. B. (2001). <em>Introduction to graph theory</em>. Prentice hall.
</div>
<div id="ref-Whittaker1990" class="csl-entry" role="listitem">
Whittaker, J. (1990). <em>Graphical models in applied multivariate statistics</em>. John Wiley; Sons.
</div>
<div id="ref-R-tourr" class="csl-entry" role="listitem">
Wickham, H., &amp; Cook, D. (2025). <em>Tourr: Tour methods for multivariate data visualisation</em>. <a href="https://github.com/ggobi/tourr">https://github.com/ggobi/tourr</a>
</div>
<div id="ref-Wickham-etal-2011" class="csl-entry" role="listitem">
Wickham, H., Cook, D., Hofmann, H., &amp; Buja, A. (2011). Tourr: An <span>R</span> package for exploring multivariate data with projections. <em>Journal of Statistical Software</em>, <em>40</em>(2). <a href="https://doi.org/10.18637/jss.v040.i02">https://doi.org/10.18637/jss.v040.i02</a>
</div>
<div id="ref-Wood:2006" class="csl-entry" role="listitem">
Wood, S. N. (2006). <em>Generalized additive models: An introduction with r</em>. Chapman; Hall/CRC Press.
</div>
<div id="ref-R-corrgram" class="csl-entry" role="listitem">
Wright, K. (2021). <em>Corrgram: Plot a correlogram</em>. <a href="https://doi.org/10.32614/CRAN.package.corrgram">https://doi.org/10.32614/CRAN.package.corrgram</a>
</div>
<div id="ref-R-loon-tour" class="csl-entry" role="listitem">
Xu, Z., &amp; Oldford, R. W. (2021). <em>Loon.tour: Tour in ’loon’</em>. <a href="https://cran.r-project.org/package=loon.tourr">https://cran.r-project.org/package=loon.tourr</a>
</div>
</div>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr />
<ol>
<li id="fn1"><p>Confidence bands allow us to visualize the uncertainty around a fitted regression curve. These can be of two types: <em>pointwise intervals</em> or <em>simultaneous intervals</em>. The default setting in `<code>ggplot2::geom_smooth()</code> calculates pointwise intervals (using <code>stats::predict.lm(..., interval="confidence")</code> at a confidence level <span class="math inline">\(1-\alpha\)</span> for the predicted response at <em>each value</em> <span class="math inline">\(x_i\)</span> of a predictor, and have the frequentist interpretation that over repeated sampling only <span class="math inline">\(100\;\alpha\)</span> of the predictions at <span class="math inline">\(x_i\)</span> will be outside that interval. In contrast, simultaneous intervals are calculated so that <span class="math inline">\(1 - \alpha\)</span> is the probability that <em>all of them</em> cover their corresponding true values simultaneously. These are necessarily wider than pointwise intervals. Commonly used methods for constructing simultaneous confidence bands in regression are the Bonferroni and Scheffé methods, which control the family-wise error rate over all values of <span class="math inline">\(x_i\)</span>. See <a href="https://en.wikipedia.org/wiki/Confidence_and_prediction_bands"></a> for precise definitions of these terms. These are different from a <em>prediction band</em>, which is used to represent the uncertainty about the value of a <strong>new</strong> data-point on the curve, but subject to the additional variance reflected in one observation.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The classic study by <span class="citation" data-cites="ClevelandMcGill:84b">Cleveland &amp; McGill (<a href="#ref-ClevelandMcGill:84b" role="doc-biblioref">1984</a>)</span>;<span class="citation" data-cites="ClevelandMcGill:85">Cleveland &amp; McGill (<a href="#ref-ClevelandMcGill:85" role="doc-biblioref">1985</a>)</span> shows that judgements of magnitude along a common scale are more accurate than those along separate, aligned scales.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>The dataset was collected by Bernard Blishen, William Carroll and Catherine Moore, but apparently unpublished. A version updated to the 1981 census is described in <span class="citation" data-cites="Blishen-etal-1987">Blishen et al. (<a href="#ref-Blishen-etal-1987" role="doc-biblioref">1987</a>)</span>.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>In R 4.5.0, a revised version of the <code>palmerpenguins</code> dataset, named <code>penguins</code> was added to the base R <span style="color: brown;"><strong>datasets</strong></span> package. This corrected a few of the problems that led me to create <code>heplots::peng</code>, but still retains the cases with missing data.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>The <span class="math inline">\(p^{th}\)</span>-quantile of a random variable <span class="math inline">\(Y\)</span> is a value, denoted by <span class="math inline">\(Q_Y(p)\)</span>, such that <span class="math inline">\(X \leq Q_Y(p)\)</span> with probability <span class="math inline">\(p\)</span>, i.e., <span class="math inline">\(\Pr (Y \leq Q_Y(p)) = p\)</span>. For continuous variables, the quantile is the inverse of the cumulative distribution function.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>After the long-nosed Cyrano de Bergerac from Edmund Rostand’s play. Actually, bird 283 is a female. But, it would be a mistake to call her “Rozanne” instead.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p>Other implementations of parallel coordinate plots in R include: <code>MASS::parcoord()</code>, <code>GGally::ggparcoord()</code> and <code>PairViz::pcp()</code>. The <strong>ggpcp</strong> version used here is the most general.<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p>This example was modified from one used by <span class="citation" data-cites="Cook-etal-2008">Cook et al. (<a href="#ref-Cook-etal-2008" role="doc-biblioref">2008</a>)</span>.<a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id = "quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/friendly\.github\.io\/Vis-MLM-book");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a  href="/03-getting_started.html" class="pagination-link" aria-label="Getting Started">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class='chapter-number'>3</span>  <span class='chapter-title'>Getting Started</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a  href="/05-pca-biplot.html" class="pagination-link" aria-label="Dimension Reduction">
        <span class="nav-page-text"><span class='chapter-number'>5</span>  <span class='chapter-title'>Dimension Reduction</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->

</body>

</html>
