<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.29">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>11&nbsp; Visualizing Multivariate Models – Visualizing Multivariate Data and Models in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./12-eqcov.html" rel="next">
<link href="./10-mlm-review.html" rel="prev">
<link href="./images/favicon/favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-0815c480559380816a4d1ea211a47e91.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-b5345210c3a79db27e5ded64c8911254.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/line-highlight-1.0.0/line-highlight.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./09-hotelling.html">Multivariate Linear Models</a></li><li class="breadcrumb-item"><a href="./11-mlm-viz.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Visualizing Multivariate Models</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./09-hotelling.html">Multivariate Linear Models</a></li><li class="breadcrumb-item"><a href="./11-mlm-viz.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Visualizing Multivariate Models</span></a></li></ol></nav>
      <h1 class="title"><span id="sec-vis-mlm" class="quarto-section-identifier"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Visualizing Multivariate Models</span></span></h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Visualizing Multivariate Data and Models in R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/friendly/vis-MLM-book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Orienting Ideas</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Exploratory Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-multivariate_plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Plots of Multivariate Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-pca-biplot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Dimension Reduction</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Univariate Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-linear_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Overview of Linear models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-linear_models-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Plots for univariate response models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-lin-mod-topics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Topics in Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-collinearity-ridge.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Collinearity &amp; Ridge Regression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Multivariate Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-hotelling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Hotelling’s <span class="math inline">\(T^2\)</span></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-mlm-review.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Multivariate Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-mlm-viz.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Visualizing Multivariate Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-eqcov.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Visualizing Equality of Covariance Matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-infl-robust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Multiviate Influence and Robust Estimation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-case-studies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Case studies</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">End matter</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./91-colophon.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Colophon</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./95-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-he-framework" id="toc-sec-he-framework" class="nav-link active" data-scroll-target="#sec-he-framework"><span class="header-section-number">11.1</span> HE plot framework</a></li>
  <li>
<a href="#he-plot-construction" id="toc-he-plot-construction" class="nav-link" data-scroll-target="#he-plot-construction"><span class="header-section-number">11.2</span> HE plot construction</a>
  <ul class="collapse">
<li><a href="#iris-ex" id="toc-iris-ex" class="nav-link" data-scroll-target="#iris-ex"><span class="header-section-number">11.2.1</span> Example: Iris data</a></li>
  <li><a href="#sec-iris-mod" id="toc-sec-iris-mod" class="nav-link" data-scroll-target="#sec-iris-mod"><span class="header-section-number">11.2.2</span> MANOVA model</a></li>
  </ul>
</li>
  <li><a href="#he-plots" id="toc-he-plots" class="nav-link" data-scroll-target="#he-plots"><span class="header-section-number">11.3</span> HE plots</a></li>
  <li><a href="#significance-scaling" id="toc-significance-scaling" class="nav-link" data-scroll-target="#significance-scaling"><span class="header-section-number">11.4</span> Significance scaling</a></li>
  <li><a href="#visualizing-contrasts-and-linear-hypotheses" id="toc-visualizing-contrasts-and-linear-hypotheses" class="nav-link" data-scroll-target="#visualizing-contrasts-and-linear-hypotheses"><span class="header-section-number">11.5</span> Visualizing contrasts and linear hypotheses</a></li>
  <li><a href="#sec-HEplot-matrices" id="toc-sec-HEplot-matrices" class="nav-link" data-scroll-target="#sec-HEplot-matrices"><span class="header-section-number">11.6</span> HE plot matrices</a></li>
  <li>
<a href="#sec-candisc" id="toc-sec-candisc" class="nav-link" data-scroll-target="#sec-candisc"><span class="header-section-number">11.7</span> Low-D views: Canonical analysis</a>
  <ul class="collapse">
<li><a href="#coeficients" id="toc-coeficients" class="nav-link" data-scroll-target="#coeficients"><span class="header-section-number">11.7.1</span> Coeficients</a></li>
  <li><a href="#canonical-scores-plot" id="toc-canonical-scores-plot" class="nav-link" data-scroll-target="#canonical-scores-plot"><span class="header-section-number">11.7.2</span> Canonical scores plot</a></li>
  <li><a href="#canonical-he-plot" id="toc-canonical-he-plot" class="nav-link" data-scroll-target="#canonical-he-plot"><span class="header-section-number">11.7.3</span> Canonical HE plot</a></li>
  </ul>
</li>
  <li>
<a href="#quantitative-predictors-mmra" id="toc-quantitative-predictors-mmra" class="nav-link" data-scroll-target="#quantitative-predictors-mmra"><span class="header-section-number">11.8</span> Quantitative predictors: MMRA</a>
  <ul class="collapse">
<li><a href="#sec-NLSY-HE" id="toc-sec-NLSY-HE" class="nav-link" data-scroll-target="#sec-NLSY-HE"><span class="header-section-number">11.8.1</span> Example: NLSY data</a></li>
  </ul>
</li>
  <li>
<a href="#sec-cancor" id="toc-sec-cancor" class="nav-link" data-scroll-target="#sec-cancor"><span class="header-section-number">11.9</span> Canonical correlation analysis</a>
  <ul class="collapse">
<li><a href="#sec-schooldata-HE" id="toc-sec-schooldata-HE" class="nav-link" data-scroll-target="#sec-schooldata-HE"><span class="header-section-number">11.9.1</span> Example: School data</a></li>
  <li><a href="#canonical-analysis" id="toc-canonical-analysis" class="nav-link" data-scroll-target="#canonical-analysis"><span class="header-section-number">11.9.2</span> Canonical analysis</a></li>
  </ul>
</li>
  <li>
<a href="#mancova-models" id="toc-mancova-models" class="nav-link" data-scroll-target="#mancova-models"><span class="header-section-number">11.10</span> MANCOVA models</a>
  <ul class="collapse">
<li><a href="#example-rohwer-data" id="toc-example-rohwer-data" class="nav-link" data-scroll-target="#example-rohwer-data"><span class="header-section-number">11.10.1</span> Example: Rohwer data</a></li>
  </ul>
</li>
  <li><a href="#what-we-have-learned" id="toc-what-we-have-learned" class="nav-link" data-scroll-target="#what-we-have-learned"><span class="header-section-number">11.11</span> What We Have Learned</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/friendly/vis-MLM-book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><!--- For HTML Only ---><!-- \require{newcommand} --><!-- %\renewcommand*{\det}[1]{\mathrm{det} (#1)} --><!-- %\renewcommand*{\det}[1]{|#1|} --><!-- \newcommand{\sizedmat}[2]{\mathord{\mathop{\mat{#1}}\limits_{(#2)}}} --><!-- \newcommand*{\E}{\mathcal{E}} --><!-- Index generation --><!-- % R packages:  indexed under both package name and packages! --><!-- % data sets:  --><!-- % R stuff --><p>The methods discussed in <a href="10-mlm-review.html" class="quarto-xref"><span>Chapter 10</span></a> provide the basis for a rather complete multivariate analysis of traditional univariate methods for the same designs. You can carry out multiple regression, ANOVA, or indeed, any classical linear model with the standard collection of analysis tools you use for a single outcome variable, but naturally extended in most cases to having several outcomes to analyse together. The key points are:</p>
<ul>
<li><p>Everything you know about the usual univariate models– regression coefficients, main effects and contrasts for factors, interactions of model terms, … applies here.</p></li>
<li><p>By a rather clever design, called “matrix algebra” the separate univariate models can be combined, by turning vectors of responses, <span class="math inline">\(\mathbf{y}_1, \mathbf{y}_2, \dots\)</span> into a matrix <span class="math inline">\(\mathbf{Y}\)</span>. Bingo! We get a multivariate extension.</p></li>
<li><p>You can treat this as a collection of separate models, one for each response, because the coefficients are the same. But, the benefit of a multivariate approach is that you also get an overall multivariate test for each term in the model.</p></li>
</ul>
<p>As nice as these mathematical and statistical ideas might be, the fact that the analysis is conducted for the response variables <em>collectively</em>, means that it may be harder to interpret and explain what this means about the separate responses. Here’s where multivariate model visualization comes to the rescue!</p>
<ul>
<li><p>The tests of multivariate models, including multivariate analysis of variance (MANOVA) for group differences and multivariate multiple regression (MMRA) can be easily visualized by plots of a hypothesis (“H”) data ellipse for the fitted values, relative to the corresponding plot of the error ellipse (“E”) of the residuals, which I call the HE plot framework.</p></li>
<li><p>For more than a few response variables, these result can be projected onto a lower-dimensional “canonical” space providing an even simpler description. Vectors for the response variables in this space show how these relate to the canonical dimension, facilitating interpretation. This answers the objection of <span class="citation" data-cites="Huang2019">Huang (<a href="95-references.html#ref-Huang2019" role="doc-biblioref">2019</a>)</span> and others that multivariate models are difficult to understand because they are framed in terms of linear combinations of the the responses. The difficulty in understanding, I believe, can be cured by methods for visualizing what these linear combinations reflect in terms of the observed variables.</p></li>
</ul>
<!-- 
- refer to Huang: "_given the challenges and assumptions involved in appropriately performing and interpreting MANOVA results, researchers are better off using other less error-prone procedures._" 
--><p><strong>Packages</strong></p>
<p>In this chapter we use the following packages. I load them now.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-forge.r-project.org/projects/car/">car</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://friendly.github.io/heplots/">heplots</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/friendly/candisc/">candisc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sec-he-framework" class="level2" data-number="11.1"><h2 data-number="11.1" class="anchored" data-anchor-id="sec-he-framework">
<span class="header-section-number">11.1</span> HE plot framework</h2>
<p><a href="09-hotelling.html" class="quarto-xref"><span>Chapter 9</span></a> illustrated the basic ideas of the framework for visualizing multivariate linear models in the context of a simple two group design, using Hotelling’s <span class="math inline">\(T^2\)</span>. The main ideas were illustrated in <a href="09-hotelling.html#fig-HE-framework" class="quarto-xref">Figure&nbsp;<span>9.9</span></a>.</p>
<p>Having described the statistical ideas behind the MLM in <a href="10-mlm-review.html" class="quarto-xref"><span>Chapter 10</span></a>, we can proceed to extend this framework to larger designs. <a href="#fig-dogfood-quartet" class="quarto-xref">Figure&nbsp;<span>11.1</span></a> illustrates these ideas using the simple one-way MANOVA design of the dogfood data from <a href="10-mlm-review.html#sec-dogfood-data" class="quarto-xref"><span>Section 10.2.1</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-dogfood-quartet" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-dogfood-quartet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/dogfood-quartet.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-dogfood-quartet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.1: <strong>Dogfood quartet</strong>: Illustration of the conceptual ideas of the HE plot framework for the dogfood data. (a) Scatterplot of the data; (b) Summary using data ellipses; (c) HE plot shows the variation in the means in relation to pooled within group variance; (d) Transformation from data space to canonical space
</figcaption></figure>
</div>
</div>
</div>
<ul>
<li><p>In data space (a), each group is summarized by its <strong>data ellipse</strong> (b), representing the means and covariances.</p></li>
<li><p>Variation against the hypothesis of equal means can be seen by the <span class="math inline">\(\mathbf{H}\)</span> ellipse in the <strong>HE plot</strong> (c), representing the data ellipse of the fitted values. Error variance is shown in the <span class="math inline">\(\mathbf{E}\)</span> ellipse, representing the pooled within-group covariance matrix, <span class="math inline">\(\mathbf{S}_p\)</span> and the data ellipse of the residuals from the model. For the dogfood data, the group means have a negative relation: longer time to start eating is associated with a smaller amount eaten.</p></li>
<li><p>The MANOVA (or Hotelling’s <span class="math inline">\(T^2\)</span>) is formally equivalent to a <strong>discriminant analysis</strong>, predicting group membership from the response variables which can be seen in data space. (The main difference is emphasis and goals: MANOVA seeks to test differences among group means, while discriminant analysis aims at classification of the observations into groups.)</p></li>
<li><p>This effectively projects the <span class="math inline">\(p\)</span>-dimensional space of the predictors into the smaller <strong>canonical space</strong> (d) that shows the greatest differences among the groups. As in a biplot, vectors show the relations of the response variables with the canonical dimensions.</p></li>
</ul>
<!--

::: {.cell layout-align="center"}
::: {.cell-output-display}
![The Hypothesis Error plot framework for a two-group design. Above: Data ellipses can be summarized in an HE plot showing the pooled within-group error ($\mathbf{E}$) ellipse and the $\mathbf{H}$ 'ellipse' for the group means. Below: Observations projected on the line joining the means give discriminant scores which correpond to a one-dimensional canonical space, represented by a boxplot of their scores and arrows reflecting the variable weights.](images/HE-framework.png){#fig-HE-framework fig-align='center' width=100%}
:::
:::

--><!--
Having described the statistical ideas behind the MLM in @sec-mlm-review, we can proceed to
extend this framework to larger designs. A conceptual overview is shown in @fig-arcmanov1 for a one-way MANOVA
design with 8 groups.


::: {.cell layout-align="center"}
::: {.cell-output-display}
![Conceptual plots showing the essential ideas behind multivariate tests, in terms of the hypothesis ($\mathbf{H}$) and error ($\mathbf{E}$) matrices for a 1-way MANOVA design with two response variables, $Y_1$ and $Y_2$](images/arcmanov.png){#fig-arcmanov1 fig-align='center' width=100%}
:::
:::

--><p>For more complex models such as MANOVA with multiple factors or multivariate multivariate regression, there is one sum of squares and products matrix (SSP), and therefore one <span class="math inline">\(\mathbf{H}\)</span> ellipse for each term in the model. For example, in a two-way MANOVA design with the model formula <code>(y1, y2) ~ A + B + A*B</code> and equal sample sizes in the groups, the total sum of squares accounted for by the model is <span class="math display">\[\begin{aligned}
\mathbf{SSP}_{\text{Model}} &amp; = \mathbf{SSP}_{A} + \mathbf{SSP}_{B} + \mathbf{SSP}_{AB} \\
                            &amp; = \mathbf{H}_A + \mathbf{H}_B + \mathbf{H}_{AB} \:\: .
\end{aligned}\]</span></p>
</section><section id="he-plot-construction" class="level2" data-number="11.2"><h2 data-number="11.2" class="anchored" data-anchor-id="he-plot-construction">
<span class="header-section-number">11.2</span> HE plot construction</h2>
<p>The HE plot is constructed to allow a direct visualization of the “size” of hypothesized terms in a multivariate linear model in relation to unexplained error variation. These can be displayed in 2D or 3D plots, so I use the term “ellipsoid” below to cover all cases.</p>
<p>Error variation is represented by a standard 68% data ellipsoid of the <span class="math inline">\(\mathbf{E}\)</span> matrix of the residuals in <span class="math inline">\(\boldsymbol{\Large\varepsilon}\)</span>. This is divided by the residual degrees of freedom, so the size of <span class="math inline">\(\mathbf{E} / \text{df}_e\)</span> is analogous to a mean square error in univariate tests. The choice of 68% coverage allows you to ``read’’ the residual standard deviation as the half-length of the shadow of the <span class="math inline">\(\mathbf{E}\)</span> ellipsoid on any axis (see <a href="03-multivariate_plots.html#fig-galton-ellipse-r" class="quarto-xref">Figure&nbsp;<span>3.10</span></a>). The <span class="math inline">\(\mathbf{E}\)</span> ellipsoid is then translated to the overall (grand) means <span class="math inline">\(\bar{\mathbf{y}}\)</span> of the variables plotted, which allows us to show the means for factor levels on the same scale, facilitating interpretation. In the notation of <a href="03-multivariate_plots.html#eq-ellE" class="quarto-xref">Equation&nbsp;<span>3.2</span></a>, the error ellipsoid is given by <span id="eq-Ec"><span class="math display">\[
\mathcal{E}_c (\bar{\mathbf{y}}, \mathbf{E}) = \bar{\mathbf{y}} \; \oplus \; c\,\mathbf{E}^{1/2} \:\: ,
\tag{11.1}\]</span></span> where <span class="math inline">\(c = \sqrt{2 F_{2, n-2}^{0.68}}\)</span> for 2D plots and <span class="math inline">\(c = \sqrt{3 F_{3, n-3}^{0.68}}\)</span> for 3D for standard 68% coverage.</p>
<p>An ellipsoid representing variation in the means of a factor (or any other term reflected in a general linear hypothesis test, <a href="10-mlm-review.html#eq-hmat" class="quarto-xref">Equation&nbsp;<span>10.7</span></a>) in the <span class="math inline">\(\mathbf{H}\)</span> matrix is simply the data ellipse of the fitted values for that term.</p>
<p>Dividing the hypothesis matrix by the error degrees of freedom, giving <span class="math inline">\(\mathbf{H} / \text{df}_e\)</span>, puts this on the same scale as the ellipse. <!-- , as shown in the left panel of \figref{fig:heplot-iris1}. --> I refer to this as <em>effect size scaling</em>, because it is similar to an effect size index used in univariate models, e.g., <span class="math inline">\(ES = (\bar{y}_1 - \bar{y}_2) / s_e\)</span> in a two-group, univariate design.</p>
<!--
The notation $\bar{\mathbf{y}} \; \oplus$ in @eq-Ec means that the error ellipsoid is
shifted to be centered at the grand means of the response variables, where the $\mat{H}$
is also centered. This allows the HE plot to also show the group means, facilitating
interpretation.
-->
<section id="iris-ex" class="level3" data-number="11.2.1"><h3 data-number="11.2.1" class="anchored" data-anchor-id="iris-ex">
<span class="header-section-number">11.2.1</span> Example: Iris data</h3>
<p>Perhaps the most famous (or infamous) dataset in the history of multivariate data analysis is that of measurements on three species of Iris flowers collected by Edgar Anderson <span class="citation" data-cites="Anderson:35">(<a href="95-references.html#ref-Anderson:35" role="doc-biblioref">1935</a>)</span> in the Gaspé Peninsula of Québec, Canada. Anderson wanted to quantify the outward appearance (“morphology”: shape, structure, color, pattern, size) of species as a method to study variation within and between such groups. Although Anderson published in the obscure <em>Bulletin of the American Iris Society</em>, R. A. Fisher <span class="citation" data-cites="Fisher:1936">(<a href="95-references.html#ref-Fisher:1936" role="doc-biblioref">1936</a>)</span> saw this as a challenge and opportunity to introduce the method now called discriminant analysis—how to find a weighted composite of variables to best discriminate among existing groups.</p>
<!-- depending on use, make some of this a text box -->
<div class="callout callout-style-default callout-note callout-titled" title="History corner">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
History corner
</div>
</div>
<div class="callout-body-container callout-body">
<p>I said “infamous” above because Fisher published in the <em>Annals of Eugenics</em>, was an ardent eugenicist himself, and the work of eugenicists was often pervaded by prejudice against racial, ethnic and disabled groups. Through guilt by association, the Iris data, having mistakenly been called “Fisher’s Iris Data”, has become deprecated, even called “racist data”.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> The voices of the <em>Setosa</em>, <em>Versicolor</em> and <em>Virginica</em> of Gaspé protest: we don’t have a racist bone in out body and nor prejudice against any other species.</p>
<p><span class="citation" data-cites="Bodmer-etal-2021">Bodmer et al. (<a href="95-references.html#ref-Bodmer-etal-2021" role="doc-biblioref">2021</a>)</span> present a careful account of Fisher’s views on eugenics within the context of his time and his contributions to modern statistical theory and practice. Fisher’s views on race were largely formed by Darwin and Galton, but “nearly all of Fisher’s statements were about populations, groups of populations, or the human species as a whole”. Regardless, the <code>iris</code> data were Anderson’s and should not be blamed. After all, if Anderson had lent his car to Fisher, would the car be tainted by Fisher’s eugenicist leanings?</p>
</div>
</div>
<!--

::: {.cell layout-align="center"}

```{.r .cell-code}
knitr::include_graphics(here::here("images/iris-flowers-labeled.png"))
```

::: {.cell-output-display}
![Three species of irises in the Anderson/Fisher data set. _Source_: P. I. Adegbite [Iris Flower Classification]( https://peaceadegbite1.medium.com/iris-flower-classification-60790e9718a1)](images/iris-flowers-labeled.png){#fig-iris-flowers fig-align='center' width=100%}
:::
:::


@fig-iris-flowers shows photos of the three iris species, _Setosa_, _Versicolor_ and _Virginica_. 
Each flower has three sepals and three petals. The sepals have brightly colored central sections.
Anderson recorded measurements of the  length and width (in cm.) of the sepals and petals on 50 flowers of each type.
-->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-iris-diagram" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/iris-diagram.jpg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.2: Diagram of an iris flower showing the measurements of petal and sepal size. Each flower has three sepals and three alternating petals. The sepals have brightly colored central sections. <em>Source</em>: <span class="citation" data-cites="DeSilva2020">Gayan De Silva (<a href="95-references.html#ref-DeSilva2020" role="doc-biblioref">2020</a>)</span>
</figcaption></figure>
</div>
</div>
</div>
<p>So that we understand what the measurements represent, <a href="#fig-iris-diagram" class="quarto-xref">Figure&nbsp;<span>11.2</span></a> superposes labels on a typical iris flower. Sepals are like ostentatious petals, with attractive decorations in the central section. Length is the distance from the center to the tip and width is the transverse dimension.</p>
<p>As always, it is useful to start with overview displays to see the data. A scatterplot matrix (<a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a>) shows that <em>versicolor</em> and <em>virginica</em> are more similar to each other than either is to <em>setosa</em>, both in their pairwise means (<em>setosa</em> are smaller) and in the slopes of regression lines. Further, the ellipses suggest that the assumption of constant within-group covariance matrices is problematic: While the shapes and sizes of the concentration ellipses for <em>versicolor</em> and <em>virginica</em> are reasonably similar, the shapes and sizes of the ellipses for <em>setosa</em> are different from the other two.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris_colors</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"darkgreen"</span>, <span class="st">"brown4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/car/man/scatterplotMatrix.html">scatterplotMatrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Sepal.Length</span> <span class="op">+</span> <span class="va">Sepal.Width</span> <span class="op">+</span> </span>
<span>                    <span class="va">Petal.Length</span> <span class="op">+</span> <span class="va">Petal.Width</span> <span class="op">|</span> <span class="va">Species</span>,</span>
<span>  data <span class="op">=</span> <span class="va">iris</span>,</span>
<span>  col <span class="op">=</span> <span class="va">iris_colors</span>,</span>
<span>  pch <span class="op">=</span> <span class="fl">15</span><span class="op">:</span><span class="fl">17</span>,</span>
<span>  smooth<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>  regLine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ellipse<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>levels<span class="op">=</span><span class="fl">0.68</span>, fill.alpha<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  diagonal <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>coords <span class="op">=</span> <span class="st">"bottomleft"</span>, </span>
<span>                cex <span class="op">=</span> <span class="fl">1.3</span>, pt.cex <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-iris-spm" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-spm-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-spm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.3: Scatterplot matrix of the <code>iris</code> dataset. The species are summarized by 68% data ellipses and linear regression lines in each pairwise plot.
</figcaption></figure>
</div>
</div>
</div>
</section><section id="sec-iris-mod" class="level3" data-number="11.2.2"><h3 data-number="11.2.2" class="anchored" data-anchor-id="sec-iris-mod">
<span class="header-section-number">11.2.2</span> MANOVA model</h3>
<p>We proceed nevertheless to fit a multivariate one-way ANOVA model to the iris data. The MANOVA model for these data addresses the question: “Do the means of the <code>Species</code> differ significantly for the sepal and petal variables taken together?” <span class="math display">\[
\mathcal{H}_0 : \mathbf{\mu}_\textrm{setosa} = \mathbf{\mu}_\textrm{versicolor} = \mathbf{\mu}_\textrm{virginica}
\]</span></p>
<p>Because there are three species, the test involves <span class="math inline">\(s = \min(p, g-1) =2\)</span> degrees of freedom, and we are entitled to represent this by two 1-df contrasts, or sub-questions. From the separation among the groups shown in <a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a> (or more botanical knowledge), it makes sense to compare:</p>
<ul>
<li>Setosa vs.&nbsp;others: <span class="math inline">\(\mathbf{c}_1 = (1,\: -\frac12, \: -\frac12)\)</span>
</li>
<li>Versicolor vs.&nbsp;Virginica: : <span class="math inline">\(\mathbf{c}_1 = (0,\: 1, \: -1)\)</span>
</li>
</ul>
<p>You can do this by putting these vectors as columns in a matrix and assigning this to the <code><a href="https://rdrr.io/r/stats/contrasts.html">contrasts()</a></code> of <code>Species</code>. It is important to do this <em>before</em> fitting with <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code>, because the contrasts in effect determine how the <span class="math inline">\(\mathbf{X}\)</span> matrix is setup, and hence the names of the coefficients representing <code>Species</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span>,<span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span>,  </span>
<span>              <span class="fl">0</span>,   <span class="fl">1</span>,  <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">3</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/contrasts.html">contrasts</a></span><span class="op">(</span><span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">C</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/contrasts.html">contrasts</a></span><span class="op">(</span><span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span></span>
<span><span class="co">#&gt;            [,1] [,2]</span></span>
<span><span class="co">#&gt; setosa      1.0    0</span></span>
<span><span class="co">#&gt; versicolor -0.5    1</span></span>
<span><span class="co">#&gt; virginica  -0.5   -1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s fit the model. As you would expect from <a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a>, the differences among groups are highly significant.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris.mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">Sepal.Length</span>, <span class="va">Sepal.Width</span>, <span class="va">Petal.Length</span>, <span class="va">Petal.Width</span><span class="op">)</span> <span class="op">~</span></span>
<span>                 <span class="va">Species</span>, data<span class="op">=</span><span class="va">iris</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/car/man/Anova.html">Anova</a></span><span class="op">(</span><span class="va">iris.mod</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Type II MANOVA Tests: Pillai test statistic</span></span>
<span><span class="co">#&gt;         Df test stat approx F num Df den Df Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; Species  2      1.19     53.5      8    290 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As a quick follow-up, it is useful to examine the univariate tests for each of the iris variables, using <code>heplots::glance()</code> or <code><a href="https://friendly.github.io/heplots/reference/uniStats.html">heplots::uniStats()</a></code>. It is of interest that the univariate <span class="math inline">\(R^2\)</span> values are much larger for the petal variables than the sepal length and width.<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. For comparison, <code><a href="https://friendly.github.io/heplots/reference/etasq.html">heplots::etasq()</a></code> gives the overall <span class="math inline">\(\eta^2\)</span> proportion of variance accounted for in all responses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/glance.html">glance</a></span><span class="op">(</span><span class="va">iris.mod</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 4 × 8</span></span>
<span><span class="co">#&gt;   response     r.squared sigma fstatistic numdf dendf  p.value  nobs</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 Sepal.Length     0.619 0.515      119.      2   147 1.67e-31   150</span></span>
<span><span class="co">#&gt; 2 Sepal.Width      0.401 0.340       49.2     2   147 4.49e-17   150</span></span>
<span><span class="co">#&gt; 3 Petal.Length     0.941 0.430     1180.      2   147 2.86e-91   150</span></span>
<span><span class="co">#&gt; 4 Petal.Width      0.929 0.205      960.      2   147 4.17e-85   150</span></span>
<span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/etasq.html">etasq</a></span><span class="op">(</span><span class="va">iris.mod</span><span class="op">)</span></span>
<span><span class="co">#&gt;         eta^2</span></span>
<span><span class="co">#&gt; Species 0.596</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>But these statistics don’t help to understand <em>how</em> the species differ. For this, we turn to HE plots.</p>
</section></section><section id="he-plots" class="level2" data-number="11.3"><h2 data-number="11.3" class="anchored" data-anchor-id="he-plots">
<span class="header-section-number">11.3</span> HE plots</h2>
<p>The <code><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot()</a></code> function takes a <code>"mlm"</code> object and produces an HE plot for one pair of variables specified by the <code>variables</code> argument. By default, it plots the first two. <a href="#fig-iris-HE1" class="quarto-xref">Figure&nbsp;<span>11.4</span></a> shows the HE plots for the two sepal and the two petal variables.</p>
<!-- duplicated the code b/c it doesn't print from the fig-iris-HE1 chunk -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">iris.mod</span>, size <span class="op">=</span> <span class="st">"effect"</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">iris.mod</span>, size <span class="op">=</span> <span class="st">"effect"</span>, variables <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-iris-HE1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-HE1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-HE1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-HE1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.4: HE plots for the multivariate model <code>iris.mod</code>. The left panel shows the plot for the Sepal variables; the right panel plots the Petal variables.
</figcaption></figure>
</div>
</div>
</div>
<p>The interpretation of the plots in <a href="#fig-iris-HE1" class="quarto-xref">Figure&nbsp;<span>11.4</span></a> is as follows:</p>
<ul>
<li><p>For the Sepal variables, length and width are positively correlated <em>within</em> species (the <span class="math inline">\(\mathbf{E}\)</span> = “Error” ellipsoid). The means of the groups (the <span class="math inline">\(\mathbf{H}\)</span> = “Species” ellipsoid), however, are negatively correlated. This plot is the HE plot representation of the data shown in row 2, column 1 of <a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a>. It reflects the relative <strong>shape</strong> of the iris sepals: shorter and wider for <em>setosa</em> than the other two species.</p></li>
<li><p>For the Petal variables length and width are again positively correlated <em>within</em> species, but now the means of the groups are positively correlated: longer petals go with wider ones across species. This reflects the relative <strong>size</strong> of the iris petals. The analogous data plot appears in row 4, column 3 of <a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a>.</p></li>
</ul>
<!--
**Older stuff**
@fig-iris-HE0 shows two views of the relationship between sepal length and sepal width for the iris data. ...

::: {.cell layout-align="center"}

```{.r .cell-code}
op <- par(mar = c(4, 4, 1, 1) + .5,
          mfrow = c(1, 2))
col <-c("blue", "darkgreen", "brown")
clr <- c(col, "red")
covEllipses(cbind(Sepal.Length, Sepal.Width) ~ Species, data=iris,
      pooled = TRUE,
      fill=TRUE,
      fill.alpha = 0.1,
      lwd = 3,
      col = clr,
      cex = 1.5, cex.lab = 1.5,
      label.pos = c(3, 1, 3, 0),
      xlim = c(4,8), ylim = c(2,4))

iris.mod <- lm(cbind(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width) ~
                 Species, data=iris)
heplot(iris.mod, size = "effect",
       cex = 1.5, cex.lab = 1.5,
       fill=TRUE, fill.alpha=c(0.3,0.1),
       xlim = c(4,8), ylim = c(2,4))
par(op)
```
:::



::: {.cell layout-align="center"}

```{.r .cell-code}
knitr::include_graphics("images/iris-HE1.png")
```

::: {.cell-output-display}
![Iris data: Data ellipses and HE plot](images/iris-HE1.png){#fig-iris-HE0 fig-align='center' width=100%}
:::
:::

--></section><section id="significance-scaling" class="level2" data-number="11.4"><h2 data-number="11.4" class="anchored" data-anchor-id="significance-scaling">
<span class="header-section-number">11.4</span> Significance scaling</h2>
<p>The geometry of ellipsoids and multivariate tests allow us to go further with another re-scaling of the <span class="math inline">\(\mathbf{H}\)</span> ellipsoid that gives a <em>visual test of significance</em> for any term in a MLM. This is done simply by dividing <span class="math inline">\(\mathbf{H} / df_e\)</span> further by the <span class="math inline">\(\alpha\)</span>-critical value of the corresponding test statistic to show the strength of evidence against the null hypothesis.</p>
<p>Among the various multivariate test statistics, Roy’s maximum root test, based on the largest eigenvalue <span class="math inline">\(\lambda_1\)</span> of <span class="math inline">\(\mathbf{H} \mathbf{E}^{-1}\)</span>, gives <span class="math inline">\(\mathbf{H} / (\lambda_\alpha df_e)\)</span> which has the visual property that the scaled <span class="math inline">\(\mathbf{H}\)</span> ellipsoid will protrude <em>somewhere</em> outside the standard <span class="math inline">\(\mathbf{E}\)</span> ellipsoid if and only if Roy’s test is significant at significance level <span class="math inline">\(\alpha\)</span>. The critical value <span class="math inline">\(\lambda_\alpha\)</span> for Roy’s test is <span class="math display">\[
\lambda_\alpha = \left(\frac{\text{df}_1}{\text{df}_2}\right) \; F_{\text{df}_1, \text{df}_2}^{1-\alpha} \:\: ,
\]</span> where <span class="math inline">\(\text{df}_1 = \max(p, \text{df}_h)\)</span> and <span class="math inline">\(\text{df}_2 = \text{df}_h + \text{df}_e - \text{df}_1\)</span>.</p>
<p>For these data, the HE plot using significance scaling is shown in the right panel of <a href="#fig-iris-HE2" class="quarto-xref">Figure&nbsp;<span>11.5</span></a>. The left panel is the same as that shown for sepal width and length in <a href="#fig-iris-HE1" class="quarto-xref">Figure&nbsp;<span>11.4</span></a>, but with axis limits to make the two plots directly comparable.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-iris-HE2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-HE2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/iris-HE2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-HE2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.5: HE plots for sepal width and sepal length in the iris dataset. Left: <em>effect</em> scaling of the <span class="math inline">\(\mathbf{H}\)</span> matrix; right: <em>significance</em> scaling, where protrusion of <span class="math inline">\(\mathbf{H}\)</span> outside <span class="math inline">\(\mathbf{E}\)</span> indicates a significant effect by Roy’s test.
</figcaption></figure>
</div>
</div>
</div>
<p>You can interpret the plot using effect scaling to indicate that the overall “size” of variation of the group means is roughly the same as that of within-group variation for the two sepal variables. Significance scaling weights the evidence against the null hypothesis that a given effect is zero.</p>
</section><section id="visualizing-contrasts-and-linear-hypotheses" class="level2" data-number="11.5"><h2 data-number="11.5" class="anchored" data-anchor-id="visualizing-contrasts-and-linear-hypotheses">
<span class="header-section-number">11.5</span> Visualizing contrasts and linear hypotheses</h2>
<p>As described in <a href="10-mlm-review.html#sec-contrasts" class="quarto-xref"><span>Section 10.3.1</span></a>, tests of linear hypotheses and contrasts represented by the general linear test <span class="math inline">\(\mathcal{H}_0: \mathbf{C} \;\mathbf{B} = \mathbf{0}\)</span> provide a powerful way to probe the <em>specific</em> effects represented within the global null hypothesis, <span class="math inline">\(\mathcal{H}_0: \mathbf{B} = \mathbf{0}\)</span>, that all effects are zero.</p>
<p>In this example the contrasts <span class="math inline">\(\mathbf{c}_1\)</span> (<code>Species1</code>) and <span class="math inline">\(\mathbf{c}_2\)</span> (<code>Species2</code>) among the iris species are orthogonal, i.e., <span class="math inline">\(\mathbf{c}_1^\top \mathbf{c}_2 = 0\)</span>. Therefore, their tests are statistically independent, and their <span class="math inline">\(\mathbf{H}\)</span> matrices are additive. They fully decompose the general question of differences among the groups into two independent questions regarding the contrasts.</p>
<p><span id="eq-H-Species"><span class="math display">\[
\mathbf{H}_\text{Species} = \mathbf{H}_\text{Species1} + \mathbf{H}_\text{Species2}
\tag{11.2}\]</span></span></p>
<p><code><a href="https://rdrr.io/pkg/car/man/linearHypothesis.html">car::linearHypothesis()</a></code> is the means for testing these statistically, and <code><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot()</a></code> provides the way to show these tests visually. Using the contrasts set up in <a href="#sec-iris-mod" class="quarto-xref"><span>Section 11.2.2</span></a>, <span class="math inline">\(\mathbf{c}_1\)</span>, representing the difference between <em>setosa</em> and the other species is labeled <code>Species1</code> and the comparison of <em>versicolor</em> with <em>virginica</em> is <code>Species2</code>. The coefficients for these in <span class="math inline">\(\mathbf{B}\)</span> give the differences in the means. The line for <code>(Intercept)</code> gives grand means of the variables.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">iris.mod</span><span class="op">)</span></span>
<span><span class="co">#&gt;             Sepal.Length Sepal.Width Petal.Length Petal.Width</span></span>
<span><span class="co">#&gt; (Intercept)        5.843       3.057        3.758       1.199</span></span>
<span><span class="co">#&gt; Species1          -0.837       0.371       -2.296      -0.953</span></span>
<span><span class="co">#&gt; Species2          -0.326      -0.102       -0.646      -0.350</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Numerical tests of hypotheses using <code><a href="https://rdrr.io/pkg/car/man/linearHypothesis.html">linearHypothesis()</a></code> can be specified in a very general way: A matrix (or vector) <span class="math inline">\(\mathbf{C}\)</span> giving linear combinations of coefficients by rows, or a character vector giving the hypothesis in symbolic form. A character variable or vector tests whether the named coefficients are different from zero for all responses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/car/man/linearHypothesis.html">linearHypothesis</a></span><span class="op">(</span><span class="va">iris.mod</span>, <span class="st">"Species1"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span>SSP<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Multivariate Tests: </span></span>
<span><span class="co">#&gt;                  Df test stat approx F num Df den Df Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; Pillai            1      0.97     1064      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Wilks             1      0.03     1064      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Hotelling-Lawley  1     29.55     1064      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Roy               1     29.55     1064      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/car/man/linearHypothesis.html">linearHypothesis</a></span><span class="op">(</span><span class="va">iris.mod</span>, <span class="st">"Species2"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span>SSP<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Multivariate Tests: </span></span>
<span><span class="co">#&gt;                  Df test stat approx F num Df den Df Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; Pillai            1     0.745      105      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Wilks             1     0.255      105      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Hotelling-Lawley  1     2.925      105      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; Roy               1     2.925      105      4    144 &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The various test statistics are all equivalent here—they give the same <span class="math inline">\(F\)</span> statistics— because they they have 1 degree of freedom.</p>
<p>In passing, from <a href="#eq-H-Species" class="quarto-xref">Equation&nbsp;<span>11.2</span></a>, note that the <em>joint</em> test of these contrasts is exactly equivalent to the overall test of <code>Species</code> (results not shown).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/car/man/linearHypothesis.html">linearHypothesis</a></span><span class="op">(</span><span class="va">iris.mod</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Species1"</span>, <span class="st">"Species2"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can show these contrasts in an HE plot by supplying a named list for the <code>hypotheses</code> argument. The names are used as labels in the plot. In the case of a 1-df multivariate test, the <span class="math inline">\(\mathbf{H}\)</span> ellipses plot as a degenerate line.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hyp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"S:Vv"</span> <span class="op">=</span> <span class="st">"Species1"</span>, <span class="st">"V:v"</span> <span class="op">=</span> <span class="st">"Species2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">iris.mod</span>, hypotheses<span class="op">=</span><span class="va">hyp</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"darkgreen"</span>, <span class="st">"darkgreen"</span><span class="op">)</span>,</span>
<span>       lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, label.pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.4</span>, <span class="fl">4.6</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-iris-contrasts" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-contrasts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-contrasts-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:60.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-contrasts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.6: HE plot for sepal length and width in the <code>iris</code> data showing the tests of the two contrasts
</figcaption></figure>
</div>
</div>
</div>
<p>This HE plot shows that, for the two sepal variables, the greatest between-species variation is accounted for by the contrast (<code>S:Vv</code>) between <em>setosa</em> and the others, for which the effect is very large in relation to error variation. The second contrast (<code>V:v</code>), between the <em>versicolor</em> and <em>virginica</em> species is relatively smaller, but still explains significant variation of the sepal variables among the species.</p>
<p>The directions of these hypotheses in a given plot show <em>how</em> the group means differ in terms of a given contrast.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> For example, the contrast <code>S:Vv</code> is the line that separates <em>setosa</em> from the others and indicates that <em>setosa</em> flowers have shorter but wider sepals.</p>
</section><section id="sec-HEplot-matrices" class="level2" data-number="11.6"><h2 data-number="11.6" class="anchored" data-anchor-id="sec-HEplot-matrices">
<span class="header-section-number">11.6</span> HE plot matrices</h2>
<p>In base R graphics, 2D scatterplots are extended to all pairwise views of multivariate data with a <code><a href="https://rdrr.io/r/graphics/pairs.html">pairs()</a></code> method. For multivariate linear models, the <span style="color: brown;"><strong>heplots</strong></span> defines a <code>pairs.mlm()</code> method to display HE plots for all pairs of the response variables.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html">pairs</a></span><span class="op">(</span><span class="va">iris.mod</span>,</span>
<span>      fill<span class="op">=</span><span class="cn">TRUE</span>, fill.alpha<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-iris-pairs" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-pairs-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.7: All pairwise HE plots for the iris data.
</figcaption></figure>
</div>
</div>
</div>
<p><a href="#fig-iris-pairs" class="quarto-xref">Figure&nbsp;<span>11.7</span></a> provides a fairly complete visualization of the results of the multivariate tests and answers the question: <strong>how</strong> do the species differ? Sepal length and the two petal variables have the group means nearly perfectly correlated, in the order <em>setosa</em> &lt; <em>versicolor</em> &lt; <em>virginica</em>. For Sepal width, however, <em>setosa</em> has the largest mean, and so the <span class="math inline">\(\mathbf{H}\)</span> ellipses show a negative correlation in the second row and column.</p>
</section><section id="sec-candisc" class="level2" data-number="11.7"><h2 data-number="11.7" class="anchored" data-anchor-id="sec-candisc">
<span class="header-section-number">11.7</span> Low-D views: Canonical analysis</h2>
<p>The HE plot framework so far provides views of all the effects in a MLM in <em>variable</em> space. We can view this in 2D for selected pairs of response variables, or for all pairwise views in scatterplot matrix format. There is also an <code><a href="https://friendly.github.io/heplots/reference/heplot3d.html">heplot3d()</a></code> function giving plots for three response variables together. The 3D plots are interactive, in that they can be rotated and zoomed by mouse control, and dynamic, in that they can be made to spin and saved as movies. To save space, these plots are not shown here.</p>
<p>However in a one-way MANOVA design with more than response three variables, it is difficult to visualize how the groups vary on <em>all</em> responses together, and how the different variables contribute to discrimination among groups. In this situation, <strong>canonical discriminant analysis</strong> (CDA) is often used, to provide a low-D visualization of between-group variation. When the predictors are also continuous, the analogous term is <strong>canonical correlation analysis</strong> (CCA). The advantage here is that we can also show the relations of the response variables to these dimensions, similar to a biplot (<a href="04-pca-biplot.html#sec-biplot" class="quarto-xref"><span>Section 4.3</span></a>) for a PCA of purely quantitative variables.</p>
<p>The key to this is the eigenvalue decomposition, <span class="math inline">\(\mathbf{H}\mathbf{E}^{-1} \lambda_i = \lambda_i \mathbf{v}_i\)</span> (<a href="10-mlm-review.html#eq-he-eigen" class="quarto-xref">Equation&nbsp;<span>10.5</span></a>) of <span class="math inline">\(\mathbf{H}\)</span> relative to <span class="math inline">\(\mathbf{E}\)</span>. The eigenvalues, <span class="math inline">\(\lambda_i\)</span>, give the “size” of each <span class="math inline">\(s\)</span> orthogonal dimensions on which the multivariate tests are based. But the corresponding eigenvectors, <span class="math inline">\(\mathbf{v}_i\)</span>, give the <em>weights</em> for the response variables in <span class="math inline">\(s\)</span> linear combinations that maximally discriminate among the groups or equivalently maximize the (canonical) <span class="math inline">\(R^2\)</span> of a linear combination of the predictor <span class="math inline">\(\mathbf{X}\)</span>s with a linear combination of the response <span class="math inline">\(\mathbf{Y}\)</span>s.</p>
<p>Thus, CDA amounts to a transformation of the <span class="math inline">\(p\)</span> responses, <span class="math inline">\(\mathbf{Y}_{n \times p}\)</span> into the canonical space, <span class="math display">\[
\mathbf{Z}_{n \times s} = \mathbf{Y} \; \mathbf{E}^{-1/2} \; \mathbf{V} \;,
\]</span> where <span class="math inline">\(\mathbf{V}\)</span> contains the eigenvectors of <span class="math inline">\(\mathbf{H}\mathbf{E}^{-1}\)</span> and <span class="math inline">\(s=\min ( p, df_h )\)</span>. It is well-known (e.g., <span class="citation" data-cites="Gittins:85">Gittins (<a href="95-references.html#ref-Gittins:85" role="doc-biblioref">1985</a>)</span>) that <em>canonical discriminant plots</em> of the first two (or three, in 3D) columns of <span class="math inline">\(\mathbf{Z}\)</span> corresponding to the largest canonical correlations provide an optimal low-D display of the variation between groups relative to variation within groups.</p>
<p>Canonical discriminant analysis is typically carried out in conjunction with a one-way MANOVA design. The <span style="color: brown;"><strong>candisc</strong></span> package <span class="citation" data-cites="R-candisc">(<a href="95-references.html#ref-R-candisc" role="doc-biblioref">Friendly &amp; Fox, 2025</a>)</span> generalizes this to multi-factor designs in the <code><a href="https://friendly.github.io/candisc/reference/candisc.html">candisc()</a></code> function. For any given term in a <code>"mlm"</code>, the <em>generalized canonical discriminant analysis</em> amounts to a standard discriminant analysis based on the <span class="math inline">\(\mathbf{H}\)</span> matrix for <em>that</em> term in relation to the full-model <span class="math inline">\(\mathbf{E}\)</span> matrix.<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
<p>Tests based on the eigenvalues <span class="math inline">\(\lambda_i\)</span>, initially stated by <span class="citation" data-cites="Bartlett1938">Bartlett (<a href="95-references.html#ref-Bartlett1938" role="doc-biblioref">1938</a>)</span>, use Wilks’ <span class="math inline">\(\Lambda\)</span> likelihood ratio tests of these. This allow you to determine the number of significant canonical dimensions, or the number of different aspects to consider for the relations between the responses and predictors. These take the form of <em>sequential</em> global tests of the hypothesis that the canonical correlation in the current row and all that follow are zero. Thus, if you find The canonical <span class="math inline">\(R^2\)</span>, <code>CanRsq</code>, gives the R-squared value of fitting the <span class="math inline">\(i\)</span>th response canonical variate to the corresponding <span class="math inline">\(i\)</span>th canonical variate for the predictors.</p>
<p>For the iris data, we get the following printed summary:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris.can</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://friendly.github.io/candisc/reference/candisc.html">candisc</a></span><span class="op">(</span><span class="va">iris.mod</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Canonical Discriminant Analysis for Species:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   CanRsq Eigenvalue Difference Percent Cumulative</span></span>
<span><span class="co">#&gt; 1  0.970     32.192       31.9  99.121       99.1</span></span>
<span><span class="co">#&gt; 2  0.222      0.285       31.9   0.879      100.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Test of H0: The canonical correlations in the </span></span>
<span><span class="co">#&gt; current row and all that follow are zero</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   LR test stat approx F numDF denDF Pr(&gt; F)    </span></span>
<span><span class="co">#&gt; 1        0.023    199.1     8   288 &lt; 2e-16 ***</span></span>
<span><span class="co">#&gt; 2        0.778     13.8     3   145 5.8e-08 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This analysis shows a very simple result: The differences among the iris species can be nearly entirely accounted for by the first canonical dimension (99.1%). Interestingly, the second dimension is also highly significant, even though it accounts for only 0.88%.</p>
<section id="coeficients" class="level3" data-number="11.7.1"><h3 data-number="11.7.1" class="anchored" data-anchor-id="coeficients">
<span class="header-section-number">11.7.1</span> Coeficients</h3>
<p>The <code><a href="https://rdrr.io/r/stats/coef.html">coef()</a></code> method for “candisc” objects returns a matrix of weights for the response variables in the canonical dimensions. By default, these are given for the response variables <em>standardized</em> to <span class="math inline">\(\bar{y}=0\)</span> and <span class="math inline">\(s^2_y = 1\)</span>. The <code>type</code> argument also allows for raw score weights (<code>type = "raw"</code>) used to compute the observation scores on <code>Can1</code>, <code>Can2</code>, … . <code>type = "structure"</code> gives the canonical structure coefficients, which are the correlations between each response and the canonical scores.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">iris.can</span>, type <span class="op">=</span> <span class="st">"std"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                Can1    Can2</span></span>
<span><span class="co">#&gt; Sepal.Length  0.427  0.0124</span></span>
<span><span class="co">#&gt; Sepal.Width   0.521  0.7353</span></span>
<span><span class="co">#&gt; Petal.Length -0.947 -0.4010</span></span>
<span><span class="co">#&gt; Petal.Width  -0.575  0.5810</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">iris.can</span>, type <span class="op">=</span> <span class="st">"structure"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                Can1  Can2</span></span>
<span><span class="co">#&gt; Sepal.Length -0.792 0.218</span></span>
<span><span class="co">#&gt; Sepal.Width   0.531 0.758</span></span>
<span><span class="co">#&gt; Petal.Length -0.985 0.046</span></span>
<span><span class="co">#&gt; Petal.Width  -0.973 0.223</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Thus, the two weighted sums for the canonical variates, using mean-centered standardized scores, can be shown as follows:</p>
<div class="cell" data-layout-align="center">
<p>Can1 = 0.43 x S.len + 0.52 x S.wid + -0.95 x P.len + -0.58 x P.wid Can2 = 0.01 x S.len + 0.74 x S.wid + -0.4 x P.len + 0.58 x P.wid</p>
</div>
</section><section id="canonical-scores-plot" class="level3" data-number="11.7.2"><h3 data-number="11.7.2" class="anchored" data-anchor-id="canonical-scores-plot">
<span class="header-section-number">11.7.2</span> Canonical scores plot</h3>
<p>The <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> method for <code>"candisc"</code> objects gives a plot of these observation scores. <code>ellipse=TRUE</code> overlays this with their standard data ellipses for each species, as shown in <a href="#fig-iris-candisc" class="quarto-xref">Figure&nbsp;<span>11.8</span></a>. The response variables are shown as vectors, using the structure coefficients, as in a biplot. Thus, the relative size of the projection of these vectors on the canonical axes reflects the correlation of the observed response on the canonical dimension. For ease of interpretation I flipped the sign of the first canonical dimension, so that the positive <code>Can1</code> direction corresponds to larger flowers.</p>
<p><strong>TODO</strong>: <code>var.labels</code> not a graphical parameter</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span><span class="op">(</span><span class="st">"\\."</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">iris.can</span>,</span>
<span>     var.labels <span class="op">=</span> <span class="va">vars</span>,</span>
<span>     var.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>     var.lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     ellipse<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>     scale <span class="op">=</span> <span class="fl">9</span>,</span>
<span>     col <span class="op">=</span> <span class="va">iris_colors</span>,</span>
<span>     pch <span class="op">=</span> <span class="fl">15</span><span class="op">:</span><span class="fl">17</span>,</span>
<span>     cex <span class="op">=</span> <span class="fl">0.7</span>, var.cex <span class="op">=</span> <span class="fl">1.25</span>,</span>
<span>     rev.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>     cex.lab <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-iris-candisc" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-candisc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-candisc-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-candisc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.8: Plot of canonical scores for the iris data.
</figcaption></figure>
</div>
</div>
</div>
<p>The interpretation of this plot is simple: in canonical space, variation of the means for the iris species is essentially one-dimensional (99.1% of the effect of <code>Species</code>), and this dimension corresponds to overall size of the iris flowers. All variables except for <code>Sepal.Width</code> are positively aligned with this axis, but the two petal variables show the greatest discrimination. The negative direction for <code>Sepal.Width</code> reflects the pattern seen in <a href="#fig-iris-pairs" class="quarto-xref">Figure&nbsp;<span>11.7</span></a>, where <em>setosa</em> have wider sepals.</p>
<p>For the second dimension, look at the projections of the variable vectors on the <code>Can2</code> axis. All are positive, but this is dominated by <code>Sepal.Width</code>. We could call this a flower shape dimension.</p>
</section><section id="canonical-he-plot" class="level3" data-number="11.7.3"><h3 data-number="11.7.3" class="anchored" data-anchor-id="canonical-he-plot">
<span class="header-section-number">11.7.3</span> Canonical HE plot</h3>
<p>For a one-way design, the <em>canonical HE plot</em> is simply the HE plot of the canonical scores in the analogous MLM model that substitutes <span class="math inline">\(\mathbf{Z}\)</span> for <span class="math inline">\(\mathbf{Y}\)</span>. In effect, it is a more compact visual summary of the plot shown in <a href="#fig-iris-candisc" class="quarto-xref">Figure&nbsp;<span>11.8</span></a>.</p>
<p>This is shown in <a href="#fig-iris-HEcan" class="quarto-xref">Figure&nbsp;<span>11.9</span></a> for the <code>iris</code> data. In canonical space, the residuals are always uncorrelated, so the <span class="math inline">\(\mathbf{E}\)</span> ellipse plots as a circle. The <span class="math inline">\(\mathbf{H}\)</span> ellipse for <code>Species</code> here reflects a data ellipse for the fitted values— group means— shown as labeled points in the plot. The differences among <code>Species</code> are so large that this plot uses <code>size = "effect"</code> scaling, making the axes comparable to those in <a href="#fig-iris-candisc" class="quarto-xref">Figure&nbsp;<span>11.8</span></a>.<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a></p>
<p>The vectors for each predictor are the same structure coefficients as in the ordinary canonical plot. They can again be reflected for interpretation and scaled in length to fill the plot window.</p>
<!-- now depends on candisc 0.9.1 -->
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">iris.can</span>,</span>
<span>       size <span class="op">=</span> <span class="st">"effect"</span>,</span>
<span>       scale <span class="op">=</span> <span class="fl">8</span>,</span>
<span>       var.labels <span class="op">=</span> <span class="va">vars</span>,</span>
<span>       var.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>       var.lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>       rev.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-iris-HEcan" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-iris-HEcan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-iris-HEcan-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iris-HEcan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.9: Canonical HE plot for the iris data.
</figcaption></figure>
</div>
</div>
</div>
<p>The collection of plots shown for the iris data here can be seen as progressive visual summaries of the data:</p>
<ul>
<li>The scatterplot matrix in <a href="#fig-iris-spm" class="quarto-xref">Figure&nbsp;<span>11.3</span></a> shows the iris flowers in the data space of the sepal and petal variables.</li>
<li>Canonical analysis substitutes for these the two linear combinations reflected in <code>Can1</code> and <code>Can2</code>. The plot in <a href="#fig-iris-candisc" class="quarto-xref">Figure&nbsp;<span>11.8</span></a> portrays <em>exactly</em> the same relations among the species, but in the reduced canonical space of only two dimensions.</li>
<li>The HE plot version, shown in <a href="#fig-iris-HEcan" class="quarto-xref">Figure&nbsp;<span>11.9</span></a> summarizes the separate data ellipses for the species with pooled, within-group variance of the <span class="math inline">\(\mathbf{E}\)</span> matrix for the canonical variables, which are always uncorrelated. The variation among the group means is reflected in the size and shape of the ellipse for the <span class="math inline">\(\mathbf{E}\)</span> matrix.</li>
</ul></section></section><section id="quantitative-predictors-mmra" class="level2" data-number="11.8"><h2 data-number="11.8" class="anchored" data-anchor-id="quantitative-predictors-mmra">
<span class="header-section-number">11.8</span> Quantitative predictors: MMRA</h2>
<p>The ideas behind HE plots extend naturally to multivariate multiple regression (MMRA). <!--
and multivariate analysis of covariance (MANCOVA). In MMRA designs, the $\mathbf{X}$ matrix contains only quantitative predictors, while in MANCOVA designs, it contains a mixture of factors and quantitative predictors (covariates), but typically there is just one “group” factor.

In the MANCOVA case, there is often a subtle difference in emphasis: true MANCOVA analyses focus on the differences among groups defined by the factors, adjusting for (or controlling for) the quantitative covariates. Analyses concerned with homogeneity of regression focus on quantitative predictors and attempt to test whether the regression relations are the same for all groups defined by the factors.
--> A purely visual feature of HE plots in these cases is that the <span class="math inline">\(\mathbf{H}\)</span> ellipse for a quantitative predictor with 1 df appears as a degenerate line. But consequently, the angles between these for different predictors has a simple interpretation as as the correlation between their predicted effects. Moreover, it is easy to show visual overall tests of <em>joint</em> linear hypotheses for two or more predictors together.</p>
<p><strong>TODO</strong>: Use these examples below</p>
<ul>
<li>
<code><a href="https://friendly.github.io/heplots/reference/NLSY.html">heplots::NLSY</a></code>: National Longitudinal Survey of Youth Data</li>
<li>
<code><a href="https://friendly.github.io/heplots/reference/schooldata.html">heplots::schooldata</a></code>: School Data -&gt; <code>R/schooldata-ex.R</code>
</li>
<li>
<code><a href="https://friendly.github.io/heplots/reference/Hernior.html">heplots::Hernior</a></code>: Recovery from Elective Herniorrhaphy -&gt; HE_mmra vignette</li>
</ul>
<section id="sec-NLSY-HE" class="level3" data-number="11.8.1"><h3 data-number="11.8.1" class="anchored" data-anchor-id="sec-NLSY-HE">
<span class="header-section-number">11.8.1</span> Example: NLSY data</h3>
<!-- **TODO**: The HE plots are nicer w/o using log(income) -->
<p>Here I’ll continue the analysis of the NLSY data from <a href="10-mlm-review.html#sec-NLSY-mmra" class="quarto-xref"><span>Section 10.5.1</span></a>. In the model <code>NLSY.mod1</code> I used only father’s income and education to predict scores in reading and math, and both of these demographic variables were highly significant.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">NLSY</span>, package <span class="op">=</span> <span class="st">"heplots"</span><span class="op">)</span></span>
<span><span class="va">NLSY.mod1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">read</span>, <span class="va">math</span><span class="op">)</span> <span class="op">~</span> <span class="va">income</span> <span class="op">+</span> <span class="va">educ</span>, </span>
<span>                data <span class="op">=</span> <span class="va">NLSY</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">NLSY.mod1</span>, </span>
<span>  fill<span class="op">=</span><span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>  cex <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  lwd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  label.pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"bottom"</span>, <span class="st">"top"</span>, <span class="st">"top"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-NLSY-heplot1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-NLSY-heplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-NLSY-heplot1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-NLSY-heplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.10: HE plot for the simple model for the NLSY data fitting reading and math scores from income and education.
</figcaption></figure>
</div>
</div>
</div>
<p>Fathers income and education are positively correlated in their effects on the outcome scores. From the angles in the plot, income is most related to the math score, while education is related to both, but slightly more to the reading score.</p>
<p>The overall joint test for both predictors can then visualized as the test of the linear hypothesis <span class="math inline">\(\mathcal{H}_0 : \mathbf{B} = [\boldsymbol{\beta}_\text{income}, \boldsymbol{\beta}_\text{educ}] = \mathbf{0}\)</span>. For <code><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot()</a></code>, we specify the names of the coefficients to be tested with the <code>hypotheses</code> argument.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">coefs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">NLSY.mod1</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "income" "educ"</span></span>
<span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">NLSY.mod1</span>, </span>
<span>       hypotheses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Overall"</span> <span class="op">=</span> <span class="va">coefs</span><span class="op">)</span>,</span>
<span>       fill<span class="op">=</span><span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>       cex <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       lwd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>       label.pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"bottom"</span>, <span class="st">"top"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-NLSY-heplot2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-NLSY-heplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-NLSY-heplot2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-NLSY-heplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.11: HE plot adding the <span class="math inline">\(\mathbf{H}\)</span> ellipse for the overall test that both predictors have no effect on the outcome scores.
</figcaption></figure>
</div>
</div>
</div>
</section></section><section id="sec-cancor" class="level2" data-number="11.9"><h2 data-number="11.9" class="anchored" data-anchor-id="sec-cancor">
<span class="header-section-number">11.9</span> Canonical correlation analysis</h2>
<p>Just as we saw for MANOVA designs, a canonical analysis for multivariate regression involves finding a low-D view of the relations between predictors and outcomes that maximally explains their relations in terms of linear combinations of each. That is, the goal is to find weights for one set of variables, say <span class="math inline">\(\mathbf{X}\)</span> <em>not</em> to predict each of the other set <span class="math inline">\(\mathbf{Y} =[\mathbf{y}_1, \mathbf{y}_2, \dots]\)</span> <em>individually</em>, but rather to also find weights for the <span class="math inline">\(\mathbf{y}\)</span>s which is most highly correlated with the linear combination of the <span class="math inline">\(\mathbf{x}\)</span>s.</p>
<p>In this sense, canonical correlation analysis (CCA) is <em>symmetric</em> in the <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> variables: the <span class="math inline">\(y\)</span> set is not considered responses. Rather the goal is simply to explain the correlations between the two sets. For a thorough treatment of this topic, see <span class="citation" data-cites="Gittins:85">Gittins (<a href="95-references.html#ref-Gittins:85" role="doc-biblioref">1985</a>)</span>.</p>
<p>Geometrically, these linear combinations are vectors representing projections in the observation space of the <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> variables, and CCA can also be thought of as minimizing the angle between these vectors or maximizing the cosine of this angle. This is illustrated in <a href="#fig-cancor-diagram" class="quarto-xref">Figure&nbsp;<span>11.12</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-cancor-diagram" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-cancor-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/cancor-diagram-udi.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-cancor-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.12: Diagram illustrating canonical correlation. For two <span class="math inline">\(y\)</span> variables, all linear combinations are vectors in their plane, and similarly for the <span class="math inline">\(x\)</span> variables. Maximizing the correlation between linear combinations of each is equivalent to making the angle <span class="math inline">\(\phi\)</span> between them as small as possible, or maximizing <span class="math inline">\(\cos({\theta})\)</span>, shown in the diagram at the right. The thick grey arrow indignates that the two planes should be overlaid at a common origin. <em>Source</em>: Re-drawn by Udi Alter following a Cross-Validated discussion by user ‘ttnphns’, https://bit.ly/4dgq2cp
</figcaption></figure>
</div>
</div>
</div>
<p>Specifically, we want to find one set of weights <span class="math inline">\(\mathbf{a}_1\)</span> for the <span class="math inline">\(x\)</span> variables and another for the <span class="math inline">\(y\)</span> variables to give the linear combinations <span class="math inline">\(\mathbf{u}_1\)</span> and <span class="math inline">\(\mathbf{v}_1\)</span>,</p>
<p><span class="math display">\[\begin{aligned}
\mathbf{u}_1 &amp; = \mathbf{X} \ \mathbf{a}_1 = a_{11} \mathbf{x}_1 + a_{12} \mathbf{x}_2 + \cdots + a_{11} \mathbf{x}_q \\
\mathbf{v}_1 &amp; = \mathbf{Y} \ \mathbf{b}_1 = b_{11} \mathbf{y}_1 + b_{12} \mathbf{y}_1 + \cdots + b_{11} \mathbf{y}_p \; ,
\end{aligned}\]</span></p>
<p>such that the correlation <span class="math inline">\(\rho_1 = \textrm{corr}(\mathbf{u}_1, \mathbf{v}_1)\)</span> is maximized, or equivalently, minimizing the angle between them.</p>
<p>Using <span class="math inline">\(\mathbf{S}_{xx}\)</span>, <span class="math inline">\(\mathbf{S}_{yy}\)</span> to represent the covariance matrices of the <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> variables, and <span class="math inline">\(\mathbf{S}_{xy}\)</span> for the cross-covariances between the two sets, the correlation between the linear combinations of each can be expressed as</p>
<p><span class="math display">\[\begin{aligned}
\rho_1 &amp; = \textrm{corr}(\mathbf{u}_1, \mathbf{v}_1)
         = \textrm{corr}(\mathbf{X} \ \mathbf{a}_1, \mathbf{Y} \ \mathbf{b}_1) \\
       &amp; = \frac{\mathbf{a}_1^\top \ \mathbf{S}_{xy} \ \mathbf{b}_1 }{\sqrt{\mathbf{a}_1^\top \ \mathbf{S}_{xx} \  \mathbf{a}_1 } \sqrt{\mathbf{b}_1^\top \ \mathbf{S}_{yy} \ \mathbf{b}_1 }}
\end{aligned}\]</span></p>
<p>But, the <span class="math inline">\(y\)</span> variables lie in a <span class="math inline">\(p\)</span>-dimensional (observation) space, and the <span class="math inline">\(x\)</span> in <span class="math inline">\(q\)</span> dimensions, so what they have common is a space of <span class="math inline">\(s = \min(p, q)\)</span> dimensions. Therefore, we can find additional pairs of canonical variables,</p>
<p><span class="math display">\[\begin{aligned}
\mathbf{u}_2 = \mathbf{X} \ \mathbf{a}_2 &amp; \quad\quad \mathbf{v}_2 = \mathbf{Y} \ \mathbf{b}_2 \\
                                         &amp; \vdots \\
\mathbf{u}_s = \mathbf{X} \ \mathbf{a}_s &amp; \quad\quad \mathbf{v}_s = \mathbf{Y} \ \mathbf{b}_s \\
\end{aligned}\]</span></p>
<p>such that each pair <span class="math inline">\((\mathbf{u}_i, \mathbf{v}_i)\)</span> has the maximum possible correlation and all distinct pairs are uncorrelated:</p>
<p><span class="math display">\[\begin{aligned}
\rho_i &amp; =\max _{\mathbf{a}_i, \mathbf{b}_i}\left\{\mathbf{u}_i^{\top} \mathbf{v}_i\right\} = \\
\left\|\mathbf{u}_i\right\| &amp; =1, \quad\left\|\mathbf{v}_i\right\|=1, \\
\mathbf{u}_i^{{\top}} \mathbf{u}_j &amp; =0, \quad \mathbf{v}_i^{\top} \mathbf{v}_j=0 \quad\quad \forall j \neq i: i, j \in\{1,2, \ldots, s\} \ .
\end{aligned}\]</span></p>
<p>In words, the correlations among canonical variables are zero except when when they are associated with the same canonical correlation or the weights <span class="math inline">\((\mathbf{a}_i, \mathbf{b}_i)\)</span> for the same pair. Alternatively, all <span class="math inline">\(p \times q\)</span> correlations the variables in <span class="math inline">\(\mathbf{Y}\)</span> and <span class="math inline">\(\mathbf{X}\)</span> are fully summarized in the <span class="math inline">\(s = \min(p, q)\)</span> canonical correlations <span class="math inline">\(\rho_i\)</span> for <span class="math inline">\(i = 1, 2, \dots, s\)</span>.</p>
<p>The solution, developed by <span class="citation" data-cites="Hotelling1936">Hotelling (<a href="95-references.html#ref-Hotelling1936" role="doc-biblioref">1936</a>)</span>, is a form of a generalized eigenvalue problem, that can be stated in two equivalent ways,</p>
<!-- $$ -->
<!-- \left(\mathbf{S}_{y y}^{-1} \ \mathbf{S}_{y x} \ \mathbf{S}_{x x}^{-1} \ \mathbf{S}_{x y}-\rho^2 \ \mathbf{I}\right) \mathbf{v} = \mathbf{0} -->
<!-- $$ -->
<p><span class="math display">\[
\begin{aligned}
&amp; \left(\mathbf{S}_{y x} \ \mathbf{S}_{x x}^{-1} \ \mathbf{S}_{x y} - \rho^2 \ \mathbf{S}_{y y}\right) \mathbf{b} = \mathbf{0} \\
&amp; \left(\mathbf{S}_{x y} \ \mathbf{S}_{y y}^{-1} \ \mathbf{S}_{y x} - \rho^2 \ \mathbf{S}_{x x}\right) \mathbf{a} = \mathbf{0} \ .
\end{aligned}
\]</span> Both equations have the same form and have the same eigenvalues. And, given the eigenvectors for one of these equations, we can find the eigenvectors for the other.</p>
<p><strong>TODO</strong>: Fill in details of canonical correlations</p>
<section id="sec-schooldata-HE" class="level3" data-number="11.9.1"><h3 data-number="11.9.1" class="anchored" data-anchor-id="sec-schooldata-HE">
<span class="header-section-number">11.9.1</span> Example: School data</h3>
<!-- **TODO**: This should be redone, omitting the influential cases, 44, 59 -->
<p>The <code>schooldata</code> dataset analyzed in <a href="10-mlm-review.html#sec-schooldata-mmra" class="quarto-xref"><span>Section 10.5.2</span></a> can also be illuminated by the methods of this chapter. There I fit the multivariate regression model predicting students scores on reading, mathematics and a measure of self-esteem using as predictors measures of parents’ education, occupation, school visits, counseling help with school assignments and number of teachers per school. But I also found two highly influential observations (cases 44, 59; see <a href="10-mlm-review.html#fig-schoolmod-infl" class="quarto-xref">Figure&nbsp;<span>10.16</span></a>) whose effect on the coefficients is rather large; so, I remove them from the analysis here.<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">schooldata</span>, package <span class="op">=</span> <span class="st">"heplots"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">44</span>, <span class="fl">59</span><span class="op">)</span></span>
<span><span class="va">OK</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">schooldata</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="va">bad</span><span class="op">)</span></span>
<span><span class="va">school.mod2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">reading</span>, <span class="va">mathematics</span>, <span class="va">selfesteem</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, </span>
<span>                  data<span class="op">=</span><span class="va">schooldata</span><span class="op">[</span><span class="va">OK</span>, <span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this model, parent’s education and occupation and their visits to the schools were highly predictive of student’s outcomes but their counseling efforts and the number of teachers in the schools did not contribute much. However, the nature of these relationships was largely uninterpreted in that analysis.</p>
<p>Here is where HE plots can help. You can think of this as a way to visualize what is entailed in the coefficients for this model by showing the <em>magnitude</em> of the predictor effects by their size and their relations to the outcome variable by their <em>direction</em>. The table of raw score coefficients isn’t very helpful in this regard.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">school.mod2</span><span class="op">)</span></span>
<span><span class="co">#&gt;             reading mathematics selfesteem</span></span>
<span><span class="co">#&gt; (Intercept)  2.7096       3.561    0.39751</span></span>
<span><span class="co">#&gt; education    0.2233       0.132   -0.01088</span></span>
<span><span class="co">#&gt; occupation   3.3336       4.284    1.79574</span></span>
<span><span class="co">#&gt; visit        0.0101      -0.123    0.20005</span></span>
<span><span class="co">#&gt; counseling  -0.3953      -0.293    0.00868</span></span>
<span><span class="co">#&gt; teacher     -0.1945      -0.360    0.01129</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="#fig-school-heplot1" class="quarto-xref">Figure&nbsp;<span>11.13</span></a> shows the HE plot for reading and mathematics scores in this model, using the default significance scaling.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">school.mod2</span>, </span>
<span>       fill<span class="op">=</span><span class="cn">TRUE</span>, fill.alpha<span class="op">=</span><span class="fl">0.1</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       cex.lab <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       label.pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"top"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"bottom"</span>, <span class="st">"bottom"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-school-heplot1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-school-heplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-school-heplot1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-school-heplot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.13: HE plot for reading and mathematics scores in the multivariate regression model…
</figcaption></figure>
</div>
</div>
</div>
<p>Parent’s occupation and education are both significant in this view, but what is more important is their orientation. Both are positively associated with reading and math scores, but education is somewhat more related to reading than to mathematics. Number of teachers and degree of parental counseling have a similar orientation, with teachers having a greater relation to mathematics scores. Visits to school and number of teachers are not significant in this plot, but both are positively correlated with reading and math and are coincident in the plot. The parent time counseling measure, while also insignificant, tilts in the opposite direction, having different signs for reading and math.</p>
<p>In the <code><a href="https://rdrr.io/r/graphics/pairs.html">pairs()</a></code> plot for all three responses (<a href="#fig-school-heplot2" class="quarto-xref">Figure&nbsp;<span>11.14</span></a>), we see something different in the relations for self-esteem. While occupation has a large positive relation in all the plots in the third row and column, education, counseling and teachers have negative relations in these plots, particularly with mathematics scores.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html">pairs</a></span><span class="op">(</span><span class="va">school.mod2</span>, </span>
<span>      fill<span class="op">=</span><span class="cn">TRUE</span>, fill.alpha<span class="op">=</span><span class="fl">0.1</span>,</span>
<span>      var.cex <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>      cex <span class="op">=</span> <span class="fl">1.3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-school-heplot2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-school-heplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-school-heplot2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-school-heplot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.14: Pairwise HE plots for the three outcome variables in the multivariate regression model …
</figcaption></figure>
</div>
</div>
</div>
</section><section id="canonical-analysis" class="level3" data-number="11.9.2"><h3 data-number="11.9.2" class="anchored" data-anchor-id="canonical-analysis">
<span class="header-section-number">11.9.2</span> Canonical analysis</h3>
<p>With <span class="math inline">\(p = 3\)</span> responses and <span class="math inline">\(q = 5\)</span> predictors there are three possible sets of canonical variables which together account for 100% of the total linear relations between them. <code>heplots::cancor()</code> gives the percentage associated with each of the eigenvalues and the canonical correlations.</p>
<p>For this dataset, the first canonical variates, with Can <span class="math inline">\(R = 0.995\)</span>, accounts for 98.6%, so you might think that that is sufficient. Yet the likelihood ratio tests show that the second set, with Can <span class="math inline">\(R = 0.774\)</span>, is also significant, even though it only accounts for 1.3%.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># bad &lt;- c(44, 59)</span></span>
<span><span class="co"># OK &lt;- (1:nrow(schooldata)) |&gt; setdiff(bad)</span></span>
<span><span class="va">school.can2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://friendly.github.io/candisc/reference/cancor.html">cancor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">reading</span>, <span class="va">mathematics</span>, <span class="va">selfesteem</span><span class="op">)</span> <span class="op">~</span></span>
<span>                        <span class="va">education</span> <span class="op">+</span> <span class="va">occupation</span> <span class="op">+</span> <span class="va">visit</span> <span class="op">+</span> <span class="va">counseling</span> <span class="op">+</span> <span class="va">teacher</span>,</span>
<span>                     data<span class="op">=</span><span class="va">schooldata</span><span class="op">[</span><span class="va">OK</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">school.can2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Canonical correlation analysis of:</span></span>
<span><span class="co">#&gt;   5   X  variables:  education, occupation, visit, counseling, teacher </span></span>
<span><span class="co">#&gt;   with  3   Y  variables:  reading, mathematics, selfesteem </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     CanR CanRSQ    Eigen  percent    cum</span></span>
<span><span class="co">#&gt; 1 0.9946 0.9892 91.41999 98.57540  98.58</span></span>
<span><span class="co">#&gt; 2 0.7444 0.5541  1.24267  1.33994  99.92</span></span>
<span><span class="co">#&gt; 3 0.2698 0.0728  0.07852  0.08466 100.00</span></span>
<span><span class="co">#&gt;                          scree</span></span>
<span><span class="co">#&gt; 1 ****************************</span></span>
<span><span class="co">#&gt; 2                             </span></span>
<span><span class="co">#&gt; 3                             </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Test of H0: The canonical correlations in the </span></span>
<span><span class="co">#&gt; current row and all that follow are zero</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    CanR LR test stat approx F numDF denDF Pr(&gt; F)    </span></span>
<span><span class="co">#&gt; 1 0.995        0.004     67.5    15   166 &lt; 2e-16 ***</span></span>
<span><span class="co">#&gt; 2 0.744        0.413      8.5     8   122 4.1e-09 ***</span></span>
<span><span class="co">#&gt; 3 0.270        0.927      1.6     3    62    0.19    </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The virtue of CCA is that <em>all</em> correlations between the X and Y variables are completely captured in the correlations between the <em>pairs</em> of canonical scores: The <span class="math inline">\(p \times q\)</span> correlations between the sets are entirely represented by the <span class="math inline">\(s = \min(p, q)\)</span> canonical ones. Whether the second dimension is useful here depends on whether it adds some interpretable increment to what is going on in these relations. One could be justifiably happy with an explanation based on the first dimension that accounts for nearly all the total association between the sets.</p>
<p>The class <code>"cancor"</code> object returned by <code><a href="https://friendly.github.io/candisc/reference/cancor.html">cancor()</a></code> contains the canonical coefficients, for which there is a <code><a href="https://rdrr.io/r/stats/coef.html">coef()</a></code> method as in <code><a href="https://friendly.github.io/candisc/reference/candisc.html">candisc()</a></code>, and also a <code><a href="https://friendly.github.io/candisc/reference/cancor.html">scores()</a></code> method to return the scores on the canonical variables, called <code>Xcan1</code>, <code>Xcan2</code>, … and <code>Ycan1</code>, <code>Ycan2</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">school.can2</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cancor"    "names"     "ndim"      "dim"       "coef"     </span></span>
<span><span class="co">#&gt;  [6] "scores"    "X"         "Y"         "weights"   "structure"</span></span>
<span><span class="co">#&gt; [11] "call"      "terms"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can use the <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> method or <code><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot()</a></code> method to visualize and help interpret the results. The <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> method plots the canonical <code>scores$X</code> against the <code>scores$Y</code> for a given dimension (selected by the <code>which</code> argument). The <code>id.n</code> argument gives a way to flag noteworthy observations.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">1.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Can R ="</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">school.can2</span><span class="op">$</span><span class="va">cancor</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     cex <span class="op">=</span> <span class="fl">1.4</span>, pos <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">school.can2</span>, which <span class="op">=</span> <span class="fl">2</span>, </span>
<span>     pch<span class="op">=</span><span class="fl">16</span>, id.n <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cex.lab <span class="op">=</span> <span class="fl">1.5</span>, id.cex <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>     ellipse.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Can R ="</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">school.can2</span><span class="op">$</span><span class="va">cancor</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     cex <span class="op">=</span> <span class="fl">1.4</span>, pos <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="va">op</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-school-can" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-school-can-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-school-can-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-school-can-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.15: Plots of canonical scores for the first two canonical dimensions of the <code>schooldata</code> dataset, omitting the two highly influential cases.
</figcaption></figure>
</div>
</div>
</div>
<p>It is worthwhile to look at an analogous plot of canonical scores for the original dataset including the two highly influential cases. As you can see in <a href="#fig-school-can0" class="quarto-xref">Figure&nbsp;<span>11.16</span></a>, cases 44 and 59 are way outside the range of the rest of the data. Their influence increases the canonical correlation to a near perfect <span class="math inline">\(\rho = 0.997\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">school.can</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://friendly.github.io/candisc/reference/cancor.html">cancor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">reading</span>, <span class="va">mathematics</span>, <span class="va">selfesteem</span><span class="op">)</span> <span class="op">~</span></span>
<span>                       <span class="va">education</span> <span class="op">+</span> <span class="va">occupation</span> <span class="op">+</span> <span class="va">visit</span> <span class="op">+</span> <span class="va">counseling</span> <span class="op">+</span> <span class="va">teacher</span>,</span>
<span>                     data<span class="op">=</span><span class="va">schooldata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">school.can</span>, </span>
<span>     pch<span class="op">=</span><span class="fl">16</span>, id.n <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cex.lab <span class="op">=</span> <span class="fl">1.5</span>, id.cex <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>     ellipse.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Can R ="</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">school.can</span><span class="op">$</span><span class="va">cancor</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     cex <span class="op">=</span> <span class="fl">1.4</span>, pos <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-school-can0" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-school-can0-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-school-can0-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-school-can0-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.16: Plots of canonical scores on the first canonical dimension for the <code>schooldata</code>, including the influential cases, which stand out as so far frome the rest of the observations.
</figcaption></figure>
</div>
</div>
</div>
<p>Plots of canonical scores tell us of the strength of the canonical dimensions, but do not help interpreting the analysis in relation to the original variables. The HE plot version for canonical correlation analysis re-fits a multivariate regression model for the Y variables against the Xs, but substitutes the canonical scores for each, essentially projecting the data into canonical space.</p>
<p><strong>TODO</strong>: Check out signs of structure coefs from <code><a href="https://friendly.github.io/candisc/reference/cancor.html">cancor()</a></code>. Would be better to reflect the vectors for <code>Ycan1</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">school.can2</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>       var.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>       asp <span class="op">=</span> <span class="cn">NA</span>, scale <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>       cex.lab <span class="op">=</span> <span class="fl">1.5</span>, cex <span class="op">=</span> <span class="fl">1.25</span>,</span>
<span>       prefix<span class="op">=</span><span class="st">"Y canonical dimension "</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-school-hecan" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-school-hecan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-school-hecan-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-school-hecan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.17: HE plot for the canonical correlation analysis of the schooldata. Vectors for the variables indicate their correlations with the canonical dimensions.
</figcaption></figure>
</div>
</div>
</div>
<p>The <span style="color: red;">red</span> variable vectors shown in these plots are intended only to show the correlations of Y variables with the canonical dimensions. The fact that they are so closely aligned reflects the fact that the first dimension accounts for nearly all of their associations with the predictors. The orientation of the <span class="math inline">\(\mathbf{H}\)</span> ellipses/lines reflects the projection of those from <a href="#fig-school-heplot2" class="quarto-xref">Figure&nbsp;<span>11.14</span></a> into canonical space</p>
<p>Only their <em>relative lengths</em> and angles with respect to the Y canonical dimensions have meaning in these plots. Relative lengths correspond to proportions of variance accounted for in the Y canonical dimensions plotted; angles between the variable vectors and the canonical axes correspond to the structure correlations. The absolute lengths of these vectors are arbitrary and are typically manipulated by the <code>scale</code> argument to provide better visual resolution and labeling for the variables.</p>
</section></section><section id="mancova-models" class="level2" data-number="11.10"><h2 data-number="11.10" class="anchored" data-anchor-id="mancova-models">
<span class="header-section-number">11.10</span> MANCOVA models</h2>
<p>HE plots for designs containing a collection of quantitative predictors and one or more factors are quite simple in MANCOVA models where the effects are additive, i.e., don’t involve interactions. They are a bit more challenging when you allow <em>separate</em> slopes for groups on all quantitative variables, because there get to be too many terms to usefully display. But these models are more complicated!</p>
<p>If the evidence for heterogeneity of regressions is not very strong, it is still useful to fit the MANCOVA model and display it in an HE plot.</p>
<p>An alternative is to fit separate models for the groups and display these as HE plots. As noted earlier (<a href="10-mlm-review.html#sec-PA-tasks" class="quarto-xref"><span>Section 10.7.1</span></a>), this is not ideal for testing hypotheses, but provides a useful and informative display of the relations between the predictors and responses and the groups effect. I illustrate these approaches for the Rohwer data, encountered in <a href="10-mlm-review.html#sec-PA-tasks" class="quarto-xref"><span>Section 10.7.1</span></a>, below.</p>
<section id="example-rohwer-data" class="level3" data-number="11.10.1"><h3 data-number="11.10.1" class="anchored" data-anchor-id="example-rohwer-data">
<span class="header-section-number">11.10.1</span> Example: Rohwer data</h3>
<p>In <a href="10-mlm-review.html#sec-PA-tasks" class="quarto-xref"><span>Section 10.7.1</span></a> I fit several models for Rohwer’s data on the relations between paired-associate tasks and scholastic performance. The first model was the MANCOVA model testing the difference between the high and low SES groups, controlling for, or taking into account differences on the paired-associate task.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Rohwer.mod1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">SAT</span>, <span class="va">PPVT</span>, <span class="va">Raven</span><span class="op">)</span> <span class="op">~</span> <span class="va">SES</span> <span class="op">+</span> <span class="va">n</span> <span class="op">+</span> <span class="va">s</span> <span class="op">+</span> <span class="va">ns</span> <span class="op">+</span> <span class="va">na</span> <span class="op">+</span> <span class="va">ss</span>, </span>
<span>                 data<span class="op">=</span><span class="va">Rohwer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>HE plots for this model for the pairs (SAT, PPVT) and (SAT, Raven) is shown in <a href="#fig-rohwer-HE-mod1-pairs" class="quarto-xref">Figure&nbsp;<span>11.18</span></a>. The result of an overall test for <em>all</em> predictors, <span class="math inline">\(\mathcal{H}_0 : \mathbf{B} = \mathbf{0}\)</span>, is added to the basic plot using the <code>hypotheses</code> argument.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="fl">5</span><span class="op">)</span>, <span class="st">"#969696"</span><span class="op">)</span></span>
<span><span class="va">covariates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">Rohwer.mod1</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html">pairs</a></span><span class="op">(</span><span class="va">Rohwer.mod1</span>, </span>
<span>       col<span class="op">=</span><span class="va">colors</span>,</span>
<span>       hypotheses<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Regr"</span> <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>       cex<span class="op">=</span><span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>, var.cex <span class="op">=</span> <span class="fl">3</span>,</span>
<span>       lwd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-rohwer-HE-mod1-pairs" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-rohwer-HE-mod1-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-rohwer-HE-mod1-pairs-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rohwer-HE-mod1-pairs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.18: All-pairs HE plot for SAT, PPVT and Raven using the MANCOVA model. The ellipses labeled ‘Regr’ show the test of the overall effect of the quantitative predictors.
</figcaption></figure>
</div>
</div>
</div>
<p>The positive effect of SES on the outcome measures is seen in all pairwise plots: the high SES group is better on all responses. The positive orientation of the <code>Regr</code> ellipses for the covariates shows that the predicted values for all three responses are positively correlated (more so for SAT and PPVT): higher performance on the paired associate tasks, in general, is associated with higher academic performance. The two significant predictors, <code>na</code> and <code>ns</code> are the only ones that extend outside the error ellipses, but their orientations differ.</p>
<section id="homogeneity-of-regression" class="level4 unnumbered"><h4 class="unnumbered anchored" data-anchor-id="homogeneity-of-regression">Homogeneity of regression</h4>
<p>A second model tested the assumption of homogeneity of regression by adding interactions of SES with the PA tasks, allowing separate slopes for the two groups on each of the other predictors.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Rohwer.mod2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">SAT</span>, <span class="va">PPVT</span>, <span class="va">Raven</span><span class="op">)</span> <span class="op">~</span> <span class="va">SES</span> <span class="op">*</span> <span class="op">(</span><span class="va">n</span> <span class="op">+</span> <span class="va">s</span> <span class="op">+</span> <span class="va">ns</span> <span class="op">+</span> <span class="va">na</span> <span class="op">+</span> <span class="va">ss</span><span class="op">)</span>,</span>
<span>                  data<span class="op">=</span><span class="va">Rohwer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This model has 11 terms, excluding the intercept: <code>SES</code>, plus 5 main effects (<span class="math inline">\(x\)</span>s) for the predictors and 5 interactions (slope differences), too many for an understandable display. To visualize this in an HE plot (<a href="#fig-rohwer-HE-mod2" class="quarto-xref">Figure&nbsp;<span>11.19</span></a>), I simplify, by showing the interaction terms <em>collectively</em> by a single ellipse, representing their joint effect, and specified as a linear hypothesis called <code>slopes</code> that picks out the interaction effects.</p>
<p>The argument <code>terms</code> limits the <span class="math inline">\(\mathbf{H}\)</span> ellipses for the right-hand-side of the model which are shown to just those terms specified. The combined effect of the interaction terms is specified as an hypothesis (<code>slopes</code>) testing the interaction terms (which have a “:” in their name). Because SES is “treatment-coded” in this model, the interaction terms reflect the difference in slopes for the high SES group compared to the low.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">(</span><span class="va">coefs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">Rohwer.mod2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "(Intercept)" "SESLo"       "n"           "s"          </span></span>
<span><span class="co">#&gt;  [5] "ns"          "na"          "ss"          "SESLo:n"    </span></span>
<span><span class="co">#&gt;  [9] "SESLo:s"     "SESLo:ns"    "SESLo:na"    "SESLo:ss"</span></span>
<span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="fl">5</span><span class="op">)</span>, <span class="st">"#969696"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">Rohwer.mod2</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">colors</span>, <span class="st">"darkgreen"</span><span class="op">)</span>, </span>
<span>       terms<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SES"</span>, <span class="st">"n"</span>, <span class="st">"s"</span>, <span class="st">"ns"</span>, <span class="st">"na"</span>, <span class="st">"ss"</span><span class="op">)</span>, </span>
<span>       hypotheses<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Regr"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="st">"s"</span>, <span class="st">"ns"</span>, <span class="st">"na"</span>, <span class="st">"ss"</span><span class="op">)</span>,</span>
<span>                       <span class="st">"Slopes"</span> <span class="op">=</span> <span class="va">coefs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">":"</span>, <span class="va">coefs</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.2</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-rohwer-HE-mod2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-rohwer-HE-mod2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-rohwer-HE-mod2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rohwer-HE-mod2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.19: HE plot for SAT and PPVT using the heterogeneous regression model. The ellipse labeled ‘Regr’ shows the test of the covariates combined, and the ellipse labeled ‘slopes’ shows the combined difference in slopes between the two groups.
</figcaption></figure>
</div>
</div>
</div>
</section><section id="separate-models" class="level4 unnumbered"><h4 class="unnumbered anchored" data-anchor-id="separate-models">Separate models</h4>
<p>When there is heterogeneity of regressions, using submodels for each of the groups has the advantage that you can easily visualize the slopes for the predictors in each of the groups, particularly if you overlay the individual HE plots. In this example, I’m using the models <code>Rohwer.sesLo</code> and <code>Rohwer.sesLo</code> fit to each of the groups.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Rohwer.sesLo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">SAT</span>, <span class="va">PPVT</span>, <span class="va">Raven</span><span class="op">)</span> <span class="op">~</span> <span class="va">n</span> <span class="op">+</span> <span class="va">s</span> <span class="op">+</span> <span class="va">ns</span> <span class="op">+</span> <span class="va">na</span> <span class="op">+</span> <span class="va">ss</span>, </span>
<span>                   data<span class="op">=</span><span class="va">Rohwer</span>, subset <span class="op">=</span> <span class="va">SES</span><span class="op">==</span><span class="st">"Lo"</span><span class="op">)</span></span>
<span><span class="va">Rohwer.sesHi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">SAT</span>, <span class="va">PPVT</span>, <span class="va">Raven</span><span class="op">)</span> <span class="op">~</span> <span class="va">n</span> <span class="op">+</span> <span class="va">s</span> <span class="op">+</span> <span class="va">ns</span> <span class="op">+</span> <span class="va">na</span> <span class="op">+</span> <span class="va">ss</span>, </span>
<span>                   data<span class="op">=</span><span class="va">Rohwer</span>, subset <span class="op">=</span> <span class="va">SES</span><span class="op">==</span><span class="st">"Hi"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here I make use of the fact that several HE plots can be overlaid using the option <code>add=TRUE</code> as shown in <a href="#fig-rohwer-HE-lohi" class="quarto-xref">Figure&nbsp;<span>11.20</span></a>. The axis limits may need adjustment in the first plot so that the second one will fit.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">Rohwer.sesLo</span>, </span>
<span>       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span>,               <span class="co"># adjust axis limits</span></span>
<span>       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">40</span>,<span class="fl">110</span><span class="op">)</span>, </span>
<span>       col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span>, </span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>       lwd<span class="op">=</span><span class="fl">2</span>, cex<span class="op">=</span><span class="fl">1.2</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://friendly.github.io/heplots/reference/heplot.html">heplot</a></span><span class="op">(</span><span class="va">Rohwer.sesHi</span>, </span>
<span>       add<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>       col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"brown"</span>, <span class="st">"black"</span><span class="op">)</span>, </span>
<span>       grand.mean<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>       error.ellipse<span class="op">=</span><span class="cn">TRUE</span>,            <span class="co"># not shown by default when add=TRUE</span></span>
<span>       fill <span class="op">=</span> <span class="cn">TRUE</span>, fill.alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>       lwd<span class="op">=</span><span class="fl">2</span>, cex<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-rohwer-HE-lohi" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-rohwer-HE-lohi-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="figs/ch11/fig-rohwer-HE-lohi-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rohwer-HE-lohi-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11.20: Overlaid HE plots for SAT and PPVT, for the low and high SES groups, when each group is fit separately.
</figcaption></figure>
</div>
</div>
</div>
<p>We can readily see the difference in means for the two SES groups (Hi has greater scores on both variables) and it also appears that the slopes of the <code>s</code> and <code>n</code> predictor ellipses are shallower for the High than the Low group, indicating greater relation with the SAT score. As well, the error ellipses show that on these measures, error variation is somewhat smaller in the low SES group.</p>
<!--
## Chapter summary

**TODO**: Insert chapter summary

* This chapter explains and illustrates the use of HE plots for understanding the effects and significance tests of model terms in multivariate linear models for factor predictors (MANOVA), quantitative predictors
(MMRA) and models for mixtures of these types, that I've collectively called MANCOVA models.
-->
</section></section></section><section id="what-we-have-learned" class="level2" data-number="11.11"><h2 data-number="11.11" class="anchored" data-anchor-id="what-we-have-learned">
<span class="header-section-number">11.11</span> What We Have Learned</h2>
<ul>
<li><p><strong>HE plots clarify complex multivariate models into enlightening visualizations</strong> - The HE plot framework brilliantly addresses the interpretability problem of multivariate models by visualizing hypothesis (H) ellipses against error (E) ellipses. Rather than navigate a confusing maze of tables of coefficients and test statistics, with HE plots you can see which effects matter, how they relate to each other, and whether they’re statistically significant. All of these benefits are given in a single, intuitive plot that reveals the geometric structure underlying your multivariate analysis.</p></li>
<li><p><strong>Canonical space is your secret weapon for high-dimensional visualization</strong> - When you have many response variables, canonical discriminant analysis and canonical correlation analysis project the complex multivariate relationships into a lower-dimensional space that captures the essential patterns. This isn’t just dimension reduction—it’s insight amplification, revealing the fundamental directions of variation that matter most while showing how your original variables contribute to these meaningful dimensions.</p></li>
<li><p><strong>Visual hypothesis testing beats p-value hunting every time</strong> - HE plots make hypothesis testing immediate and intuitive: if a hypothesis ellipse extends outside the error ellipse (under significance scaling), the effect is significant. No more scanning tables of p-values or wrestling with multiple comparisons—the geometry tells the story directly. You can even decompose overall effects into meaningful contrasts and see their individual contributions as separate ellipses.</p></li>
<li><p><strong>Ellipse orientations reveal the hidden correlational structure of your effects</strong> - The angles between hypothesis ellipses in HE plots directly show how different predictors relate to your response variables and to each other. When effect ellipses point in similar directions, those predictors have similar multivariate signatures; when they’re orthogonal, they capture independent aspects of variation. This geometric insight goes far beyond what correlation matrices can reveal.</p></li>
<li><p><strong>Multivariate models become tractable through progressive visual summaries</strong> - The chapter demonstrates a powerful visualization strategy: start with scatterplot matrices to see the raw data structure, move to HE plots to understand model effects, then project to canonical space for the clearest possible view of multivariate relationships. Each step preserves the essential information while making it more interpretable, turning the complexity of multivariate analysis into a comprehensible visual narrative.</p></li>
</ul>
<p><strong>Packages</strong>:</p>
<div class="cell" data-layout-align="center">
<pre data-code-line-numbers=""><code>#&gt; Writing packages to  C:/R/Projects/Vis-MLM-book/bib/pkgs.txt
#&gt; 9  packages used here:
#&gt;  broom, candisc, car, carData, dplyr, ggplot2, heplots, knitr, tidyr</code></pre>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list" style="display: none">
<div id="ref-Anderson:35" class="csl-entry" role="listitem">
Anderson, E. (1935). The irises of the <span>Gaspé</span> peninsula. <em>Bulletin of the American Iris Society</em>, <em>35</em>, 2–5.
</div>
<div id="ref-Bartlett1938" class="csl-entry" role="listitem">
Bartlett, M. S. (1938). Further aspects of the theory of multiple regression. <em>Mathematical Proceedings of the Cambridge Philosophical Society</em>, <em>34</em>(1), 33–40. <a href="https://doi.org/10.1017/s0305004100019897">https://doi.org/10.1017/s0305004100019897</a>
</div>
<div id="ref-Bodmer-etal-2021" class="csl-entry" role="listitem">
Bodmer, W., Bailey, R. A., Charlesworth, B., Eyre-Walker, A., Farewell, V., Mead, A., &amp; Senn, S. (2021). The outstanding scientist, r.a. Fisher: His views on eugenics and race. <em>Heredity</em>, <em>126</em>(4), 565–576. <a href="https://doi.org/10.1038/s41437-020-00394-6">https://doi.org/10.1038/s41437-020-00394-6</a>
</div>
<div id="ref-Fisher:1936" class="csl-entry" role="listitem">
Fisher, R. A. (1936). The use of multiple measurements in taxonomic problems. <em>Annals of Eugenics</em>, <em>7</em>(2), 179–188. <a href="https://doi.org/10.1111/j.1469-1809.1936.tb02137.x">https://doi.org/10.1111/j.1469-1809.1936.tb02137.x</a>
</div>
<div id="ref-R-candisc" class="csl-entry" role="listitem">
Friendly, M., &amp; Fox, J. (2025). <em>Candisc: Visualizing generalized canonical discriminant and canonical correlation analysis</em>. <a href="https://github.com/friendly/candisc/">https://github.com/friendly/candisc/</a>
</div>
<div id="ref-Friendly-etal:ellipses:2013" class="csl-entry" role="listitem">
Friendly, M., Monette, G., &amp; Fox, J. (2013). Elliptical insights: Understanding statistical methods through elliptical geometry. <em>Statistical Science</em>, <em>28</em>(1), 1–39. <a href="https://doi.org/10.1214/12-STS402">https://doi.org/10.1214/12-STS402</a>
</div>
<div id="ref-DeSilva2020" class="csl-entry" role="listitem">
Gayan De Silva. (2020). <em>Exploring the world of artificial neural networks -a beginner’s overview</em>. <a href="https://doi.org/10.13140/RG.2.2.14790.14406">https://doi.org/10.13140/RG.2.2.14790.14406</a>
</div>
<div id="ref-Gittins:85" class="csl-entry" role="listitem">
Gittins, R. (1985). <em>Canonical analysis: A review with applications in ecology</em>. Springer-Verlag.
</div>
<div id="ref-Hotelling1936" class="csl-entry" role="listitem">
Hotelling, H. (1936). Relations between two sets of variates. <em>Biometrika</em>, <em>28</em>(3/4), 321. <a href="https://doi.org/10.2307/2333955">https://doi.org/10.2307/2333955</a>
</div>
<div id="ref-Huang2019" class="csl-entry" role="listitem">
Huang, F. L. (2019). <span>MANOVA</span>: A procedure whose time has passed? <em>Gifted Child Quarterly</em>, <em>64</em>(1), 56–60. <a href="https://doi.org/10.1177/0016986219887200">https://doi.org/10.1177/0016986219887200</a>
</div>
</div>
</section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>For example, Megan Stodel in a blog post <a href="https://www.meganstodel.com/posts/no-to-iris/">Stop using iris</a> says, “<em>It is clear to me that knowingly using work that was itself used in pursuit of racist ideals is totally unacceptable.</em>” A Reddit discussion on this topic, <a href="https://www.reddit.com/r/MachineLearning/comments/ii9urh/d_is_it_socially_acceptable_to_use_the_iris/">Is it socially acceptable to use the Iris dataset?</a> has some interesting replies.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>Recall that <span class="math inline">\(R^2\)</span> for a linear model is the the proportion of variation in the response that is explained by the model, calculated as <span class="math inline">\(R^2 = \text{SS}_H / \text{SS}_T = \text{SS}_H / (\text{SS}_H + \text{SS}_E )\)</span>. For a multivariate model, these are obtained from the diagonal elements of <span class="math inline">\(\mathbf{H}\)</span> and <span class="math inline">\(\mathbf{E}\)</span>.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>That the <span class="math inline">\(\mathbf{H}\)</span> ellipses for the contrasts subtend that for the overall test of <code>Species</code> is no accident. In fact, this is true in <span class="math inline">\(p\)</span>-dimensional space for <em>any</em> linear hypothesis, and orthogonal contrasts have the additional geometric property that they form <em>conjugate axes</em> for the overall <span class="math inline">\(\mathbf{H}\)</span> ellipsoid relative to the <span class="math inline">\(\mathbf{E}\)</span> ellipsoid <span class="citation" data-cites="Friendly-etal:ellipses:2013">(<a href="95-references.html#ref-Friendly-etal:ellipses:2013" role="doc-biblioref">Friendly et al., 2013</a>)</span>.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p><code><a href="https://friendly.github.io/candisc/reference/candiscList.html">candiscList()</a></code> performs a generalized canonical discriminant analysis for <em>all</em> terms in a multivariate linear model, returning a list of the results for each factor.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>If significance scaling was used the interpretation of the canonical HE plot plot would the same as before: if the hypothesis ellipse extends beyond the error ellipse, then that dimension is significant.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>An alternative to fitting the model removing specific cases deemed troublesome is to use a <strong>robust</strong> method, such as <code>heplots::roblm()</code> which uses re-weighted least squares to down-weight observations with large residuals or other problems.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./10-mlm-review.html" class="pagination-link" aria-label="Multivariate Linear Models">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Multivariate Linear Models</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./12-eqcov.html" class="pagination-link" aria-label="Visualizing Equality of Covariance Matrices">
        <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Visualizing Equality of Covariance Matrices</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/friendly/vis-MLM-book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer><script src="site_libs/quarto-contrib/line-highlight-1.0.0/line-highlight.js" defer="true"></script>
</body></html>